---
title: "Final Seq"
output:
  html_document:
    df_print: paged
---

install.packages('devtools')
devtools::install_github(repo = 'satijalab/seurat', ref = 'develop', INSTALL_opts = '--no-lock')

```{r setup=T}


library(Seurat)
library(stringr)
library(data.table)
library(tidyverse)
library(clusterProfiler)
library(org.Hs.eg.db)
library(ggplot2)
library(cowplot)
library(SingleR)
library(xlsx)
library(writexl)
library(dplyr)
library(ggrepel)
library(svglite)

color.control = "#8B9693"
color.symptomatic = "#FE4988"
color.asymptomatic = "#2C9B5C"

color.tp1 = "#FFDFFF"
color.tp2 = "#FFADEC"
color.tp3 = "#FF7BBA"

dsCols = c("CONTROL"=color.control, "ASYMPTOMATIC"=color.asymptomatic, "SYMPTOMATIC"=color.symptomatic)
mpCols = c("1"=color.tp1, "2"=color.tp2, "3"=color.tp3)

source("sbs_plot.R")


knitr::opts_chunk$set(echo = TRUE, rows.print=25)

```

```{r}

save_plot = function(plotobj,outname, fig.width, fig.height)
{
  print(paste(outname, fig.width, fig.height))
  
  fname=paste(outname, "png", sep=".")
  print(paste("Saving to file", fname))
  png(filename=fname, width = fig.width, height = fig.height, units = 'in', res = 300)#width = fig.width*100, height=fig.height*100)
  plot(plotobj)
  dev.off()
  
  fname=paste(outname, "pdf", sep=".")
  print(paste("Saving to file", fname))
  pdf(file=fname, width = fig.width, height=fig.height)
  plot(plotobj)
  dev.off()
  

  fname=paste(outname, "svg", sep=".")
  print(paste("Saving to file", fname))
  svglite::svglite(file = fname, width = fig.width, height = fig.height)
  plot(plotobj)
  dev.off()
  
  return(plotobj)
}

```


# Initialization & Integration


```{r rows.print=36}

sampleDF = read.csv("samples2.df", sep="\t")
sampleDF

```


```{r}
files <- list.files(path="/mnt/f/covid_sc/calls2_mod/", pattern="*.h5", full.names=T, recursive=TRUE)

files = files[grepl("raw_feature_bc_matrix", files)]

allfiles.raw = list()
allABs.raw = list()

for (file in files)
{
  samplename = str_split(dirname(file), "/")[[1]][7]
  foldername = dirname(file)
  
  print(paste(samplename, foldername))
  
  h5file = Read10X_h5(file, use.names=T, unique.features = TRUE)
  
  exprData = h5file$`Gene Expression`
  
  if (samplename %in% c("20094_0010_A_B", "20094_0011_A_B"))
  {
    next
  }
    
  allfiles.raw[[samplename]] = exprData

}


```


```{r}

files <- list.files(path="/mnt/f/covid_sc/calls2_mod/", pattern="*.h5", full.names=T, recursive=TRUE)

files = files[grepl("filtered_feature_bc_matrix", files)]

allfiles.filtered = list()
allABs.filtered = list()

for (file in files)
{
  samplename = str_split(dirname(file), "/")[[1]][7]
  foldername = dirname(file)
  
  print(paste(samplename, foldername))
  
  h5file = Read10X_h5(file, use.names=T, unique.features = TRUE)
  
  exprData = h5file$`Gene Expression`
  
  if (samplename %in% c("20094_0010_A_B", "20094_0011_A_B"))
  {
    next
  }
    
  allfiles.filtered[[samplename]] = exprData
}

print(names(allfiles.filtered))

```

```{r rows.print=15}

statDF = data.frame(matrix(ncol=5,nrow=0, dimnames=list(NULL, c("FilteredAll", "FilteredFiltered", "RawAll", "RawFiltered", "RawFilteredSeurat"))))


toCellPrefix = function(x) {
    datasetprefix = x
    datasetprefix = str_replace_all(datasetprefix, "[./]", "")
    datasetprefix = str_split(datasetprefix, "_")[[1]][1]
    
    return(paste(datasetprefix, sep=""))
}

makesum = function(a, suffix)
{
  out = {}
  out["sum"] = sum(a)

  return(out)
}

makeSeuratObj = function(matrix, proj, minUMIs, plots)
{
      obj = CreateSeuratObject(matrix, project=proj)
    print("Renaming Cells")
    obj <- RenameCells(obj, add.cell.id=proj)
    
    print(paste("Seurat obj project", obj@project.name))
    
    #obj[["percent.mtrp"]] <- PercentageFeatureSet(obj, pattern = "^mt-|^Rps|^Rpl")
    
    mtPattern = "^MT-"
    rplPattern = "^RPL"
    rpsPattern = "^RPS"
    rpPattern = "^RPS|^RPL"
    
    selGenes = rownames(obj)[grepl(rownames(obj), pattern=mtPattern)]
    print(paste("Got a total of mt-Genes:", length(selGenes), paste0(head(unlist(selGenes)), collapse=", ")))
    
    selGenes = rownames(obj)[grepl(rownames(obj), pattern=rplPattern)]
    print(paste("Got a total of Rpl-Genes:", length(selGenes), paste0(head(unlist(selGenes)), collapse=", ")))
    
    selGenes = rownames(obj)[grepl(rownames(obj), pattern=rpsPattern)]
    print(paste("Got a total of Rps-Genes:", length(selGenes), paste0(head(unlist(selGenes)), collapse=", ")))
    
    selGenes = rownames(obj)[grepl(rownames(obj), pattern=rpPattern)]
    print(paste("Got a total of Rp-Genes:", length(selGenes), paste0(head(unlist(selGenes)), collapse=", ")))
    
    obj[["percent.mt"]] <- PercentageFeatureSet(obj, pattern = mtPattern)
    obj[["percent.rpl"]] <- PercentageFeatureSet(obj, pattern = rplPattern)
    obj[["percent.rps"]] <- PercentageFeatureSet(obj, pattern = rpsPattern)
    obj[["percent.rp"]] <- PercentageFeatureSet(obj, pattern = rpPattern)
    
    if (plots)
    {
      plot1 <- FeatureScatter(obj, feature1 = "nCount_RNA", feature2 = "percent.mt")
      plot2 <- FeatureScatter(obj, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
      show(plot1 + scale_x_continuous(n.breaks = 20) + scale_y_continuous(n.breaks = 20))

    }
    

    # mt content: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6072887/
    obj <- subset(obj, subset = nFeature_RNA > 200 & nFeature_RNA < 6000 & nCount_RNA > minUMIs & percent.mt < 15 & percent.rp < 40)
    
    return(obj)
}
  

objlist = list()

for (x in names(allfiles.filtered))
{

  
  print(x)

  matrix = allfiles.raw[[x]]
  origSize = dim(matrix)[2]
  
  expt.r = as(matrix, "dgTMatrix")
  expt.df = data.frame(r = expt.r@i + 1, c = expt.r@j + 1, x = expt.r@x)
  
  DT <- data.table(expt.df)
  res = DT[, as.list(makesum(x, sampleSuffix)), by = c]

  resKeep = res[res$sum > 200,]$c
  matrix = matrix[, resKeep]
    rawFilteredSize = dim(matrix)[2]
  
    rawObj = makeSeuratObj(matrix, x, 200, FALSE)
    rawFilteredCols = ncol(rawObj)
    rawObj = NULL
    
    
    matrix = allfiles.filtered[[x]]
    filteredAll = dim(matrix)[2]
    filteredObj = makeSeuratObj(matrix, x, 1000, TRUE)
    filteredFilteredCols = ncol(filteredObj)
    
    
    filteredObj <- NormalizeData(filteredObj, verbose = FALSE)
    filteredObj <- FindVariableFeatures(filteredObj, verbose = FALSE)
    
    objlist[[x]] = filteredObj
    
    statDF <- rbind(statDF, data.frame("FilteredAll"=filteredAll, "FilteredFiltered"=filteredFilteredCols, "RawAll"=origSize, "RawFiltered"=rawFilteredSize, "RawFilteredSeurat"=rawFilteredCols))
    
}

DT = NULL
expt.df = NULL
expt.r = NULL
res = NULL
matrix = NULL
rownames(statDF) = names(objlist)
statDF
```

```{r}

features <- SelectIntegrationFeatures(object.list = objlist)
rpmtFeatures = features[grep("^Rps|^Rpl|Mt-", features)]

noRPMTFeatures = features[grep("^Rps|^Rpl|Mt-", features, invert = TRUE)][1:2000]
length(noRPMTFeatures) 

objlist <- lapply(X = objlist, FUN = function(x) {
    print(x)
    suppressWarnings( x <- SCTransform(x, vars.to.regress = "percent.mt", verbose=FALSE) )
})

```

```{r}
print("cells per experiment")
print(mapply(sum, lapply(objlist, function(x) {dim(x)[2]})))
print("total cells")
print(sum(mapply(sum, lapply(objlist, function(x) {dim(x)[2]}))))

```

```{r}
save.image("/mnt/d/dev/data/Rprojs/covidSC/before_integration.final.RData", compress = FALSE)
```


```{r}

allABs.raw = NULL
allfiles.raw = NULL

allABs.filtered = NULL

```

```{r}

objlist.features <- SelectIntegrationFeatures(object.list = objlist, nfeatures = 2000)

options(future.globals.maxSize= 4000*1024^2)

objlist.list <- PrepSCTIntegration(object.list = objlist, anchor.features = objlist.features, verbose = FALSE)

objlist.anchors <- FindIntegrationAnchors(object.list = objlist.list, normalization.method = "SCT", anchor.features = objlist.features, verbose = FALSE)

save.image("/mnt/d/dev/data/Rprojs/covidSC/before_integrate.final.RData", compress = FALSE)

```

```{r}
obj.integrated <- IntegrateData(anchorset = objlist.anchors, normalization.method = "SCT", verbose = T)

save.image("/mnt/d/dev/data/Rprojs/covidSC/integrated.final.RData", compress = FALSE)

```


# Base Analysis

```{r PCA UMAP}
obj.integrated <- RunPCA(obj.integrated, verbose = FALSE)
obj.integrated <- RunUMAP(obj.integrated, dims = 1:30)

```

```{r fig.width=20, fig.height=5}
DimPlot(obj.integrated, split.by="ident")
```

```{r fig.width=12, fig.height=5}
DimPlot(obj.integrated )
```

```{r}

obj.integrated <- FindNeighbors(obj.integrated)
obj.integrated <- FindClusters(obj.integrated, resolution = 0.5)

obj.integrated$idents = Idents(obj.integrated)

```

```{r}
table(obj.integrated$orig.ident)
sum(table(obj.integrated$orig.ident))
```


```{r fig.height=40, fig.width=10}
DimPlot(obj.integrated, pt.size = 0.005, ncol=1, split.by="ident")
```

```{r fig.height=7, fig.width=15}
DimPlot(obj.integrated, pt.size = 0.005, ncol=1)
```

```{r}
save.image("/mnt/d/dev/data/Rprojs/covidSC/integrated.final.RData", compress = FALSE)
```

# Final QC

```{r}
DefaultAssay(obj.integrated) = "RNA"
obj.integrated[["percent.mt"]] = PercentageFeatureSet(obj.integrated, pattern = "^MT-")
```


```{r fig.height=8, fig.width=15}
VlnPlot(obj.integrated, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3, pt.size = 0.001, group.by = "orig.ident")
```

```{r fig.height=8, fig.width=15}
plot1 <- FeatureScatter(obj.integrated, feature1 = "nCount_RNA", feature2 = "percent.mt")
plot2 <- FeatureScatter(obj.integrated, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
plot1 + scale_x_continuous(n.breaks = 20) + scale_y_continuous(n.breaks = 20)
```

```{r fig.height=8}
plot1 = FeaturePlot(subset(obj.integrated, percent.mt < 15), c("percent.mt"))
plot2 = FeaturePlot(subset(obj.integrated, percent.mt < 10), c("percent.mt"))

plot1+plot2
```



# Cell Annotation

```{r rows.print=36}
sampleDF
```

```{r}
cellIdents = Idents(obj.integrated)
cellList = as.character(cellIdents)
names(cellList) = names(cellIdents)

new_descs = list()

for (x in c("mpoint", "library", "disease_state", "samplename"))
{
  nvec = vector(mode="character", length=length(cellList))
  names(nvec) = names(cellList)
  new_descs[[x]] = nvec
}

for (row in 1:nrow(sampleDF))
{
  srow = sampleDF[row, ]

  
  sampleName = srow$SampleName
  
  cellNames = names(cellList)
  
  
  allSampleCells = cellNames[grep(x = cellNames, pattern = paste("^", sampleName, sep=""))]
  dfRowCells = allSampleCells

  new_descs$mpoint[dfRowCells] = srow$Mpoint
  new_descs$library[dfRowCells] = srow$Library
  new_descs$samplename[dfRowCells] = srow$SampleName
  
  diseaseState = srow$Condition
  new_descs$disease_state[dfRowCells] = diseaseState
  
}



print(paste("mpoint", paste(unique(new_descs$mpoint), collapse=",") ,sep=" "))
print(paste("library", paste(unique(new_descs$library), collapse=",") ,sep=" "))
print(paste("disease_state", paste(unique(new_descs$disease_state), collapse=",") ,sep=" "))
print(paste("samplename", paste(unique(new_descs$samplename), collapse=",") ,sep=" "))

print(paste("mpoint", length(new_descs$mpoint[new_descs$mpoint == ""])))
print(paste("library", length(new_descs$library[new_descs$library == ""])))
print(paste("disease_state", length(new_descs$disease_state[new_descs$disease_state == ""])))
print(paste("samplename", length(new_descs$samplename[new_descs$samplename == ""])))


obj.integrated$mpoint = new_descs$mpoint
obj.integrated$library = new_descs$library
obj.integrated$disease_state = new_descs$disease_state
obj.integrated$samplename = new_descs$samplename

print(paste("mpoint", length(obj.integrated$mpoint), length(new_descs$mpoint), length(intersect(obj.integrated$mpoint, new_descs$mpoint))))
print(paste("library", length(obj.integrated$library), length(new_descs$library), intersect(obj.integrated$library, new_descs$library)))
print(paste("disease_state", length(obj.integrated$disease_state), length(new_descs$disease_state), intersect(obj.integrated$disease_state, new_descs$disease_state)))
print(paste("samplename", length(obj.integrated$samplename), length(new_descs$samplename), intersect(obj.integrated$samplename, new_descs$samplename)))


no_cond_cells = names(obj.integrated$disease_state[obj.integrated$disease_state == ""])

control_cells = names(obj.integrated$disease_state[obj.integrated$disease_state == "CONTROL"])
symptomatic_cells = names(obj.integrated$disease_state[obj.integrated$disease_state == "SYMPTOMATIC"])
asymptomatic_cells = names(obj.integrated$disease_state[obj.integrated$disease_state == "ASYMPTOMATIC"])
disease_cells = names(obj.integrated$disease_state[obj.integrated$disease_state == "SYMPTOMATIC"])

print(paste("no_cond_cells", length(no_cond_cells)))
print(paste("control_cells", length(control_cells)))
print(paste("covacta_cells", length(covacta_cells)))
print(paste("symptomatic_cells", length(symptomatic_cells)))
print(paste("asymptomatic_cells", length(asymptomatic_cells)))
print(paste("disease_cells", length(disease_cells)))

for (measureData in c("mpoint", "library", "disease_state", "samplename"))
{
  
  measureFeatures = unique(new_descs[[measureData]])
  
  for (mf in measureFeatures)
  {
    
    objLabels = colnames(obj.integrated)[obj.integrated[[measureData]] == mf]
    newDescLabels = names(new_descs[[measureData]][new_descs[[measureData]] == mf])
    
    print(paste(measureData,
                mf,
                length(objLabels),
                length(newDescLabels),
                length(intersect(objLabels, newDescLabels))))

  }
  
}


```

```{r}
table(obj.integrated$disease_state)
```


```{r fig.height=20, fig.width=10}
DimPlot(obj.integrated, pt.size = 0.005, ncol=1, split.by="disease_state")
```

```{r fig.height=20, fig.width=10}
DimPlot(obj.integrated, pt.size = 0.005, ncol=2, split.by="orig.ident", group.by="disease_state")
```

## Final Cosmetics

```{r}
obj.integrated
```


```{r}

obj.integrated.orig = obj.integrated

removeCells_nocond = names(obj.integrated.orig$library[obj.integrated.orig$library == ""])

print(paste("removeCells_nocond", length(removeCells_nocond)))

obj.integrated = subset(obj.integrated.orig, cells = c(removeCells_nocond), invert=TRUE)

obj.integrated

```

```{r}
obj.integrated.orig
obj.integrated
```


```{r}
DimPlot(obj.integrated, label=T)
```


```{r}
save.image("/mnt/f/dev/data/Rprojs/covidSC/analysis_start.final.RData", compress = FALSE)
saveRDS(obj.integrated.orig, "/mnt/f/dev/data/Rprojs/covidSC/obj_integrated_orig.Rds")
saveRDS(obj.integrated, "/mnt/f/dev/data/Rprojs/covidSC/obj_integrated.Rds")
```


# Analysis

## Marker Analysis


```{r eval=F}

makesummary = function(a, suffix)
{
  out = {}
  out["num"] = length(a)
  
  if (length(a) == 0)
  {
    f = c(0,0,0,0,0)
    meanA = 0
  } else {
    f = fivenum(a)
    meanA = mean(a)
  }

  out["min"] = f[1]
  out["lower_hinge"] = f[2]
  out["median"] = f[3]
  out["upper_hinge"] = f[4]
  out["max"] = f[5]
  out["mean"] = meanA
  
  names(out) = paste(names(out), suffix, sep=".")
  
  return(out)
}

getExprData = function(markerObj, markerCells, sampleSuffix, assay="RNA")
{
  expTable = GetAssayData(object = subset(x=markerObj, cells=markerCells), slot = "data", assay=assay)
  allgenes = rownames(expTable)
  cellnames = colnames(expTable)

  expt.r = as(expTable, "dgTMatrix")
  expt.df = data.frame(r = expt.r@i + 1, c = expt.r@j + 1, x = expt.r@x)

  DT <- data.table(expt.df)
  res = DT[, as.list(makesummary(x, sampleSuffix)), by = r]
  res[[paste("anum", sampleSuffix, sep=".")]] = length(cellnames)
  res$gene = allgenes[res$r]
  
  res = res[,r:=NULL]
  
  return(res)
}

getDEXpressionDF = function ( scdata, markers, assay="SCT" )
{

outDF = NULL
DefaultAssay(object=scdata) = assay  
clusterIDs = as.character(sort(unique(Idents(scdata))))

scCells = Idents(scdata)
scCells = names(scCells)
scCells = unlist(as.character(scCells))

for (clusterID in clusterIDs){
    
    print(clusterID)
    
    cellIdents = Idents(scdata)
    cellIdents.c = names(cellIdents[cellIdents == clusterID])
    cellIdents.c = unlist(lapply(cellIdents.c, as.character))  
    
    cellIdents.bg = setdiff(unlist(lapply(names(cellIdents), as.character)), cellIdents.c)
    
    expvals = getExprData(scdata, cellIdents.c, "cluster", assay=assay)
    expvals.bg = getExprData(scdata, cellIdents.bg, "bg", assay=assay)

    modmarkers = markers[[clusterID]]
    modmarkers$gene = rownames(modmarkers)
    
    markerdf = as.data.frame(modmarkers)
    
    if ((nrow(markerdf) > 0) && (nrow(expvals) > 0))
    {
      expvals = merge(markerdf, expvals, all.x=T, by.x="gene", by.y = "gene")  
    }
    
    if ((nrow(expvals) > 0) && (nrow(expvals.bg) > 0))
    {
      expvals = merge(expvals, expvals.bg, all.x=T, by.x="gene", by.y = "gene")  
    }
    
    expvals = as.data.frame(cbind(clusterID, expvals))
    
    if (!is.data.frame(outDF) || nrow(outDF)==0)
    {
    outDF = expvals
    } else {
    outDF = as.data.frame(rbind(outDF, expvals))
    }
    
}

return(outDF)

}

makeDEResults = function(inobj, assay="SCT", test="wilcox")
{
  clusterIDs = as.character(sort(unique(Idents(inobj))))
  
  retList = list()
  
  for(clusterID in clusterIDs)
  {
  
  
      cellIdents = Idents(inobj)
      cellIdents.c = names(cellIdents[cellIdents == clusterID])
      cellIdents.c = unlist(lapply(cellIdents.c, as.character))
  
      print(paste("Processing cluster", clusterID, "with a total of", length(cellIdents.c), "cells"))
  
      deMarkers = FindMarkers(inobj, assay=assay, ident.1 = cellIdents.c, test.use=test)
  
  
      retList[[clusterID]] = deMarkers
  
  }
  
  return(retList)

}


compareCells = function(scdata, cellsID1, cellsID2, suffix1, suffix2, prefix="cluster", test="t", assay="RNA", outfolder="./", all=FALSE)
{
    logfc.threshold = 0.25
    
    if (all==TRUE)
    {
    logfc.threshold = 0.01  
    }
    
    print(paste("Missing cells", cellsID1[!cellsID1 %in% colnames(obj.integrated)]))
    print(paste("Missing cells", cellsID2[!cellsID2 %in% colnames(obj.integrated)]))
    
    print(paste("Selected Cells 1", length(cellsID1)))
    print(paste("Selected Cells 2", length(cellsID2)))
    
    if ((length(cellsID1) < 3) || (length(cellsID2) < 3))
    {
      return(data.frame())
    }

    markers = FindMarkers(scdata, assay=assay, ident.1 = cellsID1, ident.2 = cellsID2, test.use=test, logfc.threshold=logfc.threshold)
    
    outvalues1 = getExprData(scdata, cellsID1, suffix1, assay=assay)
    outvalues2 = getExprData(scdata, cellsID2, suffix2, assay=assay) 
    
    
    markers$gene = rownames(markers)
    joinedData = merge(markers, outvalues1, by="gene", all=T)
    joinedData = merge(joinedData, outvalues2, by="gene", all=T)  
    
    joinedData = joinedData[!is.na(joinedData$p_val),]
    
    outfile = paste(outfolder, "/", prefix, ".", suffix1, "_", suffix2, ".tsv", sep="")
    message(outfile)
    write.table(joinedData, file=outfile, row.names = F,  quote=FALSE, sep='\t')
    
    outfile = paste(outfolder, "/", prefix, ".", suffix1, "_", suffix2, ".xlsx", sep="")
    message(outfile)
    write.xlsx(joinedData, file=outfile, row.names = F)
    
    return(joinedData)
}


```

```{r}
DefaultAssay(obj.integrated) = "RNA"
```


perform DE analysis on the whole object

```{r}
deResTT = makeDEResults(obj.integrated, assay="RNA", test="t")
exprdfTT = getDEXpressionDF(obj.integrated, deResTT, assay="RNA")
write.table(exprdfTT, "obj.integrated.final.de.t.tsv", sep="\t", row.names=F, quote = F)
write.xlsx(exprdfTT, "obj.integrated.final.de.t.xlsx", row.names = F)

```

```{r}
FeaturePlot(obj.integrated, features = c("CD8A"), order = T)

```


```{r}
exprdfTT

```


```{bash}
#wget https://raw.githubusercontent.com/mjoppich/scrnaseq_celltype_prediction/master/analyseMarkers.py
```


```{bash}
/usr/bin/python3 analyseMarkers.py -i obj.integrated.final.de.t.tsv --expressing-cell-count "num.cluster" --cluster-cell-count "anum.cluster" --expr-mean "mean.cluster" --organs "Immune system" -n 3
```

```{bash}
/usr/bin/python3 analyseMarkers.py -i obj.integrated.final.de.t.tsv --expressing-cell-count "num.cluster" --cluster-cell-count "anum.cluster" --expr-mean "mean.cluster" --organs "Immune system" --seurat

```

Cluster 4 is ADGRE1- ==> no macrophage, also possibly no monocyte.
Cluster 11 are ADGRE1+ !

```{r}

new.cluster.ids_new <- c("Monocytes;Immune system","NK cells;Immune system","NK cells;Immune system","NK cells;Immune system","Monocytes;Immune system","T cells;Immune system","B cells naive;Immune system","B cells memory;Immune system","T cells;Immune system","NK cells;Immune system","Gamma delta T cells;Immune system","Monocytes;Immune system","B cells memory;Immune system","NK cells;Immune system","Dendritic cells;Immune system","Macrophages;Immune system","Plasmacytoid dendritic cells;Immune system","Plasma cells;Immune system")

new.cluster.ids_pred <- c("Monocytes;Immune system","NK cells;Immune system","NK cells;Immune system","NK cells;Immune system","Monocytes;Immune system","T cells;Immune system","B cells naive;Immune system","B cells memory;Immune system","T cells;Immune system","NK cells;Immune system","Gamma delta T cells;Immune system","Monocytes;Immune system","B cells memory;Immune system","NK cells;Immune system","Dendritic cells;Immune system","Macrophages;Immune system","Plasmacytoid dendritic cells;Immune system","Plasma cells;Immune system")

orignames = Idents(obj.integrated)
names(new.cluster.ids) <- levels(orignames)
levels(orignames) = new.cluster.ids
obj.integrated$cellnames = orignames

new.cluster.ids <- c("Monocytes;Immune system","T cells;Immune system","NK cells;Immune system","NK cells;Immune system","T cells;Immune system","T cells;Immune system","B cells;Immune system","B cells;Immune system","T cells;Immune system","NK cells;Immune system","Gamma delta T cells;Immune system","Monocytes;Immune system","B cells;Immune system","T cells;Immune system","Dendritic cells;Immune system","Monocytes;Immune system","Plasmacytoid dendritic cells;Immune system","Plasma cells;Immune system")

orignames = Idents(obj.integrated)
names(new.cluster.ids) <- levels(orignames)
levels(orignames) = new.cluster.ids
obj.integrated$unified_cellnames = orignames

new.cluster.ids <- c("monocytes","tcell","nk","nk","tcell","tcell","bcell","bcell","tcell","nk","gdtcell","monocytes","bcell","tcell","dendritic","monocytes","pdc","plasma")

#new.cluster.ids <- c(1,2,2,2,3,3,4,4,3,2,5,1,4,2,6,1,7,8)

orignames = Idents(obj.integrated)
names(new.cluster.ids) <- levels(orignames)
levels(orignames) = new.cluster.ids
obj.integrated$shortcellnames = orignames

```

```{r fig.width=12, fig.height=5}
DimPlot(obj.integrated, group.by = "idents", label=T)
```

```{r fig.width=24, fig.height=8}

FeaturePlot(obj.integrated, features = c("NCAM1", "CD3D"), ncol = 3, order=T)

```


```{r fig.width=12, fig.height=5}
DimPlot(obj.integrated, group.by = "unified_cellnames", label=T)
```

```{r fig.width=36, fig.height=8}
cowplot::plot_grid(
  DimPlot(obj.integrated, group.by = "ident", label=T),
  DimPlot(obj.integrated, group.by = "cellnames", label=T),
  DimPlot(obj.integrated, group.by = "unified_cellnames", label=T),
  ncol = 3
)

```


```{r fig.width=15, fig.height=8}
  DimPlot(obj.integrated, group.by = "cellnames", label=T)
```

```{r fig.width=15, fig.height=8}

umap_cellnames_plot = DimPlot(obj.integrated, group.by = "unified_cellnames", label=T)

save_plot(umap_cellnames_plot, outname = "todoplots/2/umap_cellnames", fig.width=15, fig.height=8)
```

```{r fig.width=30, fig.height=8}

umap_cellnames_bytime=plot = DimPlot(subset(obj.integrated, cells=unique(c(cells.asymptomatic, cells.symptomatic))), group.by = "unified_cellnames", split.by = "mpoint", label=T)
umap_cellnames_bytime
save_plot(umap_cellnames_bytime, outname = "todoplots/3/umap_cellnames_bytime", fig.width=15, fig.height=8)
```


```{r fig.width=15, fig.height=6}

umap_cellnames_plot = DimPlot(subset(obj.integrated, cells=cells.monocyte), group.by = "unified_cellnames", split.by = "disease_state", label=T)

save_plot(umap_cellnames_plot, outname = "todoplots/14/umap_cellnames_by_ds", fig.width=15, fig.height=6)

umap_cellnames_plot = DimPlot(subset(obj.integrated, cells=cells.monocyte), group.by = "unified_cellnames", split.by = "mpoint", label=T)

save_plot(umap_cellnames_plot, outname = "todoplots/14/umap_cellnames_by_tp", fig.width=15, fig.height=6)
```

```{r}

markers.use.tt=subset(exprdfTT ,avg_logFC>0&p_val_adj<0.01&!startsWith(gene, "MT-")&!startsWith(gene, "RP"))

finalMarkers.use.tt = markers.use.tt %>% group_by(clusterID) %>% arrange(desc(abs(avg_logFC), desc(pct.1)), .by_group=TRUE) %>% dplyr::slice(1:10)
finalMarkers.use.tt

finalMarkers.use.tt.sorted = markers.use.tt %>% group_by(clusterID) %>% arrange(desc(abs(avg_logFC)), desc(pct.1), .by_group=TRUE)

write_xlsx(finalMarkers.use.tt.sorted, "obj.integrated.final.de.t.top.xlsx")
write_xlsx(finalMarkers.use.tt.sorted, "todoplots/4/obj.integrated.final.de.t.top.xlsx")

```

```{r}
mergeLists = function(lolists)
{
  moddfs = lapply(names(lolists), function(x) {
    odf = data.frame(gene=lolists[[x]]$gene, p_val_adj=lolists[[x]]$p_val_adj, avg_logFC=lolists[[x]]$avg_logFC)
    odf$clusterID = x
    
    return(odf)
  })
  
  outdf = do.call(rbind, moddfs)
  
  return(outdf)
}
```

```{r}
mergeLists(asympt_sympt_de_list)
```

```{r}

unique(unlist(lapply(str_split(names(asympt_sympt_de_list), "\\."), function(x){return(x[1])})))

```



```{r fig.width=30, fig.height=14}

markers.use.tt.asympt_sympt=subset(mergeLists(asympt_sympt_de_list) ,avg_logFC>0&p_val_adj<0.01&!startsWith(gene, "MT-")&!startsWith(gene, "RP"))

finalMarkers.use.tt.asympt_sympt = markers.use.tt.asympt_sympt %>% group_by(clusterID) %>% arrange(p_val_adj, desc(abs(avg_logFC),), .by_group=TRUE) %>% dplyr::slice(1:20)
finalMarkers.use.tt.asympt_sympt

nlabels = as.vector(outer(diseaseStates, cellTypes, paste, sep=" "))

nlabels = c(
  "Monocytes SYMPT", "Monocytes ASYMPT", "Monocytes CTRL",
  "NK cells SYMPT", "2 ASYMPT", "2 CTRL",
  "3 SYMPT", "3 ASYMPT", "3 CTRL",
  "4 SYMPT", "4 ASYMPT", "4 CTRL",
  "5 SYMPT", "5 ASYMPT", "5 CTRL",
  "6 SYMPT", "6 ASYMPT", "6 CTRL",
  "7 SYMPT", "7 ASYMPT", "7 CTRL",
  "8 SYMPT", "8 ASYMPT", "8 CTRL"
  )

p_dp_idents = DotPlot(obj.integrated, features = unique(finalMarkers.use.tt.asympt_sympt$gene), group.by = "shortcellnames", split.by = "disease_state", cols=c("#444fd9", "#444fd9", "#444fd9"), assay="RNA")+theme(axis.text.x = element_text(angle = 90, hjust = 1))
p_dp_idents
p_dp_idents + scale_y_discrete(labels= nlabels)

cellTypes = c("Monocytes", "NK cells", "T cells", "B cells", "gdT cells", "NK cells", "DC", "pDC")
diseaseStates = c("SYMPT", "ASYMPT", "CTRL")



```

```{r fig.width=30, fig.height=8}
p_dp_idents = DotPlot(obj.integrated, features = unique(finalMarkers.use.tt$gene), group.by = "idents", assay="RNA")+
  theme(axis.text.x = element_text(angle = 90, hjust = 1))

png(filename="dotplot_topgenes_by_idents.png", width = 3000, height=800)
plot(p_dp_idents)
dev.off()

pdf(file="dotplot_topgenes_by_idents.pdf", width = 30, height=8)
plot(p_dp_idents)
dev.off()

plot(p_dp_idents)

```

```{r fig.width=30, fig.height=8}


p_dp_cellnames = DotPlot(obj.integrated, features = unique(finalMarkers.use.tt$gene), group.by = "unified_cellnames", assay="RNA")+
  theme(axis.text.x = element_text(angle = 90, hjust = 1))

png(filename="dotplot_topgenes_by_cellnames.png", width = 3000, height=800)
plot(p_dp_cellnames)
dev.off()

pdf(file="dotplot_topgenes_by_cellnames.pdf", width = 30, height=8)
plot(p_dp_cellnames)
dev.off()

plot(p_dp_cellnames)

```

```{r fig.width=15, fig.height=20}

p_hm_idents = Seurat::DoHeatmap(obj.integrated, features = unique(finalMarkers.use.tt$gene), group.by = "idents", assay="integrated")

png(filename="heatmap_topgenes_by_idents.png", width = 1500, height=2000)
plot(p_hm_idents)
dev.off()

pdf(file="heatmap_topgenes_by_idents.pdf", width = 15, height=20)
plot(p_hm_idents)
dev.off()

plot(p_hm_idents)

```

```{r fig.width=15, fig.height=22}

p_hm_cellnames = Seurat::DoHeatmap(obj.integrated, features = unique(finalMarkers.use.tt$gene), group.by = "unified_cellnames", assay="integrated")

png(filename="heatmap_topgenes_by_cellnames.png", width = 1500, height=2200)
plot(p_hm_cellnames)
dev.off()

pdf(file="heatmap_topgenes_by_cellnames.pdf", width = 15, height=22)
plot(p_hm_cellnames)
dev.off()

plot(p_hm_cellnames)

```


```{r fig.width=25, fig.height=5}
p=DimPlot(obj.integrated, group.by = "unified_cellnames", split.by = "disease_state", label=F)

png("umap_cell_by_disease_state.png", width=25*100, height = 5*100)
plot(p)
dev.off()

pdf("umap_cell_by_disease_state.pdf", width=25, height=5)
plot(p)
dev.off()

p=DimPlot(obj.integrated, group.by = "unified_cellnames", split.by = "disease_state", label=T)

png("umap_cell_by_disease_state_wlabel.png", width=25*100, height = 5*100)
plot(p)
dev.off()

pdf("umap_cell_by_disease_state_wlabel.pdf", width=25, height=5)
plot(p)
dev.off()

p
```


```{r fig.width=25, fig.height=5}
p=DimPlot(obj.integrated, split.by = "disease_state", label=F)

png("umap_ident_by_disease_state.png", width=25*100, height = 5*100)
plot(p)
dev.off()

pdf("umap_ident_by_disease_state.pdf", width=25, height=5)
plot(p)
dev.off()

p=DimPlot(obj.integrated, split.by = "disease_state", label=T)

png("umap_ident_by_disease_state_wlabel.png", width=25*100, height = 5*100)
plot(p)
dev.off()

pdf("umap_ident_by_disease_state_wlabel.pdf", width=25, height=5)
plot(p)
dev.off()

p
```

```{r fig.width=12, fig.height=5}
DimPlot(obj.integrated, group.by = "orig.ident")
```

```{r fig.width=12, fig.height=5}
DimPlot(obj.integrated, group.by = "condition")
```

```{r}
DimPlot(subset(obj.integrated, disease_state == "SYMPTOMATIC"& mpoint==1), pt.size = 0.005, group.by="cellnames", label=T)
```


```{r fig.height=20, fig.width=20}

plotTimepoints = function(scobj, use.idents=NULL, outnames=NULL, xmin = -11,xmax = 15,ymin = -15,ymax = 10.5, groupby="cellnames", include.ctrl=T, label=F, fig.height=20, fig.width=20)
{
  all_elems = list()
  
  if (is.null(use.idents))
  {
    use.idents =NULL
  }

  plotIdx = 1
  ctrlLabels = c()

if (include.ctrl)
{
  all_elems[[1]] = DimPlot(subset(scobj, disease_state == "CONTROL", idents = use.idents ), pt.size = 0.005, group.by=groupby, label=label)+theme(legend.position = "none") + xlim(xmin, xmax) + ylim(ymin, ymax)
  all_elems[[2]] = NULL
  all_elems[[3]] = NULL
  
  ctrlLabels = c("CONTROL / TP1","", "")
 
  plotIdx = 4 
}

all_elems[[plotIdx]] = DimPlot(subset(scobj, disease_state == "SYMPTOMATIC"& mpoint==1, idents = use.idents), pt.size = 0.005, group.by=groupby, label=label)+theme(legend.position = "none") + xlim(xmin, xmax) + ylim(ymin, ymax)
all_elems[[plotIdx+1]] = DimPlot(subset(scobj, disease_state == "SYMPTOMATIC"& mpoint==2, idents = use.idents), pt.size = 0.005, group.by=groupby, label=label)+theme(legend.position = "none") + xlim(xmin, xmax) + ylim(ymin, ymax)
all_elems[[plotIdx+2]] = DimPlot(subset(scobj, disease_state == "SYMPTOMATIC"& mpoint==3, idents = use.idents), pt.size = 0.005, group.by=groupby, label=label)+theme(legend.position = "none") + xlim(xmin, xmax) + ylim(ymin, ymax)

plotIdx = plotIdx + 3

all_elems[[plotIdx]] = try(DimPlot(subset(scobj, disease_state == "ASYMPTOMATIC"& mpoint==1, idents = use.idents), pt.size = 0.005, group.by=groupby, label=label)+theme(legend.position = "none") + xlim(xmin, xmax) + ylim(ymin, ymax))
all_elems[[plotIdx+1]] = DimPlot(subset(scobj, disease_state == "ASYMPTOMATIC"& mpoint==2, idents = use.idents), pt.size = 0.005, group.by=groupby, label=label)+theme(legend.position = "none") + xlim(xmin, xmax) + ylim(ymin, ymax)
all_elems[[plotIdx+2]]= DimPlot(subset(scobj, disease_state == "ASYMPTOMATIC"& mpoint==3, idents = use.idents), pt.size = 0.005, group.by=groupby, label=label)+theme(legend.position = "none") + xlim(xmin, xmax) + ylim(ymin, ymax)

# extract a legend that is laid out horizontally
legend_b <- get_legend(
  all_elems[[1]] + 
    guides(color = guide_legend(nrow = 2, override.aes = list(size=10)))+
    theme(legend.position = "bottom")
)

ap = cowplot::plot_grid(
  plotlist = all_elems, align="hv",
  labels = c(ctrlLabels, "SYMPTOMATIC / TP1", "TP2", "TP3", "ASYMPTOMATIC / TP1", "TP2", "TP3"), ncol=3
)

fp = cowplot::plot_grid(ap, legend_b, ncol = 1, rel_heights = c(1, .1))

if (!is.null(outnames))
{
 for (outn in outnames)
 {
   cowplot::save_plot(outn, fp, base_height = fig.height, base_width = fig.width) 
 }
}

return(fp)
}


```

```{r fig.height=14, fig.width=20}

plotTimepoints(obj.integrated, outnames=c("umap_noctrl_cellnames_by_timepoint_wlabel.png", "umap_noctrl_cellnames_by_timepoint_wlabel.pdf"), groupby="unified_cellnames", label=T, include.ctrl = F,fig.height=14, fig.width=20)

plotTimepoints(obj.integrated, outnames=c("umap_noctrl_cellnames_by_timepoint.png", "umap_noctrl_cellnames_by_timepoint.pdf"), groupby="unified_cellnames", include.ctrl = F,fig.height=14, fig.width=20)

```

```{r fig.height=20, fig.width=20}

plotTimepoints(obj.integrated, outnames=c("umap_cellnames_by_timepoint_wlabel.png", "umap_cellnames_by_timepoint_wlabel.pdf"), groupby="unified_cellnames", label=T)

plotTimepoints(obj.integrated, outnames=c("umap_cellnames_by_timepoint.png", "umap_cellnames_by_timepoint.pdf"), groupby="unified_cellnames")

```

```{r fig.height=20, fig.width=20}

plotTimepoints(obj.integrated, outnames=c("umap_by_timepoint_wlabel.png", "umap_by_timepoint_wlabel.pdf"), groupby="idents", label=T)

plotTimepoints(obj.integrated, outnames=c("umap_by_timepoint.png", "umap_by_timepoint.pdf"), groupby="idents")

```


```{r fig.width=20, fig.height=8}
cowplot::plot_grid(
  DimPlot(obj.integrated, group.by = "ident", label=T),
  DimPlot(obj.integrated, group.by = "unified_cellnames", label=T)
)
```


```{r}

idents.tcell = c(1,4,5,8,13)
idents.gdtcell = c(10)
idents.nk = c(2,3,9)
idents.bcell = c(6,7,12)
idents.plasma = c(17)
idents.pdc = c(16)
idents.monocytes = c(0,11,15)
idents.dendritic = c(14)

idents.cd8 = c(1,4,8)
idents.cd4t = c(5)

```

```{r}

obj.integrated.subset = obj.integrated

```


```{r fig.width=20}

makeSubCluster = function( scobj, idents.sub, subclustername)
{
  
  # Subset pbmc_small cluster1 and sub-cluster
  scobj_sub <- subset(obj.integrated, idents = idents.sub)
  DefaultAssay(scobj_sub) = "integrated"
  scobj_sub <- FindNeighbors(scobj_sub, dims = 1:10, k.param = 10)
  scobj_sub <- FindClusters(scobj_sub, resolution = 0.75)
  
  entryname = paste("sub_cluster_", subclustername, sep="")
  print(entryname)
  
  # Generate a new column called sub_cluster in the metadata
  scobj[[entryname]] <- as.character(Idents(scobj))
  
  cellIdents = paste(subclustername, Idents(scobj_sub), sep="_")
  names(cellIdents) = names(Idents(scobj_sub))
  print(head(cellIdents))

  # Change the information of cells containing sub-cluster information
  scobj[[entryname]][names(cellIdents),] <- cellIdents
  
  print(head(scobj[[entryname]][names(cellIdents),]))
  
  return(scobj)
}

```

```{r}
DimPlot(subset(obj.integrated.subset, idents = idents.tcell), label=T)

```


```{r}

obj.integrated.subset = makeSubCluster(obj.integrated.subset, idents.tcell, "tcell")

p=cowplot::plot_grid(DimPlot(subset(obj.integrated.subset, idents = idents.tcell), group.by = "sub_cluster_tcell", label=T))

save_plot(p, "sub_cluster_tcell_new", fig.width=10, fig.height=10)

```

```{r}

VlnPlot(subset(obj.integrated.subset, idents = idents.tcell), features = c("CCR6", "CTLA4"), group.by = "sub_cluster_tcell")

```
```{r}
FeaturePlot(subset(obj.integrated.subset, idents = idents.tcell), feature="CCR6")
```

```{r}
FeaturePlot(subset(obj.integrated.subset, idents = idents.tcell), feature="CTLA4")
```



```{r}
DimPlot(subset(obj.integrated.subset, idents = idents.bcell), label=T)

```

```{r}

obj.integrated.subset = makeSubCluster(obj.integrated.subset, idents.bcell, "bcell")
p = cowplot::plot_grid(DimPlot(subset(obj.integrated.subset, idents = idents.bcell), group.by = "sub_cluster_bcell", label=T))

p

cowplot::save_plot("sub_cluster_bcell.png", p, base_height = 10, base_width = 10) 
cowplot::save_plot("sub_cluster_bcell.pdf", p, base_height = 10, base_width = 10) 

```

```{r fig.height=8, fig.width=20}

obj.integrated.subset = makeSubCluster(obj.integrated.subset, idents.monocytes, "monocytes")
p=cowplot::plot_grid(DimPlot(subset(obj.integrated.subset, idents = idents.monocytes), group.by = "sub_cluster_monocytes", label=T))

p

cowplot::save_plot("sub_cluster_monocytes.png", p, base_height = 10, base_width = 10) 
cowplot::save_plot("sub_cluster_monocytes.pdf", p, base_height = 10, base_width = 10) 


plotTimepoints(subset(obj.integrated.subset, idents = idents.monocytes), outnames=c("todoplots/14/umap_monocytes_by_timepoint.png", "todoplots/14/umap_monocytes_by_timepoint.pdf"), groupby="sub_cluster_monocytes", label=T, include.ctrl = T,fig.height=14, fig.width=20, xmin=5, xmax=15, ymin = -1, ymax=8)


p=DimPlot(subset(obj.integrated.subset, idents = idents.monocytes), group.by="sub_cluster_monocytes", split.by = "mpoint")
p

save_plot(p, "todoplots/14/umap_monocytes_by_mpoint", fig.height = 8, fig.width = 20)

p=DimPlot(subset(obj.integrated.subset, idents = idents.monocytes), group.by="sub_cluster_monocytes", split.by = "disease_state")
p

save_plot(p, "todoplots/14/umap_monocytes_by_disease_state", fig.height = 8, fig.width = 20)

```
a little offspin: try to get marker genes for those clusters ...

```{r}
obj.monocytes_fine = subset(obj.integrated.subset, idents = idents.monocytes)
obj.monocytes_fine_12 = obj.monocytes_fine[,obj.monocytes_fine$samplename=="20094_0012_A_B"]

sum(obj.monocytes_fine$samplename=="20094_0012_A_B")
obj.monocytes_fine_12

Idents(obj.monocytes_fine_12) = obj.monocytes_fine_12$sub_cluster_monocytes

DimPlot(obj.monocytes_fine_12, group.by = "ident", label=T)

saveRDS(obj.monocytes_fine_12, "obj.monocytes_fine_12.Rds")

```

```{r}

deResTT_mc_fine = makeDEResults(obj.monocytes_fine_12, assay="RNA", test="t")
exprdfTT_mc_fine = getDEXpressionDF(obj.monocytes_fine_12, deResTT_mc_fine, assay="RNA")
write.table(exprdfTT_mc_fine, "monocytes_fine.de.t.tsv", sep="\t", row.names=F, quote = F)


```


```{r}

obj.integrated.subset = makeSubCluster(obj.integrated.subset, idents.nk, "nk")
p=cowplot::plot_grid(DimPlot(subset(obj.integrated.subset, idents = idents.nk), group.by = "sub_cluster_nk", label=T))

p

cowplot::save_plot("sub_cluster_nk.png", p, base_height = 10, base_width = 10) 
cowplot::save_plot("sub_cluster_nk.pdf", p, base_height = 10, base_width = 10) 

```

```{r}
DimPlot(obj.integrated, label=T)

```


```{r fig.height=20, fig.width=20}

plotTimepoints(obj.integrated, outname="umap_by_timepoint_t_cell.png", groupby="ident", use.idents = idents.tcell, xmin = -11,xmax = 5,ymin = 0,ymax = 10)

```

```{r fig.height=20, fig.width=20}

plotTimepoints(obj.integrated, outname="umap_by_timepoint_b_cell.png", groupby="ident", use.idents = idents.bcell, xmin = -5,xmax = 5,ymin = -15,ymax = -5)

```

```{r fig.height=20, fig.width=20}

plotTimepoints(obj.integrated, outname="umap_by_timepoint_monocytes.png", groupby="ident", use.idents = idents.monocytes, xmin = 5,xmax = 15,ymin = -5,ymax = 5)

```

```{r}
DimPlot(obj.integrated, label=T)
```

# 3 DE analysis

```{r}

makeGrpCells = function(scdata, clusterIdents)
{
    cellIdents = Idents(scdata)
    cellIdents.1 = names(cellIdents[cellIdents %in% clusterIdents])
    cellIdents.1 = unlist(lapply(cellIdents.1, as.character))   
    
    return(cellIdents.1)
}

```

```{r}
FeaturePlot(obj.integrated, features = c("CD4"))
```


```{r}





cells.tcell = makeGrpCells(obj.integrated, idents.tcell)

cells.cd8 = makeGrpCells(obj.integrated, idents.cd8)
cells.cd4t = makeGrpCells(obj.integrated, idents.cd4t)

cells.bcell = makeGrpCells(obj.integrated, idents.bcell)
cells.monocyte = makeGrpCells(obj.integrated, idents.monocytes)
cells.nk = makeGrpCells(obj.integrated, idents.nk)

cells.plasma = makeGrpCells(obj.integrated, idents.plasma)
cells.pdc = makeGrpCells(obj.integrated, idents.pdc)
cells.gdtcell = makeGrpCells(obj.integrated, idents.gdtcell)
cells.dendritic = makeGrpCells(obj.integrated, idents.dendritic)


print(length(cells.tcell))
print(length(cells.bcell))
print(length(cells.monocyte))
print(length(cells.nk))

print(length(cells.plasma))
print(length(cells.pdc))
print(length(cells.gdtcell))
print(length(cells.dendritic))

unique(obj.integrated$unified_cellnames)

celltypenames = as.character(obj.integrated$unified_cellnames)

celltypenames[cells.tcell] = "T cells"
celltypenames[cells.bcell] = "B cells"
celltypenames[cells.monocyte] = "Monocytes"
celltypenames[cells.nk] = "NK cells"
celltypenames[cells.plasma] = "Plasma cells"
celltypenames[cells.pdc] = "Plasmacytoid dendritic cells"
celltypenames[cells.gdtcell] = "gdT cells"
celltypenames[cells.dendritic] = "Dendritic cells"
celltypenames[cells.cd8] = "CD8+ T cells"
celltypenames[cells.cd4t] = "CD4+ T cells"

obj.integrated$celltypenames = celltypenames



celltypenamestp = as.character(obj.integrated$unified_cellnames)

tpcells = list("TP1"=cells.tp1, "TP2"=cells.tp2, "TP3"=cells.tp3)

for (tpname in names(tpcells))
{
  
  tpcelllist = tpcells[[tpname]]
  
  celltypenamestp[intersect(cells.tcell, tpcelllist)] = paste("T cells", tpname)
  celltypenamestp[intersect(cells.bcell, tpcelllist)] = paste("B cells", tpname)
  celltypenamestp[intersect(cells.monocyte, tpcelllist)] = paste("Monocytes", tpname)
  celltypenamestp[intersect(cells.nk, tpcelllist)] = paste("NK cells", tpname)
  celltypenamestp[intersect(cells.plasma, tpcelllist)] = paste("Plasma cells", tpname)
  celltypenamestp[intersect(cells.pdc, tpcelllist)] = paste("Plasmacytoid dendritic cells", tpname)
  celltypenamestp[intersect(cells.gdtcell, tpcelllist)] = paste("gdT cells", tpname)
  celltypenamestp[intersect(cells.dendritic, tpcelllist)] = paste("Dendritic cells", tpname)
  celltypenamestp[intersect(cells.cd8, tpcelllist)] = paste("CD8+ T cells", tpname)
  celltypenamestp[intersect(cells.cd4t, tpcelllist)] = paste("CD4+ T cells", tpname)
  
  
}

obj.integrated$celltypenamestp = celltypenamestp


celltypenamesstate = as.character(obj.integrated$unified_cellnames)

tpcells = list("TP1"=cells.tp1, "TP2"=cells.tp2, "TP3"=cells.tp3)
statecells = list(".CTRL"=cells.control, "ASYMPT"=cells.asymptomatic, "SYMPT"=cells.symptomatic)

for (statename in names(statecells))
{
  print(statename)
  
  statenamelist = statecells[[statename]]
  
  celltypenamesstate[intersect(cells.tcell, statenamelist)] = paste("T cells", statename)
  celltypenamesstate[intersect(cells.bcell, statenamelist)] = paste("B cells", statename)
  celltypenamesstate[intersect(cells.monocyte, statenamelist)] = paste("Monocytes", statename)
  celltypenamesstate[intersect(cells.nk, statenamelist)] = paste("NK cells", statename)
  celltypenamesstate[intersect(cells.plasma, statenamelist)] = paste("Plasma cells", statename)
  celltypenamesstate[intersect(cells.pdc, statenamelist)] = paste("Plasmacytoid dendritic cells", statename)
  celltypenamesstate[intersect(cells.gdtcell, statenamelist)] = paste("gdT cells", statename)
  celltypenamesstate[intersect(cells.dendritic, statenamelist)] = paste("Dendritic cells", statename)
  celltypenamesstate[intersect(cells.cd8, statenamelist)] = paste("CD8+ T cells", statename)
  celltypenamesstate[intersect(cells.cd4t, statenamelist)] = paste("CD4+ T cells", statename)
  
  
}

obj.integrated$celltypenamesstate = celltypenamesstate


celltypenamestpstate = as.character(obj.integrated$unified_cellnames)


for (tpname in names(tpcells))
{
  
  tpcelllist = tpcells[[tpname]]
  
  for (statename in names(statecells))
  {
      statecelllist = statecells[[statename]]
    
      celltypenamestpstate[intersect(intersect(cells.tcell, tpcelllist), statecelllist)] = paste("T cells", tpname, statename)
      celltypenamestpstate[intersect(intersect(cells.bcell, tpcelllist), statecelllist)] = paste("B cells", tpname, statename)
      celltypenamestpstate[intersect(intersect(cells.monocyte, tpcelllist), statecelllist)] = paste("Monocytes", tpname, statename)
      celltypenamestpstate[intersect(intersect(cells.nk, tpcelllist), statecelllist)] = paste("NK cells", tpname, statename)
      celltypenamestpstate[intersect(intersect(cells.plasma, tpcelllist), statecelllist)] = paste("Plasma cells", tpname, statename)
      celltypenamestpstate[intersect(intersect(cells.pdc, tpcelllist), statecelllist)] = paste("Plasmacytoid dendritic cells", tpname, statename)
      celltypenamestpstate[intersect(intersect(cells.gdtcell, tpcelllist), statecelllist)] = paste("gdT cells", tpname, statename)
      celltypenamestpstate[intersect(intersect(cells.dendritic, tpcelllist), statecelllist)] = paste("Dendritic cells", tpname, statename)
      celltypenamestpstate[intersect(intersect(cells.cd8, tpcelllist), statecelllist)] = paste("CD8+ T cells", tpname, statename)
      celltypenamestpstate[intersect(intersect(cells.cd4t, tpcelllist), statecelllist)] = paste("CD4+ T cells", tpname, statename)
  
  }
  

  
}

obj.integrated$celltypenamestpstate = celltypenamestpstate

```

```{r}
 
cells.control = names(obj.integrated$disease_state[obj.integrated$disease_state == "CONTROL"])
cells.asymptomatic = names(obj.integrated$disease_state[obj.integrated$disease_state == "ASYMPTOMATIC"])
cells.symptomatic = names(obj.integrated$disease_state[obj.integrated$disease_state == "SYMPTOMATIC"])

print(length(cells.control))
print(length(cells.asymptomatic))
print(length(cells.symptomatic))

```

```{r}
 
cells.tp1 = names(obj.integrated$mpoint[obj.integrated$mpoint == 1])
cells.tp2 = names(obj.integrated$mpoint[obj.integrated$mpoint == 2])
cells.tp3 = names(obj.integrated$mpoint[obj.integrated$mpoint == 3])

print(length(cells.tp1))
print(length(cells.tp2))
print(length(cells.tp3))

cells.tp1[!cells.tp1 %in% colnames(obj.integrated)]
cells.tp2[!cells.tp2 %in% colnames(obj.integrated)]
cells.tp3[!cells.tp3 %in% colnames(obj.integrated)]

```


```{r}

allmpoints = unique(obj.integrated$mpoint)
allmpoints

compareCellsBulk = list()
clusterID = "bulk"

keyname = paste(clusterID, "symptomatic_control", sep=".")
cres = compareCells(obj.integrated,
         cells.symptomatic,
         cells.control,
         "symptomatic", "control", prefix=paste(clusterID, sep="_"), test="t", outfolder="bulk_de")

compareCellsBulk[[keyname]] = cres

keyname = paste(clusterID, "asymptomatic_control", sep=".")
cres = compareCells(obj.integrated,
         cells.asymptomatic,
         cells.control,
         "asymptomatic", "control", prefix=paste(clusterID, sep="_"), test="t", outfolder="bulk_de")

compareCellsBulk[[keyname]] = cres

keyname = paste(clusterID, "asymptomatic_symptomatic", sep=".")
cres = compareCells(obj.integrated,
         cells.asymptomatic,
         cells.symptomatic,
         "asymptomatic", "symptomatic", prefix=paste(clusterID, sep="_"), test="t", outfolder="bulk_de")

compareCellsBulk[[keyname]] = cres

for (mpoint in allmpoints)
{
  
  mpointCells = names(obj.integrated$mpoint[obj.integrated$mpoint==mpoint])
  
  print(paste("Cluster", clusterID , "for timepoint", mpoint, "with", length(mpointCells), "cells"))

  keyname = paste(clusterID, mpoint, "symptomatic_control", sep=".")
  cres = compareCells(obj.integrated,
           intersect(cells.symptomatic, mpointCells),
           intersect(cells.control, mpointCells),
           "symptomatic", "control", prefix=paste(clusterID, mpoint, sep="_"), test="t", outfolder="bulk_de")

  compareCellsBulk[[keyname]] = cres
  
  
      
  keyname = paste(clusterID, mpoint, "asymptomatic_control", sep=".")
  cres = compareCells(obj.integrated,
             intersect(cells.asymptomatic, mpointCells),
             intersect(cells.control, mpointCells),
             "asymptomatic", "control", prefix=paste(clusterID, mpoint, sep="_"), test="t", outfolder="bulk_de")
  
  compareCellsBulk[[keyname]] = cres
  
    
  keyname = paste(clusterID, mpoint, "asymptomatic_symptomatic", sep=".")
  cres = compareCells(obj.integrated,
             intersect(cells.asymptomatic, mpointCells),
             intersect(cells.symptomatic, mpointCells),
             "asymptomatic", "symptomatic", prefix=paste(clusterID, mpoint, sep="_"), test="t", outfolder="bulk_de")
  
  compareCellsBulk[[keyname]] = cres
}



```





```{r}

identResult_disease = list()

for (clusterID in sort(unique(obj.integrated$idents)))
{
  clusterCells = names(obj.integrated$idents[obj.integrated$idents==clusterID])
  
  print(paste("Cluster", clusterID , "with", length(clusterCells), "cells"))
  
  keyname = paste(clusterID, "symptomatic_control", sep=".")
  cres = compareCells(obj.integrated,
             intersect(cells.symptomatic, clusterCells),
             intersect(cells.control, clusterCells),
             "symptomatic", "control", prefix=clusterID, test="t", outfolder="detests_idents_3")
  
  identResult_disease[[keyname]] = cres
  
    
  keyname = paste(clusterID, "asymptomatic_control", sep=".")
  cres = compareCells(obj.integrated,
             intersect(cells.asymptomatic, clusterCells),
             intersect(cells.control, clusterCells),
             "asymptomatic", "control", prefix=clusterID, test="t", outfolder="detests_idents_3")
  
  identResult_disease[[keyname]] = cres
  
    
  keyname = paste(clusterID, "asymptomatic_symptomatic", sep=".")
  cres = compareCells(obj.integrated,
             intersect(cells.asymptomatic, clusterCells),
             intersect(cells.symptomatic, clusterCells),
             "asymptomatic", "symptomatic", prefix=clusterID, test="t", outfolder="detests_idents_3")
  
  identResult_disease[[keyname]] = cres
}


```

```{r}

allmpoints = unique(obj.integrated$mpoint)
allmpoints

identResult_disease_tp = list()

for (clusterID in sort(unique(obj.integrated$idents)))
{
  clusterCells = names(obj.integrated$idents[obj.integrated$idents==clusterID])
  
  for (mpoint in allmpoints)
  {
    
    mpointCells = names(obj.integrated$mpoint[obj.integrated$mpoint==mpoint])
    
    clusterMpointCells = intersect(mpointCells, clusterCells)
    
    print(paste("Cluster", clusterID , "for timepoint", mpoint, "with", length(clusterMpointCells), "cells"))

    keyname = paste(clusterID, mpoint, "symptomatic_control", sep=".")
    cres = compareCells(obj.integrated,
             intersect(cells.symptomatic, clusterMpointCells),
             intersect(cells.control, clusterMpointCells),
             "symptomatic", "control", prefix=paste(clusterID, mpoint, sep="_"), test="t", outfolder="detests_idents_4")
  
    identResult_disease_tp[[keyname]] = cres
    
    
        
    keyname = paste(clusterID, mpoint, "asymptomatic_control", sep=".")
    cres = compareCells(obj.integrated,
               intersect(cells.asymptomatic, clusterMpointCells),
               intersect(cells.control, clusterMpointCells),
               "asymptomatic", "control", prefix=paste(clusterID, mpoint, sep="_"), test="t", outfolder="detests_idents_4")
    
    identResult_disease_tp[[keyname]] = cres
    
      
    keyname = paste(clusterID, mpoint, "asymptomatic_symptomatic", sep=".")
    cres = compareCells(obj.integrated,
               intersect(cells.asymptomatic, clusterMpointCells),
               intersect(cells.symptomatic, clusterMpointCells),
               "asymptomatic", "symptomatic", prefix=paste(clusterID, mpoint, sep="_"), test="t", outfolder="detests_idents_4")
    
    identResult_disease_tp[[keyname]] = cres
  }
}


```

```{r}


cd8.sympt_ctrl = compareCells(obj.integrated,
             intersect(cells.symptomatic, cells.cd8),
             intersect(cells.control, cells.cd8),
             "symptomatic", "control", prefix="cd8", test="t", outfolder="cd8_tests")

cd8.asympt_ctrl = compareCells(obj.integrated,
             intersect(cells.asymptomatic, cells.cd8),
             intersect(cells.control, cells.cd8),
             "asymptomatic", "control", prefix="cd8", test="t", outfolder="cd8_tests")

cd8.asympt_sympt = compareCells(obj.integrated,
             intersect(cells.asymptomatic, cells.cd8),
             intersect(cells.symptomatic, cells.cd8),
             "asymptomatic", "symptomatic", prefix="cd8", test="t", outfolder="cd8_tests")

```

```{r}

makeShortDE = function( scobj, cells.target, clusterID, resfolder)
{
  
  dir.create(resfolder, showWarnings = FALSE, recursive=TRUE)

  
  disease_tp = list()
  
  allmpoints = unique(scobj$mpoint)
  allmpoints

  dis.sympt_ctrl = compareCells(obj.integrated,
               intersect(cells.symptomatic, cells.target),
               intersect(cells.control, cells.target),
               "symptomatic", "control", prefix=clusterID, test="t", outfolder=resfolder)
  
  dis.asympt_ctrl = compareCells(obj.integrated,
               intersect(cells.asymptomatic, cells.target),
               intersect(cells.control, cells.target),
               "asymptomatic", "control", prefix=clusterID, test="t", outfolder=resfolder)
  
  dis.asympt_sympt = compareCells(obj.integrated,
               intersect(cells.asymptomatic, cells.target),
               intersect(cells.symptomatic, cells.target),
               "asymptomatic", "symptomatic", prefix=clusterID, test="t", outfolder=resfolder)
  
  disease_tp[[paste(clusterID, "symptomatic_control", sep=".")]] = dis.sympt_ctrl
  disease_tp[[paste(clusterID, "asymptomatic_control", sep=".")]] = dis.asympt_ctrl
  disease_tp[[paste(clusterID, "asymptomatic_symptomatic", sep=".")]] = dis.asympt_sympt
  
  for (mpoint in allmpoints)
  {
    
    mpointCells = names(obj.integrated$mpoint[obj.integrated$mpoint==mpoint])
    
    clusterMpointCells = intersect(mpointCells, cells.target)
    
    print(paste("for timepoint", mpoint, "with", length(clusterMpointCells), "cells"))
  
    keyname = paste(clusterID, mpoint, "symptomatic_control", sep=".")
    cres = compareCells(obj.integrated,
             intersect(cells.symptomatic, clusterMpointCells),
             intersect(cells.control, clusterMpointCells),
             "symptomatic", "control", prefix=paste(clusterID, mpoint, sep="_"), test="t", outfolder=resfolder)
  
    disease_tp[[keyname]] = cres
    
    
        
    keyname = paste(clusterID, mpoint, "asymptomatic_control", sep=".")
    cres = compareCells(obj.integrated,
               intersect(cells.asymptomatic, clusterMpointCells),
               intersect(cells.control, clusterMpointCells),
               "asymptomatic", "control", prefix=paste(clusterID, mpoint, sep="_"), test="t", outfolder=resfolder)
    
    disease_tp[[keyname]] = cres
    
      
    keyname = paste(clusterID, mpoint, "asymptomatic_symptomatic", sep=".")
    cres = compareCells(obj.integrated,
               intersect(cells.asymptomatic, clusterMpointCells),
               intersect(cells.symptomatic, clusterMpointCells),
               "asymptomatic", "symptomatic", prefix=paste(clusterID, mpoint, sep="_"), test="t", outfolder=resfolder)
    
    disease_tp[[keyname]] = cres
  }

return(disease_tp)
}
```

```{r}
cd8_disease_tp = makeShortDE(obj.integrated, cells.cd8, "cd8", "todoplots/12/cd8/")
cd4t_disease_tp = makeShortDE(obj.integrated, cells.cd4t, "cd4t", "todoplots/12/cd4t/")

```

```{r}
saveRDS(cd8_disease_tp, "todoplots/12/cd8/cd8_disease_tp.Rds")
saveRDS(cd4t_disease_tp, "todoplots/12/cd4t/cd4t_disease_tp.Rds")
```

```{r}
cd8_disease_tp=readRDS("todoplots/12/cd8/cd8_disease_tp.Rds")
cd4t_disease_tp=readRDS("todoplots/12/cd4t/cd4t_disease_tp.Rds")
```


```{bash}

Rscript raAnalysis.R "todoplots/12/cd8/cd8_disease_tp.Rds" "go_universe.Rds" "human"
Rscript goAnalysis.R "todoplots/12/cd8/cd8_disease_tp.Rds" "go_universe.Rds"
Rscript gseAnalysis.R "todoplots/12/cd8/cd8_disease_tp.Rds"

Rscript raAnalysis.R "todoplots/12/cd4t/cd4t_disease_tp.Rds" "go_universe.Rds" "human"
Rscript goAnalysis.R "todoplots/12/cd4t/cd4t_disease_tp.Rds" "go_universe.Rds"
Rscript gseAnalysis.R "todoplots/12/cd4t/cd4t_disease_tp.Rds"

```

```{r}

cd8_disease_tp.ra = readRDS("todoplots/12/cd8/ra.cd8_disease_tp.Rds")
cd8_disease_tp.go = readRDS("todoplots/12/cd8/go.cd8_disease_tp.Rds")
cd8_disease_tp.gse = readRDS("todoplots/12/cd8/gse.cd8_disease_tp.Rds")

cd4t_disease_tp.ra = readRDS("todoplots/12/cd4t/ra.cd4t_disease_tp.Rds")
cd4t_disease_tp.go = readRDS("todoplots/12/cd4t/go.cd4t_disease_tp.Rds")
cd4t_disease_tp.gse = readRDS("todoplots/12/cd4t/gse.cd4t_disease_tp.Rds")

```

```{r}
t=makePlotsListsGO(cd8_disease_tp.go, cd8_disease_tp, "todoplots/12/cd8/go/")

t=makePlotsListsGO(cd8_disease_tp.ra, cd8_disease_tp, "todoplots/12/cd8/ra/", name.ggo=NULL, name.ego="rao", name.ego_up="rao_up", name.ego_down="rao_down", name.title="ReactomePA")

t=makePlotsListsGSE(cd8_disease_tp.gse, cd8_disease_tp, "todoplots/12/cd8/gse/")

t=makePlotsListsGO(cd4t_disease_tp.go, cd4t_disease_tp, "todoplots/cd4tcells/go/")

t=makePlotsListsGO(cd4t_disease_tp.ra, cd4t_disease_tp, "todoplots/cd4tcells/ra/", name.ggo=NULL, name.ego="rao", name.ego_up="rao_up", name.ego_down="rao_down", name.title="ReactomePA")

t=makePlotsListsGSE(cd4t_disease_tp.gse, cd4t_disease_tp, "todoplots/cd4tcells/gse/")
```

```{r}
data.frame(size = sort(sapply(.GlobalEnv, object.size)))
```


```{r fig.height=8, fig.width=20}
# CTSW 23, GZMM 24, KLRD1 25, PRF1 26, NKG7 27, GZMB 28, GZMH 29
features.cd8.mail=c("CTSW", "GZMM", "KLRD1", "PRF1", "NKG7" , "GZMB", "GZMH" )

vplt_cd8_genes = VlnPlot(subset(obj.integrated, idents=idents.cd8), group.by = "disease_state" , features = features.cd8.mail, assay = "RNA", ncol = 4, pt.size = 0, cols=c(color.asymptomatic, color.control, color.symptomatic) )
title <- ggdraw() + draw_label("CD8 Cells (All TPs)", fontface = 'bold')
vplt_cd8_genes=cowplot::plot_grid(title, vplt_cd8_genes, ncol = 1, rel_heights = c(0.1, 1))



dplt_cd8_genes = DotPlot(subset(obj.integrated, idents=idents.cd8), group.by = "disease_state", features = features.cd8.mail, assay = "RNA")
title <- ggdraw() + draw_label("CD8 Cells (All TPs)", fontface = 'bold')
dplt_cd8_genes=cowplot::plot_grid(title, dplt_cd8_genes, ncol = 1, rel_heights = c(0.1, 1))


save_plot(vplt_cd8_genes, "cd8_tests/cd8cells_mail_genes_violin", fig.height = 8, fig.width = 20)
save_plot(dplt_cd8_genes, "cd8_tests/cd8cells_mail_genes_dotplot", fig.height = 8, fig.width = 20)


for (mpointID in unique(obj.integrated$mpoint))
{
  
  print(mpointID)
  mpoint_vln = VlnPlot(subset(obj.integrated, idents=idents.cd8, mpoint==mpointID), group.by = "disease_state" , features = features.cd8.mail, assay = "RNA", ncol = 4, pt.size = 0, cols=mcols )
  title <- ggdraw() + draw_label(paste("CD8 Cells (TP", mpointID, ")", sep=""), fontface = 'bold')
mpoint_vln=cowplot::plot_grid(title, mpoint_vln, ncol = 1, rel_heights = c(0.1, 1))


  mpoint_vln=save_plot(mpoint_vln, paste("cd8_tests/cd8cells_mail_genes_violin", mpointID, sep="_"), fig.height = 8, fig.width = 20)
  plot(mpoint_vln)
  
  
  mpoint_dpl = DotPlot(subset(obj.integrated, idents=idents.cd8, mpoint==mpointID), group.by = "disease_state", features = features.cd8.mail, assay = "RNA")
  title <- ggdraw() + draw_label(paste("CD8 Cells (TP", mpointID, ")", sep=""), fontface = 'bold')
mpoint_dpl=cowplot::plot_grid(title, mpoint_dpl, ncol = 1, rel_heights = c(0.1, 1))

  mpoint_dpl=save_plot(mpoint_dpl, paste("cd8_tests/cd8cells_mail_genes_dotplot", mpointID, sep="_"), fig.height = 8, fig.width = 20)
  plot(mpoint_dpl)

  
}

```

```{r}
bcell.comparisons=list()

bcell.comparisons[["12_6"]] = compareCells(obj.integrated,
             makeGrpCells(obj.integrated, c(12)),
             makeGrpCells(obj.integrated, c(6)),
             "bcell12", "bcell6", prefix="bcell", test="t", outfolder="bcells_test")

bcell.comparisons[["12_7"]] = compareCells(obj.integrated,
             makeGrpCells(obj.integrated, c(12)),
             makeGrpCells(obj.integrated, c(7)),
             "bcell12", "bcell7", prefix="bcell", test="t", outfolder="bcells_test")

bcell.comparisons[["12_67"]] = compareCells(obj.integrated,
             makeGrpCells(obj.integrated, c(12)),
             makeGrpCells(obj.integrated, c(6,7)),
             "bcell12", "bcell67", prefix="bcell", test="t", outfolder="bcells_test")

```


```{r fig.height=6, fig.width=12}

features.tcell.mail = c("FOS", "GATA3", "JUNB")

vplt_tcell_genes = VlnPlot(subset(obj.integrated, idents=idents.tcell), group.by = "disease_state", features = features.tcell.mail, assay = "RNA", ncol=3, pt.size = 0, cols=dsCols)+ theme(legend.position = 'none')
title <- ggdraw() + draw_label("T-Cells (All TPs)", fontface = 'bold')
vplt_tcell_genes=cowplot::plot_grid(title, vplt_tcell_genes, ncol = 1, rel_heights = c(0.1, 1))
save_plot(vplt_tcell_genes, "todoplots/5/tcell.violin.fos_gata3_junb", 8 , 4)

dplt_tcell_genes = DotPlot(subset(obj.integrated, idents=idents.tcell), group.by = "disease_state", features = features.tcell.mail)
save_plot(dplt_tcell_genes, "todoplots/5/tcell.dotplot.fos_gata3_junb", 12 , 4)


```

```{r fig.height=6, fig.width=12}

features.nkcell.mail = c("JUN", "JUNB", "FOS", "FOSB")

vplt_nkcell_genes = VlnPlot(subset(obj.integrated, idents=idents.nk), group.by = "disease_state", features = features.nkcell.mail, assay = "RNA", ncol=4, pt.size = 0, cols=dsCols)+ theme(legend.position = 'none')
title <- ggdraw() + draw_label("NK-Cells (All TPs)", fontface = 'bold')
vplt_nkcell_genes=cowplot::plot_grid(title, vplt_nkcell_genes, ncol = 1, rel_heights = c(0.1, 1))
save_plot(vplt_nkcell_genes, "todoplots/6/nkcell.violin.jun_fosb", 12 , 4)

dplt_nkcell_genes = DotPlot(subset(obj.integrated, idents=idents.nk), group.by = "disease_state", features = features.nkcell.mail)
save_plot(dplt_nkcell_genes, "todoplots/6/nkcell.dotplot.jun_fosb", 12 , 4)

vplt_nkcell_genes = VlnPlot(subset(obj.integrated, idents=idents.nk), group.by = "disease_state", features = features.nkcell.mail[1:3], assay = "RNA", ncol=3, pt.size = 0, cols=dsCols)+ theme(legend.position = 'none')
title <- ggdraw() + draw_label("NK-Cells (All TPs)", fontface = 'bold')
vplt_nkcell_genes=cowplot::plot_grid(title, vplt_nkcell_genes, ncol = 1, rel_heights = c(0.1, 1))
save_plot(vplt_nkcell_genes, "todoplots/6/nkcell.violin.jun_fos", 12 , 4)

dplt_nkcell_genes = DotPlot(subset(obj.integrated, idents=idents.nk), group.by = "disease_state", features = features.nkcell.mail[1:3])
save_plot(dplt_nkcell_genes, "todoplots/6/nkcell.dotplot.jun_fos", 12 , 4)

```

```{r fig.height=6, fig.width=12}

features.nkcell.mail2 = c("CCL3", "CCL4", "THEMIS2")

vplt_nkcell_genes2 = VlnPlot(subset(obj.integrated, idents=idents.nk), group.by = "disease_state", features = features.nkcell.mail2, assay = "RNA", ncol=3, pt.size = 0, cols=dsCols)+ theme(legend.position = 'none')
title <- ggdraw() + draw_label("NK-Cells (All TPs)", fontface = 'bold')
vplt_nkcell_genes2=cowplot::plot_grid(title, vplt_nkcell_genes2, ncol = 1, rel_heights = c(0.1, 1))
save_plot(vplt_nkcell_genes2, "todoplots/7/nkcell.violin.ccl", 12 , 4)

dplt_nkcell_genes2 = DotPlot(subset(obj.integrated, idents=idents.nk), group.by = "disease_state", features = features.nkcell.mail2)
save_plot(dplt_nkcell_genes2, "todoplots/7/nkcell.dotplot.ccl", 12 , 4)

dplt_nkcell_genes2_ns = DotPlot(subset(obj.integrated, idents=idents.nk), group.by = "disease_state", features = features.nkcell.mail2, scale=F)
save_plot(dplt_nkcell_genes2_ns, "todoplots/7/nkcell.dotplot_ns.ccl", 12 , 4)

```

# Die Gene CTSW, PRF1, NKG7, GZMB, GZMK, GZMH, GNLY in der CD8+ Population (Populationen 1, 4, 8) der T Zellen als Violin und als Dot plot (asympt. vs. sympt. vs. ctrl)

```{r fig.height=8, fig.width=10}

features.cd8.mail = c("CTSW", "PRF1", "NKG7", "GZMB", "GZMH", "GZMK", "GNLY")

vplt_cd8_genes = VlnPlot(subset(obj.integrated, idents=idents.cd8), group.by = "disease_state", features = features.cd8.mail, assay = "RNA", ncol=4, pt.size = 0, cols=dsCols)+ theme(legend.position = 'none')
title <- ggdraw() + draw_label("CD8-Cells (All TPs)", fontface = 'bold')
vplt_cd8_genes=cowplot::plot_grid(title, vplt_cd8_genes, ncol = 1, rel_heights = c(0.1, 1))
save_plot(vplt_cd8_genes, "todoplots/8/cd8cells.violin.CTSW", 12 , 8)

dplt_cd8_genes = DotPlot(subset(obj.integrated, idents=idents.cd8), group.by = "disease_state", features = features.cd8.mail)
save_plot(dplt_cd8_genes, "todoplots/8/cd8cells.dotplot.CTSW", 12 , 4)


## TP ALL
vplt_cd8_genes = VlnPlot(subset(obj.integrated, idents=idents.cd8), group.by = "mpoint", features = features.cd8.mail, assay = "RNA", ncol=4, pt.size = 0, cols=mpCols)+ theme(legend.position = 'none')
title <- ggdraw() + draw_label("CD8-Cells (By TPs)", fontface = 'bold')
vplt_cd8_genes=cowplot::plot_grid(title, vplt_cd8_genes, ncol = 1, rel_heights = c(0.1, 1))
save_plot(vplt_cd8_genes, "todoplots/8/cd8cells.tp_all.violin.CTSW", 12 , 8)

dplt_cd8_genes = DotPlot(subset(obj.integrated, idents=idents.cd8), group.by = "mpoint", features = features.cd8.mail)
save_plot(dplt_cd8_genes, "todoplots/8/cd8cells.tp_all.dotplot.CTSW", 12 , 4)

## SYMPT
vplt_cd8_genes = VlnPlot(subset(obj.integrated, cells=intersect(cells.cd8, cells.symptomatic)), group.by = "mpoint", features = features.cd8.mail, assay = "RNA", ncol=4, pt.size = 0, cols=mpCols)+ theme(legend.position = 'none')
title <- ggdraw() + draw_label("CD8-Cells (symptomatic, by TPs)", fontface = 'bold')
vplt_cd8_genes=cowplot::plot_grid(title, vplt_cd8_genes, ncol = 1, rel_heights = c(0.1, 1))
save_plot(vplt_cd8_genes, "todoplots/8/cd8cells.tp_sympt.violin.CTSW", 12 , 8)

dplt_cd8_genes = DotPlot(subset(obj.integrated, cells=intersect(cells.cd8, cells.symptomatic)), group.by = "mpoint", features = features.cd8.mail)
save_plot(dplt_cd8_genes, "todoplots/8/cd8cells.tp_sympt.dotplot.CTSW", 12 , 4)

## ASYMPT
vplt_cd8_genes = VlnPlot(subset(obj.integrated, cells=intersect(cells.cd8, cells.asymptomatic)), group.by = "mpoint", features = features.cd8.mail, assay = "RNA", ncol=4, pt.size = 0, cols=mpCols)+ theme(legend.position = 'none')
title <- ggdraw() + draw_label("CD8-Cells (asymptomatic, by TPs)", fontface = 'bold')
vplt_cd8_genes=cowplot::plot_grid(title, vplt_cd8_genes, ncol = 1, rel_heights = c(0.1, 1))
save_plot(vplt_cd8_genes, "todoplots/8/cd8cells.tp_asympt.violin.CTSW", 12 , 8)

dplt_cd8_genes = DotPlot(subset(obj.integrated, cells=intersect(cells.cd8, cells.asymptomatic)), group.by = "mpoint", features = features.cd8.mail)
save_plot(dplt_cd8_genes, "todoplots/8/cd8cells.tp_asympt.dotplot.CTSW", 12 , 4)

plotElems = list()
plotElems[["Control"]] = list(cells=intersect(cells.control, cells.cd8), label="Control")
plotElems[["Symptomatic"]] = list(cells=intersect(cells.symptomatic, cells.cd8), label="Symptomatic")
plotElems[["Asymptomatic"]] = list(cells=intersect(cells.asymptomatic, cells.cd8), label="Asymptomatic")

sbsPlot_cd8_todo = makeSideBySideDotPlot(obj.integrated, plotElems = plotElems, featureGenes=features.cd8.mail, group.by = "mpoint", rel.width.legend = 0.005)
plot(sbsPlot_cd8_todo)

save_plot(sbsPlot_cd8_todo, "todoplots/8/cd8cells.by.tp.ds",fig.height=6, fig.width=10)


```
# Die Gene GZMB, GZMH,S100A8, S100A9 in NK Zellen, hier exakt dasselbe wie in Punkt 6

```{r}

idents.nk

```


```{r fig.height=8, fig.width=10}

features.nk.detail = c("GZMB", "LGALS1", "S100A8", "S100A9")

vplt_nk_gzbm_genes = VlnPlot(subset(obj.integrated, idents=idents.nk), group.by = "disease_state", features = features.nk.detail, assay = "RNA", ncol=4, pt.size = 0, cols=dsCols)+ theme(legend.position = 'none')
title <- ggdraw() + draw_label("NK-Cells (All TPs)", fontface = 'bold')
vplt_nk_gzbm_genes=cowplot::plot_grid(title, vplt_nk_gzbm_genes, ncol = 1, rel_heights = c(0.1, 1))
save_plot(vplt_nk_gzbm_genes, "todoplots/9/nkcells.violin.GZMB", 12 , 4)

dplt_nk_gzbm_genes = DotPlot(subset(obj.integrated, idents=idents.nk), group.by = "disease_state", features = features.nk.detail)
save_plot(dplt_nk_gzbm_genes, "todoplots/9/nkcells.dotplot.GZMB", 12 , 4)

vplt_nk_gzbm_genes = VlnPlot(subset(obj.integrated, idents=idents.nk), group.by = "mpoint", features = features.nk.detail, assay = "RNA", ncol=4, pt.size = 0, cols=mpCols)+ theme(legend.position = 'none')
title <- ggdraw() + draw_label("NK-Cells (By TPs)", fontface = 'bold')
vplt_nk_gzbm_genes=cowplot::plot_grid(title, vplt_nk_gzbm_genes, ncol = 1, rel_heights = c(0.1, 1))
save_plot(vplt_nk_gzbm_genes, "todoplots/9/nkcells.tp.violin.GZMB", 12 , 4)

dplt_nk_gzbm_genes = DotPlot(subset(obj.integrated, idents=idents.nk), group.by = "mpoint", features = features.nk.detail)
save_plot(dplt_nk_gzbm_genes, "todoplots/9/nkcells.tp.dotplot.GZMB", 12 , 4)


#SYMPT

vplt_nk_gzbm_genes = VlnPlot(subset(obj.integrated, cells=intersect(cells.nk, cells.symptomatic)), group.by = "mpoint", features = features.nk.detail, assay = "RNA", ncol=4, pt.size = 0, cols=mpCols)+ theme(legend.position = 'none')
title <- ggdraw() + draw_label("NK-Cells (Symptomatic, By TPs)", fontface = 'bold')
vplt_nk_gzbm_genes=cowplot::plot_grid(title, vplt_nk_gzbm_genes, ncol = 1, rel_heights = c(0.1, 1))
save_plot(vplt_nk_gzbm_genes, "todoplots/9/nkcells.tp_sympt.violin.GZMB", 12 , 4)

dplt_nk_gzbm_genes = DotPlot(subset(obj.integrated, cells=intersect(cells.nk, cells.symptomatic)), group.by = "mpoint", features = features.nk.detail)
save_plot(dplt_nk_gzbm_genes, "todoplots/9/nkcells.tp_sympt.dotplot.GZMB", 12 , 4)


#ASYMPT

vplt_nk_gzbm_genes = VlnPlot(subset(obj.integrated, cells=intersect(cells.nk, cells.asymptomatic)), group.by = "mpoint", features = features.nk.detail, assay = "RNA", ncol=4, pt.size = 0, cols=mpCols)+ theme(legend.position = 'none')
title <- ggdraw() + draw_label("NK-Cells (Asymptomatic, By TPs)", fontface = 'bold')
vplt_nk_gzbm_genes=cowplot::plot_grid(title, vplt_nk_gzbm_genes, ncol = 1, rel_heights = c(0.1, 1))
save_plot(vplt_nk_gzbm_genes, "todoplots/9/nkcells.tp_asympt.violin.GZMB", 12 , 4)

dplt_nk_gzbm_genes = DotPlot(subset(obj.integrated, cells=intersect(cells.nk, cells.asymptomatic)), group.by = "mpoint", features = features.nk.detail)
save_plot(dplt_nk_gzbm_genes, "todoplots/9/nkcells.tp_asympt.dotplot.GZMB", 12 , 4)


plotElems = list()
plotElems[["Control"]] = list(cells=intersect(cells.control, cells.nk), label="Control")
plotElems[["Symptomatic"]] = list(cells=intersect(cells.symptomatic, cells.nk), label="Symptomatic")
plotElems[["Asymptomatic"]] = list(cells=intersect(cells.asymptomatic, cells.nk), label="Asymptomatic")

sbsPlot_nk_gzbm_genes = makeSideBySideDotPlot(obj.integrated, plotElems = plotElems, featureGenes=features.nk.detail, group.by = "mpoint", rel.width.legend = 0.005)
plot(sbsPlot_nk_gzbm_genes)

save_plot(sbsPlot_nk_gzbm_genes, "todoplots/9/genes_by_mpoint",fig.height=6, fig.width=10)

```


```{r}
source("sbs_plot.R")

```


```{r fig.height=5, fig.width=10}

plotElems = list()
plotElems[["Non-pneumonic"]] = list(cells=intersect(cells.monocyte, cells.asymptomatic), label="Monocytes ASYM")
plotElems[["Pneumonic"]] = list(cells=intersect(cells.monocyte, cells.symptomatic), label="Monocytes SYM")

sbsPlot = makeSideBySideDotPlot(obj.integrated, plotElems = plotElems, featureGenes=unique(c("MX1", "MX2", "XAF1", "IFI44L", "IFI44", "IFI6", "LY6E", "ISG15", "IFITM1")), group.by = "mpoint", rel.width.legend = 0.01, plot.title = "", cols=c("blue", "grey", "red"))
sbsPlot = cowplot::plot_grid(ggdraw() + draw_label("Monocytes", fontface='bold'), sbsPlot, ncol =1, rel_heights = c(0.1,1.0))

save_plot(sbsPlot, "last_minute_plots/sbsdp_involving_isgs_monocytes", fig.width = 10, fig.height = 5)

```
```{r fig.height=5, fig.width=10}

plotElems = list()
plotElems[["Non-pneumonic"]] = list(cells=intersect(cells.bcell, cells.asymptomatic), label="B-cells ASYM")
plotElems[["Pneumonic"]] = list(cells=intersect(cells.bcell, cells.symptomatic), label="B-cells SYM")

sbsPlot = makeSideBySideDotPlot(obj.integrated, plotElems = plotElems, featureGenes=unique(c("MX1", "MX2", "XAF1", "IFI44L", "IFI44", "IFI6", "LY6E", "ISG15", "IFITM1")), group.by = "mpoint", rel.width.legend = 0.01, plot.title = "", cols=c("blue", "grey", "red"))
sbsPlot = cowplot::plot_grid(ggdraw() + draw_label("B-cells", fontface='bold'), sbsPlot, ncol =1, rel_heights = c(0.1,1.0))

save_plot(sbsPlot, "last_minute_plots/sbsdp_involving_isgs_bcells", fig.width = 10, fig.height = 5)

```

```{r fig.height=5, fig.width=10}

plotElems = list()
plotElems[["Non-pneumonic"]] = list(cells=intersect(cells.cd4t, cells.asymptomatic), label="CD4+ T-cells ASYM")
plotElems[["Pneumonic"]] = list(cells=intersect(cells.cd4t, cells.symptomatic), label="CD4+ T-cells SYM")

sbsPlot = makeSideBySideDotPlot(obj.integrated, plotElems = plotElems, featureGenes=unique(c("MX1", "MX2", "XAF1", "IFI44L", "IFI44", "IFI6", "LY6E", "ISG15", "IFITM1")), group.by = "mpoint", rel.width.legend = 0.01, plot.title = "", cols=c("blue", "grey", "red"))
sbsPlot = cowplot::plot_grid(ggdraw() + draw_label("CD4+ T-cells", fontface='bold'), sbsPlot, ncol =1, rel_heights = c(0.1,1.0))

save_plot(sbsPlot, "last_minute_plots/sbsdp_involving_isgs_CD4_t_cells", fig.width = 10, fig.height = 5)

```

```{r fig.height=5, fig.width=10}

plotElems = list()
plotElems[["Non-pneumonic"]] = list(cells=intersect(cells.nk, cells.asymptomatic), label="NK-cells ASYM")
plotElems[["Pneumonic"]] = list(cells=intersect(cells.nk, cells.symptomatic), label="NK-cells SYM")

sbsPlot = makeSideBySideDotPlot(obj.integrated, plotElems = plotElems, featureGenes=unique(c("MX1", "MX2", "XAF1", "IFI44L", "IFI44", "IFI6", "LY6E", "ISG15", "IFITM1")), group.by = "mpoint", rel.width.legend = 0.01, plot.title = "", cols=c("blue", "grey", "red"))
sbsPlot = cowplot::plot_grid(ggdraw() + draw_label("NK-cells", fontface='bold'), sbsPlot, ncol =1, rel_heights = c(0.1,1.0))

save_plot(sbsPlot, "last_minute_plots/sbsdp_involving_isgs_NK_cells", fig.width = 10, fig.height = 5)

```



```{r fig.height=5, fig.width=10}

plotElems = list()
plotElems[["Non-pneumonic"]] = list(cells=intersect(cells.nk, cells.asymptomatic), label="NK-cells ASYM")
plotElems[["Pneumonic"]] = list(cells=intersect(cells.nk, cells.symptomatic), label="NK-cells SYM")

sbsPlot = makeSideBySideDotPlot(obj.integrated, plotElems = plotElems, featureGenes=unique(c("JUN", "JUNB", "FOS", "FOSB", "CCL3", "CCL4", "THEMIS2", "GZMB", "S100A8", "S100A9")), group.by = "mpoint", rel.width.legend = 0.01, plot.title = "", cols=c("blue", "grey", "red"))
sbsPlot = cowplot::plot_grid(ggdraw() + draw_label("NK-cells", fontface='bold'), sbsPlot, ncol =1, rel_heights = c(0.1,1.0))

save_plot(sbsPlot, "last_minute_plots/sbsdp_concentrating_NK_cells", fig.width = 10, fig.height = 5)

```
```{r fig.height=5, fig.width=10}

plotElems = list()
plotElems[["Non-pneumonic"]] = list(cells=intersect(cells.monocyte, cells.asymptomatic), label="Monocytes ASYM")
plotElems[["Pneumonic"]] = list(cells=intersect(cells.monocyte, cells.symptomatic), label="Monocytes SYM")

sbsPlot = makeSideBySideDotPlot(obj.integrated, plotElems = plotElems, featureGenes=unique(c("AREG", "EREG", "CD83", "TMEM176B", "EGR3", "HBEGF", "IL1RN")), group.by = "mpoint", rel.width.legend = 0.01, plot.title = "", cols=c("blue", "grey", "red"))
sbsPlot = cowplot::plot_grid(ggdraw() + draw_label("Monocytes", fontface='bold'), sbsPlot, ncol =1, rel_heights = c(0.1,1.0))

save_plot(sbsPlot, "last_minute_plots/sbsdp_involving_state_of_monocytes", fig.width = 10, fig.height = 5)

```

```{r fig.height=8, fig.width=10}

detailPlotList = list("tcells"=idents.tcell, "nkcells"=idents.nk, "bcells"=idents.bcell, "cd4tcells"=idents.cd4t, "cd8cells"=idents.cd8, "monocytes"=idents.monocytes)

features.detailplots = c("MX1", "MX2", "XAF1", "IFI44L", "IFI44", "IFI6", "LY6E", "ISG15", "IFITM1")

allMpoints = unique(obj.integrated$mpoint)
allDiseaseStates = unique(obj.integrated$disease_state)

for (cellName in names(detailPlotList))
{
  selIdents = detailPlotList[[cellName]]
  
  vplt = VlnPlot(subset(obj.integrated, idents=selIdents), group.by = "disease_state", features = features.detailplots, assay = "RNA", ncol=3, pt.size = 0, cols=dsCols)+ theme(legend.position = 'none')
  title <- ggdraw() + draw_label(paste(cellName, "(All TPs)"), fontface = 'bold')
  vplt=cowplot::plot_grid(title, vplt, ncol = 1, rel_heights = c(0.1, 1))
  save_plot(vplt, paste("todoplots/10_11/",cellName,"/", cellName,".violin.details", sep=""), 12 , 8)
  
  dplt = DotPlot(subset(obj.integrated, idents=selIdents), group.by = "disease_state", features = features.detailplots)
  save_plot(dplt, paste("todoplots/10_11/",cellName,"/", cellName,".dotplot.details", sep=""), 12 , 4)
  
  vplt_tp = VlnPlot(subset(obj.integrated, idents=selIdents), group.by = "mpoint", features = features.detailplots, assay = "RNA", ncol=4, pt.size = 0, cols=mpCols)+ theme(legend.position = 'none')
  title <- ggdraw() + draw_label(paste(cellName, "(By TPs)"), fontface = 'bold')
  vplt_tp=cowplot::plot_grid(title, vplt_tp, ncol = 1, rel_heights = c(0.1, 1))
  save_plot(vplt_tp, paste("todoplots/10_11/",cellName,"/", cellName,".violin.tp.details", sep=""), 12 , 8)
  
  dplt_tp = DotPlot(subset(obj.integrated, idents=selIdents), group.by = "mpoint", features = features.detailplots)
  save_plot(dplt_tp, paste("todoplots/10_11/",cellName,"/", cellName,".dotplot.tp.details", sep=""), 12 , 4)
  
  selCells = makeGrpCells(scdata = obj.integrated, clusterIdents = selIdents)

  for (ds in allDiseaseStates)
  {
    
    dsCells = names(obj.integrated$disease_state[obj.integrated$disease_state==ds])
    
    selDsCells = intersect(dsCells, selCells)
    selDsCells = setdiff(selDsCells, cells.control)
    
    print(paste(ds, length(selDsCells)))
    
    if (length(selDsCells) == 0)
    {
      next;
    }
    
    vplt_tp = VlnPlot(subset(obj.integrated, cells=selDsCells), group.by = "mpoint", features = features.detailplots, assay = "RNA", ncol=3, pt.size = 0, cols=mpCols)+ theme(legend.position = 'none')
  title <- ggdraw() + draw_label(paste(cellName, "(",ds,")"), fontface = 'bold')
  vplt_tp=cowplot::plot_grid(title, vplt_tp, ncol = 1, rel_heights = c(0.1, 1))
  save_plot(vplt_tp, paste("todoplots/10_11/",cellName,"/", cellName,".violin.tp.",ds,".details", sep=""), 12 , 8)
  
  dplt_tp = DotPlot(subset(obj.integrated, cells=selDsCells), group.by = "mpoint", features = features.detailplots)
  save_plot(dplt_tp, paste("todoplots/10_11/",cellName,"/", cellName,".dotplot.tp.",ds,".details", sep=""), 12 , 4)
    
  }
  
  plotElems = list()
  plotElems[["Control"]] = list(cells=intersect(cells.control, selCells), label="Control")
  plotElems[["Symptomatic"]] = list(cells=intersect(cells.symptomatic, selCells), label="Symptomatic")
  plotElems[["Asymptomatic"]] = list(cells=intersect(cells.asymptomatic, selCells), label="Asymptomatic")
  
  sbsPlot_nk_gzbm_genes = makeSideBySideDotPlot(obj.integrated, plotElems = plotElems, featureGenes=features.detailplots, group.by = "mpoint", rel.width.legend = 0.005)
  plot(sbsPlot_nk_gzbm_genes)
  
  save_plot(sbsPlot_nk_gzbm_genes, paste("todoplots/10_11/",cellName,"/", cellName,".genes_by_mpoint.details", sep=""),fig.height=6, fig.width=10)
  
}






```


```{r fig.height=6, fig.width=12}

features.tcell.mail = c("FOS", "JUN", "LY6E")


vplt_tcell_genes_2 = VlnPlot(subset(obj.integrated, idents=idents.tcell), group.by = "disease_state", split.by ="idents", features = features.tcell.mail, assay = "RNA", ncol=4, pt.size = 0.0001)+ theme(legend.position = 'right')
title <- ggdraw() + draw_label("T-Cells (All TPs)", fontface = 'bold')
vplt_tcell_genes_2=cowplot::plot_grid(title, vplt_tcell_genes_2, ncol = 1, rel_heights = c(0.1, 1))
vplt_tcell_genes_2


```

```{r fig.width=30, fig.height=15}

VlnPlot(obj.integrated, features=c("IFITM1", "IFI44L", "JUN", "JUNB", "FOS", "FOSB", "CCL3", "CCL4", "THEMIS2", "GZMB", "S100A8", "S100A9"), group.by = "cellnames", split.by = "mpoint")

```

```{r}

DotPlot(obj.integrated, features=c("IFITM1", "IFI44L", "JUN", "JUNB", "FOS", "FOSB", "CCL3", "CCL4", "THEMIS2", "GZMB", "S100A8", "S100A9"), group.by = "idents")

```



```{r}

tcell.sympt_ctrl = compareCells(obj.integrated,
             intersect(cells.symptomatic, cells.tcell),
             intersect(cells.control, cells.tcell),
             "symptomatic", "control", prefix="tcell", test="t", outfolder="detests_3")

bcell.sympt_ctrl = compareCells(obj.integrated,
             intersect(cells.symptomatic, cells.bcell),
             intersect(cells.control, cells.bcell),
             "symptomatic", "control", prefix="bcell", test="t", outfolder="detests_3")

nk.sympt_ctrl = compareCells(obj.integrated,
             intersect(cells.symptomatic, cells.nk),
             intersect(cells.control, cells.nk),
             "symptomatic", "control", prefix="nk", test="t", outfolder="detests_3")

monocytes.sympt_ctrl = compareCells(obj.integrated,
             intersect(cells.symptomatic, cells.monocyte),
             intersect(cells.control, cells.monocyte),
             "symptomatic", "control", prefix="monocytes", test="t", outfolder="detests_3")

gdtcell.sympt_ctrl = compareCells(obj.integrated,
             intersect(cells.symptomatic, cells.gdtcell),
             intersect(cells.control, cells.gdtcell),
             "symptomatic", "control", prefix="gdtcell", test="t", outfolder="detests_3")

plasma.sympt_ctrl = compareCells(obj.integrated,
             intersect(cells.symptomatic, cells.plasma),
             intersect(cells.control, cells.plasma),
             "symptomatic", "control", prefix="plasma", test="t", outfolder="detests_3")

pdc.sympt_ctrl = compareCells(obj.integrated,
             intersect(cells.symptomatic, cells.pdc),
             intersect(cells.control, cells.pdc),
             "symptomatic", "control", prefix="pdc", test="t", outfolder="detests_3")

dendritic.sympt_ctrl = compareCells(obj.integrated,
             intersect(cells.symptomatic, cells.dendritic),
             intersect(cells.control, cells.dendritic),
             "symptomatic", "control", prefix="dendritic", test="t", outfolder="detests_3")

#
#
#
#


tcell.asympt_ctrl = compareCells(obj.integrated,
             intersect(cells.asymptomatic, cells.tcell),
             intersect(cells.control, cells.tcell),
             "asymptomatic", "control", prefix="tcell", test="t", outfolder="detests_3")

bcell.asympt_ctrl = compareCells(obj.integrated,
             intersect(cells.asymptomatic, cells.bcell),
             intersect(cells.control, cells.bcell),
             "asymptomatic", "control", prefix="bcell", test="t", outfolder="detests_3")

nk.asympt_ctrl = compareCells(obj.integrated,
             intersect(cells.asymptomatic, cells.nk),
             intersect(cells.control, cells.nk),
             "asymptomatic", "control", prefix="nk", test="t", outfolder="detests_3")

monocytes.asympt_ctrl = compareCells(obj.integrated,
             intersect(cells.asymptomatic, cells.monocyte),
             intersect(cells.control, cells.monocyte),
             "asymptomatic", "control", prefix="monocytes", test="t", outfolder="detests_3")

gdtcell.asympt_ctrl = compareCells(obj.integrated,
             intersect(cells.asymptomatic, cells.gdtcell),
             intersect(cells.control, cells.gdtcell),
             "asymptomatic", "control", prefix="gdtcell", test="t", outfolder="detests_3")

plasma.asympt_ctrl = compareCells(obj.integrated,
             intersect(cells.asymptomatic, cells.plasma),
             intersect(cells.control, cells.plasma),
             "asymptomatic", "control", prefix="plasma", test="t", outfolder="detests_3")

pdc.asympt_ctrl = compareCells(obj.integrated,
             intersect(cells.asymptomatic, cells.pdc),
             intersect(cells.control, cells.pdc),
             "asymptomatic", "control", prefix="pdc", test="t", outfolder="detests_3")

dendritic.asympt_ctrl = compareCells(obj.integrated,
             intersect(cells.asymptomatic, cells.dendritic),
             intersect(cells.control, cells.dendritic),
             "asymptomatic", "control", prefix="dendritic", test="t", outfolder="detests_3")

#
#
#
#


tcell.asympt_sympt = compareCells(obj.integrated,
             intersect(cells.asymptomatic, cells.tcell),
             intersect(cells.symptomatic, cells.tcell),
             "asymptomatic", "symptomatic", prefix="tcell", test="t", outfolder="detests_3")

bcell.asympt_sympt = compareCells(obj.integrated,
             intersect(cells.asymptomatic, cells.bcell),
             intersect(cells.symptomatic, cells.bcell),
             "asymptomatic", "symptomatic", prefix="bcell", test="t", outfolder="detests_3")

nk.asympt_sympt = compareCells(obj.integrated,
             intersect(cells.asymptomatic, cells.nk),
             intersect(cells.symptomatic, cells.nk),
             "asymptomatic", "symptomatic", prefix="nk", test="t", outfolder="detests_3")

monocytes.asympt_sympt = compareCells(obj.integrated,
             intersect(cells.asymptomatic, cells.monocyte),
             intersect(cells.symptomatic, cells.monocyte),
             "asymptomatic", "symptomatic", prefix="monocytes", test="t", outfolder="detests_3")

gdtcell.asympt_sympt = compareCells(obj.integrated,
             intersect(cells.asymptomatic, cells.gdtcell),
             intersect(cells.symptomatic, cells.gdtcell),
             "asymptomatic", "symptomatic", prefix="gdtcell", test="t", outfolder="detests_3")

plasma.asympt_sympt = compareCells(obj.integrated,
             intersect(cells.asymptomatic, cells.plasma),
             intersect(cells.symptomatic, cells.plasma),
             "asymptomatic", "symptomatic", prefix="plasma", test="t", outfolder="detests_3")

pdc.asympt_sympt = compareCells(obj.integrated,
             intersect(cells.asymptomatic, cells.pdc),
             intersect(cells.symptomatic, cells.pdc),
             "asymptomatic", "symptomatic", prefix="pdc", test="t", outfolder="detests_3")

dendritic.asympt_sympt = compareCells(obj.integrated,
             intersect(cells.asymptomatic, cells.dendritic),
             intersect(cells.symptomatic, cells.dendritic),
             "asymptomatic", "symptomatic", prefix="dendritic", test="t", outfolder="detests_3")


asympt_sympt_de_list = list(
  tcell.asympt_sympt=tcell.asympt_sympt,
  bcell.asympt_sympt=bcell.asympt_sympt,
  nk.asympt_sympt=nk.asympt_sympt,
  monocytes.asympt_sympt=monocytes.asympt_sympt,
  gdtcell.asympt_sympt=gdtcell.asympt_sympt,
  plasma.asympt_sympt=plasma.asympt_sympt,
  pdc.asympt_sympt=pdc.asympt_sympt,
  dendritic.sympt_ctrl=dendritic.sympt_ctrl,

  tcell.asympt_ctrl=tcell.asympt_ctrl,
  bcell.asympt_ctrl=bcell.asympt_ctrl,
  nk.asympt_ctrl=nk.asympt_ctrl,
  monocytes.asympt_ctrl=monocytes.asympt_ctrl,
  gdtcell.asympt_ctrl=gdtcell.asympt_ctrl,
  plasma.asympt_ctrl=plasma.asympt_ctrl,
  pdc.asympt_ctrl=pdc.asympt_ctrl,
  dendritic.asympt_ctrl=dendritic.asympt_ctrl,
  
  tcell.sympt_ctrl=tcell.sympt_ctrl,
  bcell.sympt_ctrl=bcell.sympt_ctrl,
  nk.sympt_ctrl=nk.sympt_ctrl,
  monocytes.sympt_ctrl=monocytes.sympt_ctrl,
  gdtcell.sympt_ctrl=gdtcell.sympt_ctrl,
  plasma.sympt_ctrl=plasma.sympt_ctrl,
  pdc.sympt_ctrl=pdc.sympt_ctrl,
  dendritic.asympt_sympt=dendritic.asympt_sympt
)

```

```{r eval=F}
lapply(names(asympt_sympt_de_list), function(x) {
  
  joinedData = asympt_sympt_de_list[[x]]
  outfolder="detests_3"
  
  outfile=paste(outfolder, "/", x,".tsv", sep="")
  write.table(joinedData, file=outfile, row.names = F,  quote=FALSE, sep='\t')
  
  outfile=paste(outfolder, "/", x,".xlsx", sep="")
  write.xlsx(joinedData, file=outfile, row.names = F)
})
```

```{r}

tcell.asympt_sympt_tp1 = compareCells(obj.integrated,
             intersect(intersect(cells.asymptomatic, cells.tp1), cells.tcell),
             intersect(intersect(cells.symptomatic, cells.tp1), cells.tcell),
             "asymptomatic", "symptomatic", prefix="tcell.tp1", test="t", outfolder="detests_4")

tcell.asympt_sympt_tp2 = compareCells(obj.integrated,
             intersect(intersect(cells.asymptomatic, cells.tp2), cells.tcell),
             intersect(intersect(cells.symptomatic, cells.tp2), cells.tcell),
             "asymptomatic", "symptomatic", prefix="tcell.tp2", test="t", outfolder="detests_4")

tcell.asympt_sympt_tp3 = compareCells(obj.integrated,
             intersect(intersect(cells.asymptomatic, cells.tp3), cells.tcell),
             intersect(intersect(cells.symptomatic, cells.tp3), cells.tcell),
             "asymptomatic", "symptomatic", prefix="tcell.tp3", test="t", outfolder="detests_4")

#
#
#
#
#
#
#
#

bcell.asympt_sympt_tp1 = compareCells(obj.integrated,
             intersect(intersect(cells.asymptomatic, cells.tp1), cells.bcell),
             intersect(intersect(cells.symptomatic, cells.tp1), cells.bcell),
             "asymptomatic", "symptomatic", prefix="bcell.tp1", test="t", outfolder="detests_4")

bcell.asympt_sympt_tp2 = compareCells(obj.integrated,
             intersect(intersect(cells.asymptomatic, cells.tp2), cells.bcell),
             intersect(intersect(cells.symptomatic, cells.tp2), cells.bcell),
             "asymptomatic", "symptomatic", prefix="bcell.tp2", test="t", outfolder="detests_4")

bcell.asympt_sympt_tp3 = compareCells(obj.integrated,
             intersect(intersect(cells.asymptomatic, cells.tp3), cells.bcell),
             intersect(intersect(cells.symptomatic, cells.tp3), cells.bcell),
             "asymptomatic", "symptomatic", prefix="bcell.tp3", test="t", outfolder="detests_4")

#
#
#
#
#
#
#
#

nk.asympt_sympt_tp1 = compareCells(obj.integrated,
             intersect(intersect(cells.asymptomatic, cells.tp1), cells.nk),
             intersect(intersect(cells.symptomatic, cells.tp1), cells.nk),
             "asymptomatic", "symptomatic", prefix="nk.tp1", test="t", outfolder="detests_4")

nk.asympt_sympt_tp2 = compareCells(obj.integrated,
             intersect(intersect(cells.asymptomatic, cells.tp2), cells.nk),
             intersect(intersect(cells.symptomatic, cells.tp2), cells.nk),
             "asymptomatic", "symptomatic", prefix="nk.tp2", test="t", outfolder="detests_4")

nk.asympt_sympt_tp3 = compareCells(obj.integrated,
             intersect(intersect(cells.asymptomatic, cells.tp3), cells.nk),
             intersect(intersect(cells.symptomatic, cells.tp3), cells.nk),
             "asymptomatic", "symptomatic", prefix="nk.tp3", test="t", outfolder="detests_4")

#
#
#
#
#
#
#
#

monocytes.asympt_sympt_tp1 = compareCells(obj.integrated,
             intersect(intersect(cells.asymptomatic, cells.tp1), cells.monocyte),
             intersect(intersect(cells.symptomatic, cells.tp1), cells.monocyte),
             "asymptomatic", "symptomatic", prefix="monocytes.tp1", test="t", outfolder="detests_4")

monocytes.asympt_sympt_tp2 = compareCells(obj.integrated,
             intersect(intersect(cells.asymptomatic, cells.tp2), cells.monocyte),
             intersect(intersect(cells.symptomatic, cells.tp2), cells.monocyte),
             "asymptomatic", "symptomatic", prefix="monocytes.tp2", test="t", outfolder="detests_4")

monocytes.asympt_sympt_tp3 = compareCells(obj.integrated,
             intersect(intersect(cells.asymptomatic, cells.tp3), cells.monocyte),
             intersect(intersect(cells.symptomatic, cells.tp3), cells.monocyte),
             "asymptomatic", "symptomatic", prefix="monocytes.tp3", test="t", outfolder="detests_4")

#
#
#
#
#
#
#
#

gdtcell.asympt_sympt_tp1 = compareCells(obj.integrated,
             intersect(intersect(cells.asymptomatic, cells.tp1), cells.gdtcell),
             intersect(intersect(cells.symptomatic, cells.tp1), cells.gdtcell),
             "asymptomatic", "symptomatic", prefix="gdtcell.tp1", test="t", outfolder="detests_4")

gdtcell.asympt_sympt_tp2 = compareCells(obj.integrated,
             intersect(intersect(cells.asymptomatic, cells.tp2), cells.gdtcell),
             intersect(intersect(cells.symptomatic, cells.tp2), cells.gdtcell),
             "asymptomatic", "symptomatic", prefix="gdtcell.tp2", test="t", outfolder="detests_4")

gdtcell.asympt_sympt_tp3 = compareCells(obj.integrated,
             intersect(intersect(cells.asymptomatic, cells.tp3), cells.gdtcell),
             intersect(intersect(cells.symptomatic, cells.tp3), cells.gdtcell),
             "asymptomatic", "symptomatic", prefix="gdtcell.tp3", test="t", outfolder="detests_4")

#
#
#
#
#
#
#
#

plasma.asympt_sympt_tp1 = compareCells(obj.integrated,
             intersect(intersect(cells.asymptomatic, cells.tp1), cells.plasma),
             intersect(intersect(cells.symptomatic, cells.tp1), cells.plasma),
             "asymptomatic", "symptomatic", prefix="plasma.tp1", test="t", outfolder="detests_4")

plasma.asympt_sympt_tp2 = compareCells(obj.integrated,
             intersect(intersect(cells.asymptomatic, cells.tp2), cells.plasma),
             intersect(intersect(cells.symptomatic, cells.tp2), cells.plasma),
             "asymptomatic", "symptomatic", prefix="plasma.tp2", test="t", outfolder="detests_4")

plasma.asympt_sympt_tp3 = compareCells(obj.integrated,
             intersect(intersect(cells.asymptomatic, cells.tp3), cells.plasma),
             intersect(intersect(cells.symptomatic, cells.tp3), cells.plasma),
             "asymptomatic", "symptomatic", prefix="plasma.tp3", test="t", outfolder="detests_4")

#
#
#
#
#
#
#
#

pdc.asympt_sympt_tp1 = compareCells(obj.integrated,
             intersect(intersect(cells.asymptomatic, cells.tp1), cells.pdc),
             intersect(intersect(cells.symptomatic, cells.tp1), cells.pdc),
             "asymptomatic", "symptomatic", prefix="pdc.tp1", test="t", outfolder="detests_4")

pdc.asympt_sympt_tp2 = compareCells(obj.integrated,
             intersect(intersect(cells.asymptomatic, cells.tp2), cells.pdc),
             intersect(intersect(cells.symptomatic, cells.tp2), cells.pdc),
             "asymptomatic", "symptomatic", prefix="pdc.tp2", test="t", outfolder="detests_4")

pdc.asympt_sympt_tp3 = compareCells(obj.integrated,
             intersect(intersect(cells.asymptomatic, cells.tp3), cells.pdc),
             intersect(intersect(cells.symptomatic, cells.tp3), cells.pdc),
             "asymptomatic", "symptomatic", prefix="pdc.tp3", test="t", outfolder="detests_4")

#
#
#
#
#
#
#
#

dendritic.asympt_sympt_tp1 = compareCells(obj.integrated,
             intersect(intersect(cells.asymptomatic, cells.tp1), cells.dendritic),
             intersect(intersect(cells.symptomatic, cells.tp1), cells.dendritic),
             "asymptomatic", "symptomatic", prefix="dendritic.tp1", test="t", outfolder="detests_4")

dendritic.asympt_sympt_tp2 = compareCells(obj.integrated,
             intersect(intersect(cells.asymptomatic, cells.tp2), cells.dendritic),
             intersect(intersect(cells.symptomatic, cells.tp2), cells.dendritic),
             "asymptomatic", "symptomatic", prefix="dendritic.tp2", test="t", outfolder="detests_4")

dendritic.asympt_sympt_tp3 = compareCells(obj.integrated,
             intersect(intersect(cells.asymptomatic, cells.tp3), cells.dendritic),
             intersect(intersect(cells.symptomatic, cells.tp3), cells.dendritic),
             "asymptomatic", "symptomatic", prefix="dendritic.tp3", test="t", outfolder="detests_4")

tp_de_list = list(
  pdc.asympt_sympt_tp1=pdc.asympt_sympt_tp1,
  pdc.asympt_sympt_tp2=pdc.asympt_sympt_tp2,
  pdc.asympt_sympt_tp3=pdc.asympt_sympt_tp3,
  
  plasma.asympt_sympt_tp1=plasma.asympt_sympt_tp1,
  plasma.asympt_sympt_tp2=plasma.asympt_sympt_tp2,
  plasma.asympt_sympt_tp3=plasma.asympt_sympt_tp3,
  
  gdtcell.asympt_sympt_tp1=gdtcell.asympt_sympt_tp1,
  gdtcell.asympt_sympt_tp2=gdtcell.asympt_sympt_tp2,
  gdtcell.asympt_sympt_tp3=gdtcell.asympt_sympt_tp3,
  
  monocytes.asympt_sympt_tp1=monocytes.asympt_sympt_tp1,
  monocytes.asympt_sympt_tp2=monocytes.asympt_sympt_tp2,
  monocytes.asympt_sympt_tp3=monocytes.asympt_sympt_tp3,
  
  nk.asympt_sympt_tp1=nk.asympt_sympt_tp1,
  nk.asympt_sympt_tp2=nk.asympt_sympt_tp2,
  nk.asympt_sympt_tp3=nk.asympt_sympt_tp3,
  
  bcell.asympt_sympt_tp1=bcell.asympt_sympt_tp1,
  bcell.asympt_sympt_tp2=bcell.asympt_sympt_tp2,
  bcell.asympt_sympt_tp3=bcell.asympt_sympt_tp3,
  
  tcell.asympt_sympt_tp1=tcell.asympt_sympt_tp1,
  tcell.asympt_sympt_tp2=tcell.asympt_sympt_tp2,
  tcell.asympt_sympt_tp3=tcell.asympt_sympt_tp3,
  
  dendritic.asympt_sympt_tp1=dendritic.asympt_sympt_tp1,
  dendritic.asympt_sympt_tp2=dendritic.asympt_sympt_tp2,
  dendritic.asympt_sympt_tp3=dendritic.asympt_sympt_tp3
)


```

```{r}
lapply(names(tp_de_list), function(x) {
  
  joinedData = tp_de_list[[x]]
  outfolder="detests_4"
  
  outfile=paste(outfolder, "/", x,".tsv", sep="")
  write.table(joinedData, file=outfile, row.names = F,  quote=FALSE, sep='\t')
  
  outfile=paste(outfolder, "/", x,".xlsx", sep="")
  write.xlsx(joinedData, file=outfile, row.names = F)
})
```

## specific checks

```{r fig.height=8, fig.width=20}

genes.ifn = grep("^IFN", rownames(obj.integrated), value = T)

# CTSW 23, GZMM 24, KLRD1 25, PRF1 26, NKG7 27, GZMB 28, GZMH 29
vplt_ifn_genes = VlnPlot(obj.integrated, group.by = "disease_state" , features = genes.ifn , assay = "RNA", ncol = 4, pt.size = 0.001 )

dplt_ifn_genes = DotPlot(obj.integrated, group.by = "disease_state" , features = genes.ifn, assay = "RNA")+ theme(axis.text.x = element_text(angle = 90, hjust = 1))

save_plot(vplt_ifn_genes, "ifnplots/ifn_mail_genes_violin", fig.height = 24, fig.width = 20)
save_plot(dplt_ifn_genes, "ifnplots/ifn_mail_genes_dotplot", fig.height = 8, fig.width = 20)


for (clusterID in unique(obj.integrated$idents))
{
  vplt_ifn_genes = VlnPlot(subset(obj.integrated, idents=clusterID), group.by = "disease_state" , features = genes.ifn, assay = "RNA", ncol = 4, pt.size = 0.001 )+ ggtitle(paste("Cluster", clusterID))
dplt_ifn_genes = DotPlot(subset(obj.integrated, idents=clusterID), group.by = "disease_state" , features = genes.ifn, assay = "RNA") +theme(axis.text.x = element_text(angle = 90, hjust = 1)) + ggtitle(paste("Cluster", clusterID))

save_plot(vplt_ifn_genes, paste("ifnplots/ifn_mail_genes_violin", clusterID, sep="."), fig.height = 24, fig.width = 20)
save_plot(dplt_ifn_genes, paste("ifnplots/ifn_mail_genes_dotplot",clusterID, sep="."), fig.height = 8, fig.width = 20)
}

for (clusterID in unique(obj.integrated$unified_cellnames))
{
  clustername = str_replace_all(clusterID, " ", "_")
  
  vplt_ifn_genes = VlnPlot(subset(obj.integrated, unified_cellnames=clusterID), group.by = "disease_state" , features = genes.ifn, assay = "RNA", ncol = 4, pt.size = 0.001 )+ ggtitle(paste("Cluster", clusterID))
dplt_ifn_genes = DotPlot(subset(obj.integrated, unified_cellnames=clusterID), group.by = "disease_state" , features = genes.ifn, assay = "RNA") +theme(axis.text.x = element_text(angle = 90, hjust = 1))+ ggtitle(paste("Cluster", clusterID))

save_plot(vplt_ifn_genes, paste("ifnplots/ifn_mail_genes_violin", clustername, sep="."), fig.height = 24, fig.width = 20)
save_plot(dplt_ifn_genes, paste("ifnplots/ifn_mail_genes_dotplot",clustername, sep="."), fig.height = 8, fig.width = 20)
}

```

## GSE ANALYSIS

```{r}

saveRDS(rownames(obj.integrated), "go_universe.Rds")

saveRDS(asympt_sympt_de_list, "asympt_sympt_de_list.Rds")
saveRDS(tp_de_list, "tp_de_list.Rds")

```

```{bash}

mkdir -p ra_sympts
mkdir -p ra_tp

mkdir -p go_sympts
mkdir -p go_tp

mkdir -p gse_sympts
mkdir -p gse_tp
```

```{bash}

rm ra_sympts/*
rm ra_tp/*

Rscript raAnalysis.R "asympt_sympt_de_list.Rds" "go_universe.Rds" "human"
Rscript raAnalysis.R "tp_de_list.Rds" "go_universe.Rds" "human"

```

```{bash}

rm go_sympts/*
rm go_tp/*

Rscript goAnalysis.R "asympt_sympt_de_list.Rds" "go_universe.Rds"
Rscript goAnalysis.R "tp_de_list.Rds" "go_universe.Rds"

```

```{bash}

rm gse_sympts/*
rm gse_tp/*

Rscript gseAnalysis.R "asympt_sympt_de_list.Rds"
Rscript gseAnalysis.R "tp_de_list.Rds"

```


```{r}
asympt_sympt_de_list.ra = readRDS("ra.asympt_sympt_de_list.Rds")
tp_de_list.ra = readRDS("ra.tp_de_list.Rds")

asympt_sympt_de_list.go = readRDS("go.asympt_sympt_de_list.Rds")
tp_de_list.go = readRDS("go.tp_de_list.Rds")

asympt_sympt_de_list.gse = readRDS("gse.asympt_sympt_de_list.Rds")
tp_de_list.gse = readRDS("gse.tp_de_list.Rds")

```

```{r}
tp_de_list.ra=NULL
tp_de_list.go=NULL
tp_de_list.gse=NULL
```


```{r}

deGenes.Res.state.df = lapply(names(asympt_sympt_de_list), function(x) {
  return(list("degenes"=nrow(asympt_sympt_de_list[[x]][ asympt_sympt_de_list[[x]]$p_val_adj < 0.05, ])
, "valid"=asympt_sympt_de_list.gse[[x]]$valid))}) %>% map(as.data.frame) %>% bind_rows()
rownames(deGenes.Res.state.df) = names(asympt_sympt_de_list)
deGenes.Res.state.df

deGenes.Res.tp.df = lapply(names(tp_de_list), function(x) {
  return(list("degenes"=nrow(tp_de_list[[x]][ tp_de_list[[x]]$p_val_adj < 0.05, ])
, "valid"=tp_de_list.gse[[x]]$valid))}) %>% map(as.data.frame) %>% bind_rows()
rownames(deGenes.Res.tp.df) = names(tp_de_list)
deGenes.Res.tp.df

```

```{r}

plotGO = function(clusterID, expression, gseRes, invLFC, outfolder=NULL, name.title="GO enrich", name.ggo="ggo", name.ego="ego", name.ego_up="ego_up", name.ego_down="ego_down")
{
  
  cid = as.character(clusterID)
  
  
  
  if (!is.null(outfolder))
  {
    
    if ((!is.null(name.ggo)) && (!is.null(gseRes[[cid]][[name.ggo]])) && (nrow(gseRes[[cid]][[name.ggo]]@result) > 0))
    {
      write.xlsx(gseRes[[cid]][[name.ggo]]@result, paste(outfolder, paste(cid, "ggo.xlsx", sep="."), sep="/"))  
    }
    
    if ((!is.null(name.ego)) && (!is.null(gseRes[[cid]][[name.ego]])) && (nrow(gseRes[[cid]][[name.ego]]@result) > 0))
    {
      write.xlsx(gseRes[[cid]][[name.ego]]@result, paste(outfolder, paste(cid, "ego.xlsx", sep="."), sep="/"))  
    }
    
    if ((!is.null(name.ego_up)) && (!is.null(gseRes[[cid]][[name.ego_up]])) && (nrow(gseRes[[cid]][[name.ego_up]]@result) > 0))
    {
      write.xlsx(gseRes[[cid]][[name.ego_up]]@result, paste(outfolder, paste(cid, "ego_up.xlsx", sep="."), sep="/"))  
    }
    
    if ((!is.null(name.ego_down)) && (!is.null(gseRes[[cid]][[name.ego_down]])) && (nrow(gseRes[[cid]][[name.ego_down]]@result) > 0))
    {
      write.xlsx(gseRes[[cid]][[name.ego_down]]@result, paste(outfolder, paste(cid, "ego_down.xlsx", sep="."), sep="/"))  
    }
    
  }
  geneVec = expression[[cid]]$avg_logFC
  if ("gene" %in% colnames(expression[[cid]]))
  {
    names(geneVec) = expression[[cid]]$gene
  } else {
    names(geneVec) = rownames(expression[[cid]])
  }
  
  if (invLFC)
  {
    geneVec = -geneVec
  }
  
  subPlots = list()
  
  if (!is.null(name.ego))
  {
    subPlots[["enrich_go"]]=gseRes[[cid]][[name.ego]]
  }
  if 
  (!is.null(name.ego_up))
  {
    subPlots[["enrich_go_up"]]=gseRes[[cid]][[name.ego_up]]
  }
  
  if (!is.null(name.ego_down))
  {
    subPlots[["enrich_go_down"]]=gseRes[[cid]][[name.ego_down]]
  }
  
  subPlotPlots = list()
  
  for (gsePlotName in names(subPlots))
  {
      gsePlot = subPlots[[gsePlotName]]
      
      if (is.null(gsePlot))
      {
        next
      }
      
      print(gsePlotName)
      subsetPlotDF = gsePlot@result[gsePlot@result$p.adjust<0.05,]
      print(paste(nrow(subsetPlotDF), ncol(subsetPlotDF)))
      
      if ((nrow(subsetPlotDF) > 0) && (ncol(subsetPlotDF) > 1))
      {
        p=cnetplot(gsePlot, categorySize="pvalue", foldChange=geneVec, color="qvalue", colorEdge=T, showCategory=10) + ggtitle(paste("Cluster", cid, name.title, "(",gsePlotName,")"))
      
      if (!is.null(outfolder))
      {
  
        png(paste(outfolder, paste(cid, gsePlotName, "png", sep="."), sep="/"), width=25*100, height = 10*100)
        show(p)
        dev.off()
        
        pdf(paste(outfolder, paste(cid, gsePlotName, "pdf", sep="."), sep="/"), width=25, height = 10)
        show(p)
        dev.off()
      }
      
      subPlotPlots[[gsePlotName]] = p
      }
      

    
  }

  return(subPlotPlots)
  
}

makePlotsListsGO = function( obj.gse, obj.de, outfolder,name.ggo="ggo", name.ego="ego", name.ego_up="ego_up", name.ego_down="ego_down", name.title="GO enrich")
{
  dir.create(outfolder, showWarnings = FALSE, recursive=TRUE)

  
  return(
      lapply(names(obj.gse),
         function(x) {
           print(x)
            if (obj.gse[[x]]$valid)
            {
              print("calling plot")
              return(plotGO(x, obj.de, obj.gse, FALSE, outfolder,name.title=name.title, name.ggo=name.ggo, name.ego=name.ego, name.ego_up=name.ego_up, name.ego_down=name.ego_down))
            }
            
            return(NULL)
          }
  )
  )
}

```


```{r fig.height=6, fig.width=15}

makePlotsListsGO(asympt_sympt_de_list.go, asympt_sympt_de_list, "./go_sympts/")

```

```{r fig.height=6, fig.width=15}

makePlotsListsGO(tp_de_list.go, tp_de_list, "./go_tp/")

```

```{r fig.height=6, fig.width=15}

makePlotsListsGO(asympt_sympt_de_list.ra, asympt_sympt_de_list, "./ra_sympts/", name.ggo=NULL, name.ego="rao", name.ego_up="rao_up", name.ego_down="rao_down", name.title="ReactomePA")

```



```{r fig.height=6, fig.width=15}

makePlotsListsGO(tp_de_list.ra, tp_de_list, outfolder = "./ra_tp/", name.ggo=NULL, name.ego="rao", name.ego_up="rao_up", name.ego_down="rao_down", name.title="ReactomePA")

```

```{r}
plotGSE = function(clusterID, expression, gseRes, invLFC, outfolder=NULL)
{
  
  cid = as.character(clusterID)
  
  gsePlot = gseRes[[cid]]$gsesym

geneVec = expression[[cid]]$avg_logFC

if (invLFC)
{
  geneVec = -geneVec
}

if ("gene" %in% colnames(expression[[cid]]))
{
  names(geneVec) = expression[[cid]]$gene  
} else {
  names(geneVec) = rownames(expression[[cid]])
}


p=cnetplot(gsePlot, categorySize="pvalue", foldChange=geneVec, colorEdge=T, showCategory=10) + ggtitle(paste("Cluster", cid, "GSE"))

if (!is.null(outfolder))
{
  
  write.xlsx(gsePlot, paste(outfolder, paste(cid, "xlsx", sep="."), sep="/"))
  
  png(paste(outfolder, paste(cid, "png", sep="."), sep="/"), width=25*100, height = 10*100)
  show(p)
  dev.off()
  
  pdf(paste(outfolder, paste(cid, "pdf", sep="."), sep="/"), width=25, height = 10)
  show(p)
  dev.off()
}

return(p)
  
}

makePlotsListsGSE = function( obj.gse, obj.de, outfolder)
{
  
    dir.create(outfolder, showWarnings = FALSE, recursive=TRUE)

  return(
      lapply(names(obj.gse),
         function(x) {
           print(x)
            if ((obj.gse[[x]]$valid) && (nrow(obj.gse[[x]]$gsesym@result) > 0))
            {
              return(plotGSE(x, obj.de, obj.gse, FALSE, outfolder))
            }
            
            return(NULL)
          }
  )
  )
}

```

```{r fig.height=6, fig.width=15}

makePlotsListsGSE(asympt_sympt_de_list.gse, asympt_sympt_de_list, "./gse_sympts/")

```

```{r fig.height=6, fig.width=15}

makePlotsListsGSE(tp_de_list.gse, tp_de_list, "./gse_tp/")

```



```{r}
getCellCountDF = function(scobj, prefix="", group_by="orig.ident", select_by="idents", split_by=NULL, relative=F, outname=NULL,show.percent=F)
{
  
  allClusters = as.character(sort(unique(scobj[[select_by]][,])))
  
  if (!is.null(split_by))
  {
    allSplits = as.character(sort(unique(scobj[[split_by]][,])))
  }
  
  cellCounts = list()
  print(allClusters)
  for (clusterID in allClusters)
  {
    print(clusterID)
    
    if (!is.null(split_by))
    {
      clusterList = list()
      clusterList[["cluster"]] = clusterID
      cs = scobj[,scobj[[select_by]] == clusterID]
      clusterList[["all"]] = nrow(cs[[group_by]])

      for (split_cat in allSplits)
      {
          print(paste(clusterID, split_cat))
        
          cs = scobj[,scobj[[select_by]] == clusterID & scobj[[split_by]] == split_cat]
          allElems = table(cs[[group_by]])
          cs = NULL
        
          for (grp in names(allElems))
          {
            if (!is.null(prefix))
            {
              ngrp = paste(prefix, paste(split_cat, grp, sep="."), sep=".")
  
            } else {
              ngrp = paste(split_cat, grp, sep=".")
            }
            clusterList[[ngrp]] = allElems[[grp]]
          }
      }
      
    } else {
      
        cs = scobj[,scobj[[select_by]] == clusterID]
      
        allElems = table(cs[[group_by]])
        clusterList = list()
        clusterList[["cluster"]] = clusterID
        clusterList[["all"]] = nrow(cs[[group_by]])
        cs = NULL
        
        for (grp in names(allElems))
        {
          if (!is.null(prefix))
          {
            ngrp = paste(prefix, grp, sep=".")

          } else {
            ngrp = grp
          }
          clusterList[[ngrp]] = allElems[[grp]]
        }
    }
    

    
    
    cellCounts[[clusterID]] = clusterList
    
  }
  df_bysamplerep = cellCounts %>% map(as.data.frame) %>% bind_rows()
  df_bysamplerep[is.na(df_bysamplerep)] <- 0
  
  rownames(df_bysamplerep) = df_bysamplerep$cluster
  df_bysamplerep$cluster = NULL
  
  
  if (relative)
  {
    df_bysamplerep = sweep(df_bysamplerep,2,colSums(df_bysamplerep),"/")
    
    if (show.percent)
    {
      df_bysamplerep = df_bysamplerep*100;
    }
  }
  

  totals=t(colSums(df_bysamplerep))
  totals.df = data.frame(totals)
  rownames(totals.df) = "Total"
  df_bysamplerep=rbind(df_bysamplerep, totals.df)
  
  df_bysamplerep = cbind("cluster"=rownames(df_bysamplerep), df_bysamplerep)
  rownames(df_bysamplerep) = NULL
  
  if (!is.null(outname))
  {
    write.table(df_bysamplerep, file=outname, row.names = F,  quote=FALSE, sep='\t')
  }
  
  return(df_bysamplerep)
  #  
}

```


```{r fig.height=7, fig.width=8}

countByUCellname = getCellCountDF(obj.integrated, prefix="TP", select_by = "unified_cellnames", group_by="mpoint", split_by="disease_state", relative=TRUE, show.percent=T, outname="cellcounts_cellnames_diseasestate.tsv")
countByUCellname

head(colnames(subset(obj.integrated, mpoint==1 & disease_state == "CONTROL")))

cbc_cname = melt(countByUCellname)
cbc_cname$variable_factors = factor(cbc_cname$variable, levels=c("all", "TP.CONTROL.1","TP.ASYMPTOMATIC.1","TP.SYMPTOMATIC.1" , "TP.ASYMPTOMATIC.2", "TP.SYMPTOMATIC.2", "TP.ASYMPTOMATIC.3",  "TP.SYMPTOMATIC.3"))
cbc_cname = cbc_cname[!cbc_cname$cluster=="Total",]

p=DimPlot(obj.integrated, group.by = "unified_cellnames")

cbc_cname$cluster = factor(cbc_cname$cluster, levels=levels(p$data$unified_cellnames))
cbc_cname

plot_distribution_celltype = cbc_cname %>%
ggplot(aes(variable_factors, value, fill=cluster)) + 
geom_col()+#scale_fill_manual()+
geom_text_repel(aes(label = paste(round(value, digits = 1))),  size=2, position = position_stack(vjust = .5), direction ="both", segment.size  = 0.5, segment.color = "black", min.segment.length=0,
  arrow = arrow(length = unit(0.015, "cm"), angle = 0, type = "closed", ends = "first"), force=2) + ylab("Percentage")+xlab("Condition")+
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust=0.5)) + theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(), panel.background = element_blank(), axis.line = element_line(colour = "black"))

plot(plot_distribution_celltype)

save_plot(plot_distribution_celltype, "todoplots/1/frequency_celltype", fig.height=7, fig.width=8)

```



```{r fig.height=12, fig.width=8}

countByIdentSplitDisease = getCellCountDF(obj.integrated, prefix="TP", select_by = "idents", group_by="mpoint", split_by="disease_state", relative=TRUE, show.percent = T, outname="cellcounts_idents_mpoint_disease_state.tsv")
countByIdentSplitDisease

cbc_byidents = melt(countByIdentSplitDisease)

cbc_byidents$variable_factors = factor(cbc$variable, levels=c("all", "TP.CONTROL.1","TP.ASYMPTOMATIC.1","TP.SYMPTOMATIC.1" , "TP.ASYMPTOMATIC.2", "TP.SYMPTOMATIC.2", "TP.ASYMPTOMATIC.3",  "TP.SYMPTOMATIC.3"))

cbc_byidents = cbc_byidents[!cbc_byidents$cluster=="Total",]

p=DimPlot(obj.integrated, group.by = "idents")

cbc_byidents$cluster = factor(cbc_byidents$cluster, levels=levels(p$data$idents))
cbc_byidents

#mutate(cluster = fct_reorder(cluster, as.numeric(cluster))) %>%

plot_distribution_idents = cbc_byidents %>%
  ggplot(aes(variable_factors, value, fill=cluster)) + 
  geom_col()+#scale_fill_manual(values=cbPalette_byidents$hex)+
  geom_text_repel(aes(label = paste(round(value, digits = 1))),
                  size=2,
                  position = position_stack(vjust = .5),
                  direction ="both",
                  segment.size  = 0.2,
                  segment.color = "black",
                  min.segment.length=0,
                  arrow = arrow(length = unit(0.00001, "npc"), angle = 0, type = "closed", ends = "first"),
                  force=2) +
  ylab("Percentage")+
  xlab("Condition")+ 
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust=0.5)) + 
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(), panel.background = element_blank(), axis.line = element_line(colour = "black"))

save_plot(plot_distribution_idents, "todoplots/1/frequency_idents", fig.height=12, fig.width=8)

```


```{r rows.print=30}
cellCountByCellname = getCellCountDF(obj.integrated, select_by = "unified_cellnames")
cellCountByCellname


write.table(cellCountByCellname, file="cellcounts_by_samplename.tsv", row.names = F,  quote=FALSE, sep='\t')
```

```{r rows.print=30}
cellCountAll = getCellCountDF(obj.integrated)
cellCountAll

write.table(cellCountAll, file="cellcounts_by_origident.tsv", row.names = F,  quote=FALSE, sep='\t')
```


```{r rows.print=30}

cellCountCondition = getCellCountDF(obj.integrated, "condition")
cellCountCondition

write.table(cellCountCondition, file="cellcounts_by_condition.tsv", row.names = F,  quote=FALSE, sep='\t')
```

```{r rows.print=30}

cellCountConditionCellname = getCellCountDF(obj.integrated, "condition", "unified_cellnames")
cellCountConditionCellname

write.table(cellCountConditionCellname, file="cellcounts_by_condition_for_cellname.tsv", row.names = F,  quote=FALSE, sep='\t')
```

```{r}
save.image("/mnt/f/dev/data/Rprojs/covidSC/analysis_end.current.RData", compress = FALSE)

```


## Side By Side DotPlot

```{r}


```


```{r fig.height=10, fig.width=16}

plotElems = list()
plotElems[["Control"]] = list(cells=cells.control, label="Control")
plotElems[["Symptomatic"]] = list(cells=cells.symptomatic, label="Symptomatic")
plotElems[["Asymptomatic"]] = list(cells=cells.asymptomatic, label="Asymptomatic")

sbsPlot = makeSideBySideDotPlot(obj.integrated, plotElems = plotElems, featureGenes=unique(c("MT2A", "ISG15", "LY6E", "IFIT1", "IFIT2", "IFIT3", "IFITM1", "IFITM3", "IFI44L", "IFI6", "MX1", "MX2 IFI27; IFI44L", "IFI27", "RSAD2", "SIGLEC1", "IFIT1", "IS15", "IRF7", "IRF9", "XAF1")), group.by = "unified_cellnames",rel.width.legend=0.09)
plot(sbsPlot)

```

```{r fig.height=10, fig.width=11}

plotElems = list()
plotElems[["Symptomatic TP1"]] = list(cells=intersect(cells.symptomatic, cells.tp1), label="Symptomatic TP1")
plotElems[["Asymptomatic TP1"]] = list(cells=intersect(cells.asymptomatic, cells.tp1), label="Asymptomatic TP1")

sbsPlot = makeSideBySideDotPlot(obj.integrated, plotElems = plotElems, featureGenes=unique(c("MT2A", "ISG15", "LY6E", "IFIT1", "IFIT2", "IFIT3", "IFITM1", "IFITM3", "IFI44L", "IFI6", "MX1", "MX2 IFI27; IFI44L", "IFI27", "RSAD2", "SIGLEC1", "IFIT1", "IS15", "IRF7", "IRF9", "XAF1")), group.by = "unified_cellnames")
plot(sbsPlot)

```

```{r fig.height=10, fig.width=11}

plotElems = list()
plotElems[["Symptomatic TP2"]] = list(cells=intersect(cells.symptomatic, cells.tp2), label="Symptomatic TP2")
plotElems[["Asymptomatic TP2"]] = list(cells=intersect(cells.asymptomatic, cells.tp2), label="Asymptomatic TP2")

sbsPlot = makeSideBySideDotPlot(obj.integrated, plotElems = plotElems, featureGenes=unique(c("MT2A", "ISG15", "LY6E", "IFIT1", "IFIT2", "IFIT3", "IFITM1", "IFITM3", "IFI44L", "IFI6", "MX1", "MX2 IFI27; IFI44L", "IFI27", "RSAD2", "SIGLEC1", "IFIT1", "IS15", "IRF7", "IRF9", "XAF1")), group.by = "unified_cellnames")
plot(sbsPlot)

```

```{r fig.height=6, fig.width=10}

plotElems = list()
plotElems[["Control"]] = list(cells=intersect(cells.control, cells.tcell), label="Control")
plotElems[["Symptomatic"]] = list(cells=intersect(cells.symptomatic, cells.tcell), label="Symptomatic")
plotElems[["Asymptomatic"]] = list(cells=intersect(cells.asymptomatic, cells.tcell), label="Asymptomatic")

sbsPlot = makeSideBySideDotPlot(obj.integrated, plotElems = plotElems, featureGenes=unique(c("MT2A", "ISG15", "LY6E", "IFIT1", "IFIT2", "IFIT3", "IFITM1", "IFITM3", "IFI44L", "IFI6", "MX1", "MX2 IFI27; IFI44L", "IFI27", "RSAD2", "SIGLEC1", "IFIT1", "IS15", "IRF7", "IRF9", "XAF1")), group.by = "mpoint", rel.width.legend = 0.01, plot.title = "")
cowplot::plot_grid(ggdraw() + draw_label("T cells", fontface='bold'), sbsPlot, cols=1, rel_heights = c(0.1,0.9))

plot()

```

```{r fig.height=6, fig.width=10}

plotElems = list()
plotElems[["Control"]] = list(cells=cells.control, label="Control")
plotElems[["Symptomatic"]] = list(cells=cells.symptomatic, label="Symptomatic")
plotElems[["Asymptomatic"]] = list(cells=cells.asymptomatic, label="Asymptomatic")

sbsPlot = makeSideBySideDotPlot(obj.integrated, plotElems = plotElems, featureGenes=unique(c("CXCR6","CCR3","CCR2","FYCO1","LZTFL1","MAT2B","TNFSF15","OAS3","ICAM5","IFNAR2","DPP9")), group.by = "mpoint", rel.width.legend = 0.01, plot.title = "")
cowplot::plot_grid(ggdraw() + draw_label("", fontface='bold'), sbsPlot, cols=1, rel_heights = c(0.1,0.9))

plot()

```


```{r fig.height=6, fig.width=10}

plotElems = list()
plotElems[["TP1"]] = list(cells=intersect(cells.nk, cells.asymptomatic), label="NK ASYM")
plotElems[["TP2"]] = list(cells=intersect(cells.nk, cells.asymptomatic), label="NK ASYM")
plotElems[["TP3"]] = list(cells=intersect(cells.nk, cells.asymptomatic), label="NK ASYM")

sbsPlot = makeSideBySideDotPlot(obj.integrated, plotElems = plotElems, featureGenes=unique(c("GZMB", "S100A8", "S100A9", "LGALS1", "LGALS3")), group.by = "mpoint", rel.width.legend = 0.01, plot.title = "")
cowplot::plot_grid(ggdraw() + draw_label("NK cells", fontface='bold'), sbsPlot, cols=1, rel_heights = c(0.1,0.9))

plot()

#save_plot(sbsPlot_cd8_todo, "todoplots/15_leo/nk.asym.mpoint",fig.height=6, fig.width=10)


```
```{r}

dplt_15leo = DotPlot(subset(obj.integrated, cells=intersect(cells.nk, cells.asymptomatic)), group.by = "mpoint", features = unique(c("GZMB", "S100A8", "S100A9", "LGALS1", "LGALS3")), assay = "RNA") + xlab("Feature") + ylab("Timepoint")+theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust=0.5))
title <- ggdraw() + draw_label("Asympt NK Cells (by TPs)", fontface = 'bold')
dplt_15leo=cowplot::plot_grid(title, dplt_15leo, ncol = 1, rel_heights = c(0.1, 1))
dplt_15leo

save_plot(dplt_15leo, "todoplots/15_leo/genes.asympt.nk", fig.height = 5, fig.width = 10)

dplt_15leo = DotPlot(subset(obj.integrated, cells=intersect(cells.cd8, cells.asymptomatic)), group.by = "mpoint", features = unique(c("GZMB", "PRF1", "GZMH", "NKG7", "LYZ", "CTSW", "GNLY", "GZMK")), assay = "RNA") + xlab("Feature") + ylab("Timepoint")+theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust=0.5))
title <- ggdraw() + draw_label("Asympt CD8 Cells (by TPs)", fontface = 'bold')
dplt_15leo=cowplot::plot_grid(title, dplt_15leo, ncol = 1, rel_heights = c(0.1, 1))
dplt_15leo

save_plot(dplt_15leo, "todoplots/15_leo/genes.asympt.cd8", fig.height = 5, fig.width = 10)


dplt_15leo = DotPlot(subset(obj.integrated, cells=cells.tcell), group.by = "mpoint", features = unique(c("JUNB", "FOS", "FOSB", "GATA3" )), assay = "RNA") + xlab("Feature") + ylab("Timepoint")+theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust=0.5))
title <- ggdraw() + draw_label("T Cells (by TPs)", fontface = 'bold')
dplt_15leo=cowplot::plot_grid(title, dplt_15leo, ncol = 1, rel_heights = c(0.1, 1))
dplt_15leo

save_plot(dplt_15leo, "todoplots/15_leo/genes.tcell", fig.height = 5, fig.width = 10)


dplt_15leo = DotPlot(subset(obj.integrated, cells=intersect(cells.nk, cells.asymptomatic)), group.by = "mpoint", features = unique(c("FOS", "JUNB", "CCL4", "THEMIS2", "CCL3", "JUN", "FOSB")), assay = "RNA") + xlab("Feature") + ylab("Timepoint")+theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust=0.5))
title <- ggdraw() + draw_label("NK Cells (by TPs)", fontface = 'bold')
dplt_15leo=cowplot::plot_grid(title, dplt_15leo, ncol = 1, rel_heights = c(0.1, 1))
dplt_15leo

save_plot(dplt_15leo, "todoplots/15_leo/genes.nk", fig.height = 5, fig.width = 10)


dplt_15leo = DotPlot(subset(obj.integrated, cells=intersect(cells.monocyte, cells.asymptomatic)), group.by = "mpoint", features = unique(c("AREG", "EREG", "CD83", "TMEM176B", "EGR3", "HBEGF", "IL1RN" )), assay = "RNA") + xlab("Feature") + ylab("Timepoint") +theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust=0.5))
title <- ggdraw() + draw_label("Asympt Monocytes Cells (by TPs)", fontface = 'bold')
dplt_15leo=cowplot::plot_grid(title, dplt_15leo, ncol = 1, rel_heights = c(0.1, 1))
dplt_15leo

save_plot(dplt_15leo, "todoplots/15_leo/genes.monocytes", fig.height = 5, fig.width = 10)

```



And another round

```{r}

dplt_16leo = DotPlot(subset(obj.integrated, cells=intersect(cells.cd4t, cells.asymptomatic)), group.by = "mpoint", features = unique(c("JUNB", "FOS", "FOSB", "GATA3")), assay = "RNA") + xlab("Feature") + ylab("Timepoint")+theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust=0.5))
title <- ggdraw() + draw_label("Asympt CD4+ Cells (by TPs)", fontface = 'bold')
dplt_16leo=cowplot::plot_grid(title, dplt_16leo, ncol = 1, rel_heights = c(0.1, 1))

save_plot(dplt_16leo, "todoplots/16_leo/genes.asympt.cd4t", fig.height = 5, fig.width = 10)

dplt_16leo = DotPlot(subset(obj.integrated, cells=intersect(cells.tcell, cells.asymptomatic)), group.by = "mpoint", features = unique(c("JUNB", "FOS", "FOSB", "GATA3")), assay = "RNA") + xlab("Feature") + ylab("Timepoint")+theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust=0.5))
title <- ggdraw() + draw_label("Asympt T Cells (by TPs)", fontface = 'bold')
dplt_16leo=cowplot::plot_grid(title, dplt_16leo, ncol = 1, rel_heights = c(0.1, 1))

save_plot(dplt_16leo, "todoplots/16_leo/genes.asympt.tcell", fig.height = 5, fig.width = 10)

dplt_16leo = DotPlot(subset(obj.integrated, cells=intersect(cells.nk, cells.asymptomatic)), group.by = "mpoint", features = unique(c("FOS", "JUNB", "CCL4", "THEMIS2", "CCL2", "JUN", "FOSB", "GZMB", "S100A8", "S100A9", "LGALS1", "LGALS3")), assay = "RNA") + xlab("Feature") + ylab("Timepoint")+theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust=0.5))
title <- ggdraw() + draw_label("Asympt NK Cells (by TPs)", fontface = 'bold')
dplt_16leo=cowplot::plot_grid(title, dplt_16leo, ncol = 1, rel_heights = c(0.1, 1))

save_plot(dplt_16leo, "todoplots/16_leo/genes.asympt.nkcell", fig.height = 5, fig.width = 10)

dplt_16leo = DotPlot(subset(obj.integrated, cells=intersect(cells.monocyte, cells.asymptomatic)), group.by = "mpoint", features = unique(c("AREG", "EREG", "CD83", "TMEM176B", "EGR3", "HBEGF", "IL1RN" )), assay = "RNA") + xlab("Feature") + ylab("Timepoint")+theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust=0.5))
title <- ggdraw() + draw_label("Asympt Monocytes (by TPs)", fontface = 'bold')
dplt_16leo=cowplot::plot_grid(title, dplt_16leo, ncol = 1, rel_heights = c(0.1, 1))

save_plot(dplt_16leo, "todoplots/16_leo/genes.asympt.monocytes", fig.height = 5, fig.width = 10)

dplt_16leo = DotPlot(subset(obj.integrated, cells=intersect(cells.cd8, cells.symptomatic)), group.by = "mpoint", features = unique(c("CTSW", "PRF1", "NKG7", "GZMB", "GZMH", "GNLY")), assay = "RNA") + xlab("Feature") + ylab("Timepoint")+theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust=0.5))
title <- ggdraw() + draw_label("Sympt CD8+ Cells (by TPs)", fontface = 'bold')
dplt_16leo=cowplot::plot_grid(title, dplt_16leo, ncol = 1, rel_heights = c(0.1, 1))

save_plot(dplt_16leo, "todoplots/16_leo/genes.sympt.cd8cell", fig.height = 5, fig.width = 10)

```






```{r}
dplt_16leo = DotPlot(subset(obj.integrated, cells=intersect(cells.cd4t, cells.symptomatic)), group.by = "mpoint", features = unique(c("JUNB", "FOS", "FOSB", "GATA3")), assay = "RNA") + xlab("Feature") + ylab("Timepoint")+theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust=0.5))
title <- ggdraw() + draw_label("Sympt CD4+ Cells (by TPs)", fontface = 'bold')
dplt_16leo=cowplot::plot_grid(title, dplt_16leo, ncol = 1, rel_heights = c(0.1, 1))

save_plot(dplt_16leo, "todoplots/16_leo_rev/genes.sympt.cd4t", fig.height = 5, fig.width = 10)

dplt_16leo = DotPlot(subset(obj.integrated, cells=intersect(cells.tcell, cells.symptomatic)), group.by = "mpoint", features = unique(c("JUNB", "FOS", "FOSB", "GATA3")), assay = "RNA") + xlab("Feature") + ylab("Timepoint")+theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust=0.5))
title <- ggdraw() + draw_label("Sympt T Cells (by TPs)", fontface = 'bold')
dplt_16leo=cowplot::plot_grid(title, dplt_16leo, ncol = 1, rel_heights = c(0.1, 1))

save_plot(dplt_16leo, "todoplots/16_leo_rev/genes.sympt.tcell", fig.height = 5, fig.width = 10)

dplt_16leo = DotPlot(subset(obj.integrated, cells=intersect(cells.nk, cells.symptomatic)), group.by = "mpoint", features = unique(c("FOS", "JUNB", "CCL4", "THEMIS2", "CCL2", "JUN", "FOSB", "GZMB", "S100A8", "S100A9", "LGALS1", "LGALS3")), assay = "RNA") + xlab("Feature") + ylab("Timepoint")+theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust=0.5))
title <- ggdraw() + draw_label("Sympt NK Cells (by TPs)", fontface = 'bold')
dplt_16leo=cowplot::plot_grid(title, dplt_16leo, ncol = 1, rel_heights = c(0.1, 1))

save_plot(dplt_16leo, "todoplots/16_leo_rev/genes.sympt.nkcell", fig.height = 5, fig.width = 10)

dplt_16leo = DotPlot(subset(obj.integrated, cells=intersect(cells.monocyte, cells.symptomatic)), group.by = "mpoint", features = unique(c("AREG", "EREG", "CD83", "TMEM176B", "EGR3", "HBEGF", "IL1RN" )), assay = "RNA") + xlab("Feature") + ylab("Timepoint")+theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust=0.5))
title <- ggdraw() + draw_label("Sympt Monocytes (by TPs)", fontface = 'bold')
dplt_16leo=cowplot::plot_grid(title, dplt_16leo, ncol = 1, rel_heights = c(0.1, 1))

save_plot(dplt_16leo, "todoplots/16_leo_rev/genes.sympt.monocytes", fig.height = 5, fig.width = 10)

dplt_16leo = DotPlot(subset(obj.integrated, cells=intersect(cells.cd8, cells.asymptomatic)), group.by = "mpoint", features = unique(c("CTSW", "PRF1", "NKG7", "GZMB", "GZMH", "GNLY")), assay = "RNA") + xlab("Feature") + ylab("Timepoint")+theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust=0.5))
title <- ggdraw() + draw_label("Asympt CD8+ Cells (by TPs)", fontface = 'bold')
dplt_16leo=cowplot::plot_grid(title, dplt_16leo, ncol = 1, rel_heights = c(0.1, 1))

save_plot(dplt_16leo, "todoplots/16_leo_rev/genes.asympt.cd8cell", fig.height = 5, fig.width = 10)
```






```{r fig.height=8, fig.width=10}

plotElems = list()
plotElems[["Asympt"]] = list(cells=intersect(cells.monocyte, cells.asymptomatic), label="Monocytes ASYM")
plotElems[["Sympt"]] = list(cells=intersect(cells.monocyte, cells.symptomatic), label="Monocytes SYM")

sbsPlot = makeSideBySideDotPlot(obj.integrated, plotElems = plotElems, featureGenes=unique(c("AREG", "EREG", "CD83", "TMEM176B", "EGR3", "HBEGF", "IL1RN")), group.by = "mpoint", rel.width.legend = 0.01, plot.title = "")
cowplot::plot_grid(ggdraw() + draw_label("Monocytes", fontface='bold'), sbsPlot, ncol =1, rel_heights = c(0.1,0.9))

plot()

```


# Interferons

```{r}

interferonGenes = c("MT2A", "ISG15", "LY6E", "IFIT1", "IFIT2", "IFIT3", "IFITM1", "IFITM3", "IFI44L", "IFI6", "MX1", "IFI27",  "IFI44L", "RSAD2", "SIGLEC1", "IFIT1", "ISG15")

obj.integrated.data.ifn.small = GetAssayData(obj.integrated, assay = "RNA", slot = "data")
containedIfnGenes = intersect(interferonGenes, rownames(obj.integrated.data))
containedIfnGenes
obj.integrated.data.ifn.small = obj.integrated.data[containedIfnGenes, ]
dim(obj.integrated.data.ifn.small)
```



```{r}

interferonGenes = c("MX1","MX2","MT2A","RSAD2","SIGLEC1","IFIT1","IFI44","IFI27","IFITM10","IFIT3","IFI6","IFIT1P1","IFIT2","IFITM9P","IFI30","IFITM3","IFI44L","IFIT1B","IFITM5","IFITM3P2","IFITM3P9","IFI27L2","IFIT5","IFITM1","IFIH1","IFI16","IFI35","IFI27L1","IRF6","IRF7","IRF9","IRF5","IRF4","IRF2BP2","IRF2BPL","IRF3","IRF1-AS1","ISG15","ISG20L2","ISG20","XAF1","LY6E","IFITM2","IFITM3P6","IFITM3P3","IFIT6P","IRF2BP1","IRF5P1","IRF2","IRF1","IRF8","IFITM3P7","IFITM4P","IFITM3P1","IFITM3P8")

obj.integrated.data = GetAssayData(obj.integrated, assay = "RNA", slot = "data")
containedIfnGenes = intersect(interferonGenes, rownames(obj.integrated.data))
containedIfnGenes
obj.integrated.data.ifn = obj.integrated.data[containedIfnGenes, ]
dim(obj.integrated.data.ifn)

```
```{r}

ifn_score = colMeans(obj.integrated.data.ifn)
ifn_score_small_genelist = colMeans(obj.integrated.data.ifn.small)

```

```{r}

makeIFNScore = function(scobj, ifn_score)
{
  
  ifnScoreDFs = list()
subsetList = list("all"=colnames(scobj), "Monocytes"=cells.monocyte, "NK-cells"=cells.nk, "T-cells"=cells.tcell, "gdT-Cells"=cells.gdtcell, "Dendritic Cells"=cells.dendritic, "pDCs"=cells.pdc, "Plasma Cells"=cells.plasma, "CD8+ Cells"=cells.cd8, "CD4+ T-cells"=cells.cd4t, "B-cells"=cells.bcell)

for (sname in names(subsetList))
{
  print(sname)
  selCells = subsetList[[sname]]
  
  ifn_score.control = ifn_score[intersect(cells.control, selCells)]
  ifn_score.symptomatic = ifn_score[intersect(cells.symptomatic, selCells)]
  ifn_score.asymptomatic = ifn_score[intersect(cells.asymptomatic, selCells)]
  
  df.control = data.frame("cond"=rep("Control", length(ifn_score.control)), "score"=ifn_score.control)
  df.symptomatic = data.frame("cond"=rep("Symptomatic", length(ifn_score.symptomatic)), "score"=ifn_score.symptomatic)
  df.asymptomatic = data.frame("cond"=rep("Asymptomatic", length(ifn_score.asymptomatic)), "score"=ifn_score.asymptomatic)
  
  df.all = rbind(df.control, df.symptomatic, df.asymptomatic)
  
  df.all$mpoint = obj.integrated$mpoint[rownames(df.all)]
  df.all$tp = df.all$mpoint
  df.all[df.all$cond == "Control",]$tp = "0"
  
  ifnScoreDFs[[sname]] = df.all
}

return(ifnScoreDFs)

}


ifnScoreDFs = makeIFNScore(obj.integrated, ifn_score)
ifnScoreDFs_small = makeIFNScore(obj.integrated, ifn_score_small_genelist)

sIfnScore = ifn_score[names(obj.integrated$idents)]
names(sIfnScore) = names(obj.integrated$idents)

obj.integrated$ifn_score = sIfnScore

```
```{r}

allPlotDF = data.frame()

for (subsetName in names(ifnScoreDFs))
{
  
  
  plotDF = ifnScoreDFs[[subsetName]]

ctrlDF = plotDF[plotDF$cond == "Control", ]
plotDF = plotDF[plotDF$cond != "Control", ]

ctrlDF$cond = "Symptomatic"
plotDF = rbind(plotDF, ctrlDF)
ctrlDF$cond = "Asymptomatic"
plotDF = rbind(plotDF, ctrlDF)

plotDF$tp = factor(plotDF$tp, labels=c("0", "1", "2", "3"))
plotDF$cond = factor(plotDF$cond, labels=c("Asymptomatic", "Symptomatic"))
plotDF$celltype = subsetName

allPlotDF = rbind(allPlotDF, plotDF)


p = ggplot(plotDF, aes(x = tp, y = score, fill = cond)) +
  geom_boxplot(position = position_dodge(width = 0.9))+
  stat_summary(
    fun = median,
    geom = 'line',
    aes(group = cond, colour = cond),
    position = position_dodge(width = 0.9) #this has to be added
  ) + ggtitle(subsetName) + ylim(0,2.0)+theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
panel.background = element_blank(), axis.line = element_line(colour = "black"), text = element_text(size = 20)) + labs(x="TimePoint", y="ISG Score")
  

save_plot(p, paste("isg_longitudinal_plots/sympt_asympt_", str_replace_all(string = subsetName, pattern=" ", replacement="_"), sep=""), fig.width = 8, fig.height = 6)
}

allPlotDF$cell = rownames(allPlotDF)
allPlotDF
```


```{r fig.width=16, fig.height=12}

usePlotDF=NULL
usePlotDF = data.frame(allPlotDF)
usePlotDF$cond_ct = paste(usePlotDF$celltype, usePlotDF$cond, sep=" ")
usePlotDF = usePlotDF[is.element(usePlotDF$celltype, c("T-cells", "B-cells", "NK-cells", "Monocytes")),]
usePlotDF$cell = substring(usePlotDF$cell, 1, nchar(usePlotDF$cell)-1)
usePlotDF

celltype2color = list("T-cells"="#CD9600", "B-cells"="#00BE67", "Monocytes"="#F8766D", "NK-cells"="#7CAE00")

usePlotDF$color = as.character(unlist(celltype2color[usePlotDF$celltype]))

usePlotDF$celltype_fac=factor(usePlotDF$celltype, unique(usePlotDF$celltype))
use.colors = celltype2color[levels(usePlotDF$celltype_fac)]
names(use.colors) = levels(usePlotDF$celltype_fac)
usePlotDF$cond = as.character(usePlotDF$cond)
usePlotDF = usePlotDF[usePlotDF$tp!=0, ]

geneExpr = GetAssayData(obj.integrated, assay = "RNA", slot="data")

for (isGene in interferonGenes)
{
  usePlotDF[[isGene]] = geneExpr[isGene, ][usePlotDF$cell]
}
usePlotDF

allPlotDF.sympt = usePlotDF[usePlotDF$cond == "Symptomatic", ]
allPlotDF.asympt = usePlotDF[usePlotDF$cond == "Asymptomatic", ]


```

```{r fig.width=16, fig.height=12}


p = ggplot(allPlotDF.sympt, aes(x = tp, y = score, fill=celltype_fac, alpha=0.5)) +
  geom_boxplot(position = position_dodge(width = 0.9) )+scale_fill_manual(name = "Cell type", values = unlist(use.colors))+ stat_summary(
    fun = median,
    geom = 'smooth',
    aes(group = celltype_fac, colour = celltype_fac),
    position = position_dodge(width = 0.9) #this has to be added
  ) +scale_colour_manual(name = "Cell type", values = unlist(use.colors)) +
  ggtitle("Celltype and Condition ISG scores (Symptomatic)") + ylim(0,2.0) +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
panel.background = element_blank(), axis.line = element_line(colour = "black"), text = element_text(size = 20)) +
  labs(x="TimePoint", y="ISG Score")
plot(save_plot(p, paste("isg_longitudinal_plots/subset_sympt", sep=""), fig.width = 8, fig.height = 6))

  p = ggplot(allPlotDF.asympt, aes(x = tp, y = score, fill=celltype_fac)) +
  geom_boxplot(position = position_dodge(width = 0.9), alpha=0.5 )+ scale_fill_manual(name = "Cell type", values = unlist(use.colors)) + stat_summary(
    fun = median,
    geom = 'smooth',
    aes(group = celltype_fac, colour = celltype_fac),
    position = position_dodge(width = 0.9) #this has to be added
  ) + scale_colour_manual(name = "Cell type", values = unlist(use.colors)) +
    ggtitle("Celltype and Condition ISG scores (Asymptomatic)")+ ylim(0,2.0)+theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
panel.background = element_blank(), axis.line = element_line(colour = "black"), text = element_text(size = 20)) + labs(x="TimePoint", y="ISG Score")
plot(save_plot(p, paste("isg_longitudinal_plots/subset_asympt", sep=""), fig.width = 8, fig.height = 6))



plotGeneExpr = function( indf.sympt, indf.asympt, geneID)
{
  p = ggplot(indf.sympt, aes_string(x = "tp", y = geneID, fill="celltype_fac")) +
  geom_boxplot(position = position_dodge(width = 0.9), alpha=0.5 )+scale_fill_manual(name = "Cell type", values = unlist(use.colors))+ stat_summary(
    fun = median,
    geom = 'smooth',
    aes(group = celltype_fac, colour = celltype_fac),
    position = position_dodge(width = 0.9) #this has to be added
  ) +scale_colour_manual(name = "Cell type", values = unlist(use.colors)) +
  ggtitle(paste(geneID, " Expression (Symptomatic)", sep="")) + ylim(0,8.0) +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
panel.background = element_blank(), axis.line = element_line(colour = "black"), text = element_text(size = 20)) +
  labs(x="TimePoint", y="Expression")
plot(save_plot(p, paste("isg_longitudinal_plots/GeneExpr_sympt_", geneID, sep=""), fig.width = 8, fig.height = 6))

  p = ggplot(indf.asympt, aes_string(x = "tp", y = geneID, fill="celltype_fac")) +
  geom_boxplot(position = position_dodge(width = 0.9), alpha=0.5 )+ scale_fill_manual(name = "Cell type", values = unlist(use.colors)) + stat_summary(
    fun = median,
    geom = 'smooth',
    aes(group = celltype_fac, colour = celltype_fac),
    position = position_dodge(width = 0.9) #this has to be added
  ) + scale_colour_manual(name = "Cell type", values = unlist(use.colors)) +
    ggtitle(paste(geneID, " Expression (Asymptomatic)", sep=""))+ ylim(0,8.0)+theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
panel.background = element_blank(), axis.line = element_line(colour = "black"), text = element_text(size = 20)) + labs(x="TimePoint", y="Expression")
plot(save_plot(p, paste("isg_longitudinal_plots/GeneExpr_asympt_", geneID, sep=""), fig.width = 8, fig.height = 6))

}

for (isGene in interferonGenes)
{
plotGeneExpr(allPlotDF.sympt, allPlotDF.asympt, isGene)  
}


```

```{r fig.width=12, fig.height=36}
VlnPlot(obj.integrated, interferonGenes, pt.size=0, group.by="unified_cellnames", split.by="mpoint", ncol=2)
```


```{r}

p=DimPlot(obj.integrated, group.by="unified_cellnames")
g <- ggplot_build(p)
cellname2color = unique(data.frame(group=g$data[[1]]$group, color=g$data[[1]]$colour, cellnames=levels(obj.integrated$unified_cellnames)[g$data[[1]]$group]))
cellname2color
plot(p)

celltype2color = list("T-cells"="#CD9600", "B-cells"="#00BE67", "Monocytes"="#F8766D", "NK-cells"="#7CAE00")

```

```{r}

FeaturePlot(obj.integrated, features=c("ifn_score"), order=T)

```

```{r}
rownames(obj.integrated.data.ifn)
```




```{r}
setdiff(interferonGenes, containedIfnGenes)
```

```{r}
for (ct in unique(obj.integrated$celltypenames))
{
  print(paste(ct, length(obj.integrated$celltypenames[obj.integrated$celltypenames == ct])))
}
```

```{r}
for (ct in unique(obj.integrated$celltypenamestp))
{
  print(paste(ct, length(obj.integrated$celltypenamestp[obj.integrated$celltypenamestp == ct])))
}
```

```{r}

DimPlot(obj.integrated, group.by = "celltypenames", label=T)

```
```{r fig.width=30, fig.height=25}

VlnPlot(subset(obj.integrated, disease_state=="SYMPTOMATIC"), features=interferonGenes, group.by = "celltypenames", split.by = "mpoint", pt.size = 0, cols=mpCols)+ theme(legend.position = 'bottom')

```

```{r fig.width=30, fig.height=25}

VlnPlot(subset(obj.integrated, disease_state=="ASYMPTOMATIC"), features=interferonGenes, group.by = "celltypenames", split.by = "mpoint", pt.size = 0, cols=mpCols)+ theme(legend.position = 'bottom')

```

```{r fig.width=30, fig.height=25}

VlnPlot(obj.integrated, features=interferonGenes, group.by = "celltypenames", split.by = "mpoint", pt.size = 0, cols=mpCols)+ theme(legend.position = 'bottom')

```
```{r fig.width=30, fig.height=8}

mapal <- colorRampPalette(RColorBrewer::brewer.pal(9,"YlOrRd"))(128)

DoHeatmap(subset(obj.integrated, downsample=500), slot="data", features = interferonGenes, group.by = "celltypenamestp", disp.min=0, disp.max=4)+scale_fill_gradientn(colours = rev(mapal)) + theme(legend.position="bottom")


```

```{r, fig.height=10, fig.width=15}

ctname2cells = list(
  "T cells" = names(obj.integrated$celltypenames[obj.integrated$celltypenames=="T cells"]),
  "B cells" = names(obj.integrated$celltypenames[obj.integrated$celltypenames=="B cells"]),
  "Monocytes" = names(obj.integrated$celltypenames[obj.integrated$celltypenames=="Monocytes"]),
  "NK cells" = names(obj.integrated$celltypenames[obj.integrated$celltypenames=="NK cells"]),
  "Plasma cells" = names(obj.integrated$celltypenames[obj.integrated$celltypenames=="Plasma cells"]),
  "Plasmacytoid dendritic cells" = names(obj.integrated$celltypenames[obj.integrated$celltypenames=="Plasmacytoid dendritic cells"]),
  "gdT cells" = names(obj.integrated$celltypenames[obj.integrated$celltypenames=="gdT cells"]),
  "Dendritic cells" = names(obj.integrated$celltypenames[obj.integrated$celltypenames=="Dendritic cells"]),
  "CD8+ T cells" = names(obj.integrated$celltypenames[obj.integrated$celltypenames=="CD8+ T cells"]),
  "CD4+ T cells" = names(obj.integrated$celltypenames[obj.integrated$celltypenames=="CD4+ T cells"])
)

for (ctname in names(ctname2cells))
{
  
  p = DoHeatmap(subset(obj.integrated, cells=ctname2cells[[ctname]], downsample=500), slot="data", features = interferonGenes, group.by = "celltypenamestpstate", disp.min=0, disp.max=4)+scale_fill_gradientn(colours = rev(mapal)) + theme(legend.position="bottom")
  plot(p)
  
}

```

```{r fig.width=30, fig.height=50}


ctname2cells = list(
  "T cells" = names(obj.integrated$celltypenames[obj.integrated$celltypenames=="T cells"]),
  "B cells" = names(obj.integrated$celltypenames[obj.integrated$celltypenames=="B cells"]),
  "Monocytes" = names(obj.integrated$celltypenames[obj.integrated$celltypenames=="Monocytes"]),
  "NK cells" = names(obj.integrated$celltypenames[obj.integrated$celltypenames=="NK cells"]),
  "Plasma cells" = names(obj.integrated$celltypenames[obj.integrated$celltypenames=="Plasma cells"]),
  "Plasmacytoid dendritic cells" = names(obj.integrated$celltypenames[obj.integrated$celltypenames=="Plasmacytoid dendritic cells"]),
  "gdT cells" = names(obj.integrated$celltypenames[obj.integrated$celltypenames=="gdT cells"]),
  "Dendritic cells" = names(obj.integrated$celltypenames[obj.integrated$celltypenames=="Dendritic cells"]),
  "CD8+ T cells" = names(obj.integrated$celltypenames[obj.integrated$celltypenames=="CD8+ T cells"]),
  "CD4+ T cells" = names(obj.integrated$celltypenames[obj.integrated$celltypenames=="CD4+ T cells"])
)

selcells = list("ASYMPT.CTRL"=c(cells.asymptomatic, cells.control), "SYMPT.CTRL"=c(cells.symptomatic, cells.control), "ASYMPT"=cells.asymptomatic, "SYMPT"=cells.symptomatic)


for (ctname in names(ctname2cells))
{
  
  for (subsetname in names(selcells))
  {
      subsetcells = selcells[[subsetname]]

  
  p = VlnPlot(subset(obj.integrated, cells=intersect(ctname2cells[[ctname]], subsetcells)), features = interferonGenes, group.by = "celltypenamesstate", split.by = "mpoint", cols=mpCols, pt.size = 0)+theme(legend.position="bottom")
  
  
  title <- ggdraw() + draw_label(ctname, fontface = 'bold')

  cpg = cowplot::plot_grid(title, p, ncol = 1, rel_heights = c(0.05, 1))
  
  cowplot::save_plot(paste("violins_ifn/", subsetname, "_", str_replace_all(ctname, ";| |\\+", "_"), ".png", sep=""), cpg, base_width = 15, base_height = 45)
  
  }
  
}



```

```{r fig.height=6, fig.width=12}


for (ctname in names(ctname2cells))
{
  
  for (subsetname in names(selcells))
  {
      subsetcells = selcells[[subsetname]]
      
      p = DotPlot(subset(obj.integrated, cells=intersect(ctname2cells[[ctname]], subsetcells)), features = interferonGenes, group.by = "celltypenamestpstate", scale=T)
      p=p+ theme(axis.text.x = element_text(angle = 90, hjust=1, vjust = 0.5))
      p$guides$colour$title="Scaled Average Expression"
    
      title <- ggdraw() + draw_label(paste(ctname, subsetname, sep=" "), fontface = 'bold')
    
      cpg = cowplot::plot_grid(title, p, ncol = 1, rel_heights = c(0.05, 1))
      
      cowplot::save_plot(paste("violins_ifn/","dp_",subsetname, "_", str_replace_all(ctname, ";| |\\+", "_"), ".png", sep=""), cpg, base_width = 12, base_height = 6)
      
  }
  
}

```



```{r}
head(obj.integrated.data.ifn)
```


```{r}
library(psych)
```





```{r}
library(plyr)
library(ggpubr)
```



```{r}

makeIFNPlots = function(inDF, sname, subsetCells, outname)
{

  plotDF = inDF[subsetCells,]
  
  allConds = unique(plotDF$cond)
  
  print(allConds)
  
  for (cond in allConds)
  {
    #print(paste(sname, subsetCells, cond, summary(plotDF[plotDF$cond == cond,]$score)))
  }
  

  combs = combinations(length(allConds), 2, allConds, repeats.allowed = F)
  my_comparisons = lapply(seq_len(nrow(combs)), function(i) combs[i,])
  
  print(my_comparisons)
  
  p=ggboxplot(plotDF, x = "cond", y = "score",
          color = "cond", palette = dsrCols)+
  stat_compare_means(comparisons = my_comparisons, method = "t.test") + labs(title=paste("IFN-Score", sname))
  
  save_plot(p, paste(outname, "_boxplot", sep=""), fig.width = 8, fig.height = 6)
  plot(p)
  
  
  p=FeaturePlot(subset(obj.integrated, cells=subsetCells), features=c("ifn_score"), order=T) + labs(title=paste("IFN-Score", sname))
  save_plot(p, paste(outname, "_featureplot", sep=""), fig.width = 8, fig.height = 6)
  plot(p)
  
}

```

```{r}
summary(udf[udf$cond == "Symptomatic",]$score)

```

```{r}

dsrCols = c("Control"="#8B9693", "Asymptomatic"="#2C9B5C", "Symptomatic"="#FE4988")

printIFNData = function(inDF, sname, subsetCells, outname)
{

  plotDF = inDF[subsetCells,]
  
  allConds = unique(plotDF$cond)
  
  for (cond in sort(allConds))
  {
    writeLines(paste(str_replace_all(sname, " ", "_"), cond, paste(summary(plotDF[plotDF$cond == cond,]$score), collapse="\t"), sep='\t'))
  }
  
}

for (sname in names(ifnScoreDFs_small))
{
  ssname = str_replace_all(sname, " ", "_")

  inDF = ifnScoreDFs_small[[sname]]
  
  printIFNData(inDF, sname, subsetList[[sname]], paste("todoplots/13/", ssname, sep=""))
  
  for (mpoint in allmpoints)
  {
    
    mpointCells = names(obj.integrated$mpoint[obj.integrated$mpoint==mpoint])
    selCells = intersect(subsetList[[sname]], mpointCells)
    

    printIFNData(inDF, paste(sname, "TP", mpoint, sep="_"), selCells, paste("todoplots/13/tp", mpoint, ".", ssname, sep=""))


  
  }
  
}

```


```{r}
df.all
```


```{r}

dsCols
dsrCols = c("Control"="#8B9693", "Asymptomatic"="#2C9B5C", "Symptomatic"="#FE4988")

for (sname in names(ifnScoreDFs))
{
  ssname = str_replace_all(sname, " ", "_")
  print(paste(sname, ssname))
  
  inDF = ifnScoreDFs[[sname]]
  
  makeIFNPlots(inDF, sname, subsetList[[sname]], paste("todoplots/13/", ssname, sep=""))
  
  
  for (mpoint in allmpoints)
  {
    
    mpointCells = names(obj.integrated$mpoint[obj.integrated$mpoint==mpoint])
    selCells = intersect(subsetList[[sname]], mpointCells)
    
    print(paste("Cluster", sname , "for timepoint", mpoint, "with", length(selCells), "cells"))
    
    makeIFNPlots(inDF, paste(sname, "TP", mpoint), selCells, paste("todoplots/13/tp", mpoint, ".", ssname, sep=""))


  
  }
  
}

```
```{r}
ifnScoreDFs_small$all
```


```{r}
unique(ifnScoreDFs_small$all$cond)
```





```{r}
save.image("/mnt/f/dev/data/Rprojs/covidSC/analysis_end.currentifn.RData", compress = FALSE)
```

```{r}
saveRDS(obj.integrated, "obj.integrated.ifn_score.Rds")
```

# Time Analysis (DE)

```{r}
```


```{r}
allCellTypes = list(
  "Tcells" = names(obj.integrated$celltypenames[obj.integrated$celltypenames=="T cells"]),
  "Bcells" = names(obj.integrated$celltypenames[obj.integrated$celltypenames=="B cells"]),
  "Monocytes" = names(obj.integrated$celltypenames[obj.integrated$celltypenames=="Monocytes"]),
  "NKcells" = names(obj.integrated$celltypenames[obj.integrated$celltypenames=="NK cells"]),
  "Plasmacells" = names(obj.integrated$celltypenames[obj.integrated$celltypenames=="Plasma cells"]),
  "pDC" = names(obj.integrated$celltypenames[obj.integrated$celltypenames=="Plasmacytoid dendritic cells"]),
  "gdTcells" = names(obj.integrated$celltypenames[obj.integrated$celltypenames=="gdT cells"]),
  "DC" = names(obj.integrated$celltypenames[obj.integrated$celltypenames=="Dendritic cells"]),
  "CD8+Tcells" = names(obj.integrated$celltypenames[obj.integrated$celltypenames=="CD8+ T cells"]),
  "CD4+Tcells" = names(obj.integrated$celltypenames[obj.integrated$celltypenames=="CD4+ T cells"]),
  "all"=names(obj.integrated$celltypenames),
  "TcellsAny" = names(obj.integrated$unified_cellnames[obj.integrated$unified_cellnames=="T cells;Immune system"])
)

#allAddCellTypes = list("TcellsAny" = names(obj.integrated$unified_cellnames[obj.integrated$unified_cellnames=="T cells;Immune system"]))

lapply(allCellTypes, length)
allResults=list()
for (celltype in names(allCellTypes))
{
  
  cells.celltype = allCellTypes[[celltype]]
  
  ctrl.symp.tp1 = compareCells(obj.integrated,
                  intersect(cells.control, cells.celltype),
                  intersect(intersect(cells.symptomatic, cells.tp1), cells.celltype),
                  "control", "symptomatic", prefix=paste(celltype, "tp1", sep="."), test="t", outfolder="detests_time")
  ctrl.symp.tp2 = compareCells(obj.integrated,
                  intersect(cells.control, cells.celltype),
                  intersect(intersect(cells.symptomatic, cells.tp2), cells.celltype),
                  "control", "symptomatic", prefix=paste(celltype, "tp2", sep="."), test="t", outfolder="detests_time")
  ctrl.symp.tp3 = compareCells(obj.integrated,
                  intersect(cells.control, cells.celltype),
                  intersect(intersect(cells.symptomatic, cells.tp3), cells.celltype),
                  "control", "symptomatic", prefix=paste(celltype, "tp3", sep="."), test="t", outfolder="detests_time")

  
  ctrl.asymp.tp1 = compareCells(obj.integrated,
                  intersect(cells.control, cells.celltype),
                  intersect(intersect(cells.asymptomatic, cells.tp1), cells.celltype),
                  "control", "asymptomatic", prefix=paste(celltype, "tp1", sep="."), test="t", outfolder="detests_time")
  ctrl.asymp.tp2 = compareCells(obj.integrated,
                  intersect(cells.control, cells.celltype),
                  intersect(intersect(cells.asymptomatic, cells.tp2), cells.celltype),
                  "control", "asymptomatic", prefix=paste(celltype, "tp2", sep="."), test="t", outfolder="detests_time")
  ctrl.asymp.tp3 = compareCells(obj.integrated,
                  intersect(cells.control, cells.celltype),
                  intersect(intersect(cells.asymptomatic, cells.tp3), cells.celltype),
                  "control", "asymptomatic", prefix=paste(celltype, "tp3", sep="."), test="t", outfolder="detests_time")
  
  
  ctrl.covid.tp1 = compareCells(obj.integrated,
                  intersect(cells.control, cells.celltype),
                  intersect(cells.tp1, cells.celltype),
                  "control", "covid", prefix=paste(celltype, "tp1", sep="."), test="t", outfolder="detests_time")
  ctrl.covid.tp2 = compareCells(obj.integrated,
                  intersect(cells.control, cells.celltype),
                  intersect(cells.tp2, cells.celltype),
                  "control", "covid", prefix=paste(celltype, "tp2", sep="."), test="t", outfolder="detests_time")
  ctrl.covid.tp3 = compareCells(obj.integrated,
                  intersect(cells.control, cells.celltype),
                  intersect(cells.tp3, cells.celltype),
                  "control", "covid", prefix=paste(celltype, "tp3", sep="."), test="t", outfolder="detests_time")

  
  listSympt =  list("tp1"=ctrl.symp.tp1,  "tp2"=ctrl.symp.tp2,  "tp3"=ctrl.symp.tp3)
  listAsympt = list("tp1"=ctrl.asymp.tp1, "tp2"=ctrl.asymp.tp2, "tp3"=ctrl.asymp.tp3)
  listCovid = list("tp1"=ctrl.covid.tp1, "tp2"=ctrl.covid.tp2, "tp3"=ctrl.covid.tp3)

  
  allResults[[celltype]] = list("sympt"=listSympt, "asympt"=listAsympt, "covid"=listCovid)
  
}



saveRDS(allResults, "detests_time/de_results.Rds")



```

```{r}

allResultsBins = list()

for (celltype in names(allResults))
{
  celltype.logfcs = c()
short.results = list()

ctresult = allResults[[celltype]]

my_merge_outer <- function(df1, df2){                                # Create own merging function
  merge(df1, df2, by = "gene", all=TRUE)
}

for (condtp in names(ctresult))
{
  for (tp in names(ctresult[[condtp]]))
  {
      celltype.logfcs = c(celltype.logfcs, ctresult[[condtp]][[tp]]$avg_log2FC)
  }
}




createBins = function(indf, col.fc, col.sig, bounds)
{
  
  smalldf = indf[, c("gene", col.fc, col.sig)]
  colnames(smalldf) = c("gene", "fc", "sig")
  
  binFun = function(x)
  {

    fc = as.numeric(x[2])
    sig = as.numeric(x[3])
    
    #print(x)
    #print(bounds)
    
    if ((is.na(sig)) || is.nan(sig) ||(sig > 0.05))
    {
      return(0)
    }
    
    curBin = -(length(bounds)/2)
    
    if (fc >= bounds[length(bounds)])
    {
      return(length(bounds)/2)
    }
    
    for (i in 1:length(bounds))
    {
      #print(paste(i, curBin+i-1, fc, "<", bounds[i], fc < bounds[i]))
      if (fc < bounds[i])
      {
        return(curBin + i -1)
      }
    }
    
    print(x)
    return(10)
  }
    
  
  bins = apply(smalldf, 1, binFun)
  
  return(bins)
  
}

#testDF = data.frame(gene=c("gene1", "gene2", "gene3", "gene4", "gene5"),
#                    log2FC = c(-1, -0.55, 0.1, 0.55, 1),
#                    p_val_adj = c(0.01, 0.01, 0.01, 0.01,0.01)
#)
#createBins(testDF, "log2FC", "p_val_adj", c(-0.75, -0.5, 0.5, 0.75))


quantileVec = as.numeric(quantile(abs(celltype.logfcs), probs = c(0.25, 0.75)))
quantileVec = c(-rev(quantileVec), quantileVec)
quantileVec

print(celltype)
print(quantileVec)

allBinNames = c()

for (condtp in names(ctresult))
{
  
  condtp.results = list()
  for (tp in names(ctresult[[condtp]]))
  {
    
      binName = paste("bins", condtp, tp, sep=".")
      allBinNames = c(allBinNames, binName)
    
      celltype.logfcs = c(celltype.logfcs, ctresult[[condtp]][[tp]]$avg_log2FC)
      
      resDF = data.frame(gene=ctresult[[condtp]][[tp]]$gene)
      
      if (nrow(resDF) > 0)
      {
          resDF[[paste("avg_log2FC", condtp, tp, sep=".")]] = -(ctresult[[condtp]][[tp]]$avg_log2FC)
      resDF[[paste("p_val_adj", condtp, tp, sep=".")]] = ctresult[[condtp]][[tp]]$p_val_adj
      resDF[[paste("pct", condtp, tp, sep=".")]] = ctresult[[condtp]][[tp]]$pct.2
      

      resDF[[binName]] = bins = createBins(resDF, paste("avg_log2FC", condtp, tp, sep="."), paste("p_val_adj", condtp, tp, sep="."), quantileVec)
      
          condtp.results[[tp]] = resDF
      }


  }
  
  mergeDF = Reduce(my_merge_outer, condtp.results) 
  short.results[[condtp]] = mergeDF

}

celltype.results = Reduce(my_merge_outer, short.results)
celltype.results

print(celltype)
print(allBinNames)
print(grep("^bins.", colnames(celltype.results), value = T))

for (binName in allBinNames)
{
  
  if (! binName %in% colnames(celltype.results))
  {
    celltype.results[[binName]] = 0
  }
}

for (binID in grep("^bins.", colnames(celltype.results), value = T))
{
  celltype.results[[binID]][is.na(celltype.results[[binID]])] = 0
  
  for (condtp in names(short.results))
  {
      if (binID %in% colnames(short.results[[condtp]]))
      {
    short.results[[condtp]][[binID]][is.na(short.results[[condtp]][[binID]])] = 0    
      }
    
  }

}

allResultsBins[[celltype]] = celltype.results
}

```

```{r}
allResultsBins[["all"]][allResultsBins[["all"]]$gene %in% interferonGenes,]
```

```{r}
grep("^bins.", colnames(allResultsBins[["Tcells"]]), value = T)
```

```{r}

makeFuzzyPlot = function( resultsDF, selState = "sympt", logged=T, filterThresh=0, use.levels=list("-2"="DOWN", "-1"="down", "0"="No Reg.", "1"="up", "2"="UP"),map.labels=NULL, highlight=NULL, breaks=500, list.name=NULL, bins.name=NULL)
{
  
statePattern = paste("\\.", selState, sep="")

ctr = resultsDF[, c("gene", grep(statePattern, grep("^bins.", colnames(resultsDF), value = T), value = T))]
ctr$bins.0ctrl.tp0 = 0

if (!is.null(map.labels))
{
  map.labels = c("bins.0ctrl.tp0"="TP0", map.labels)
}

ctr = ctr[, c("gene", sort(grep("^bins.", colnames(ctr), value = T)))]
rownames(ctr) = ctr$gene
inGenes = ctr$gene
ctr$gene = NULL

#print(ctr)



num2str = use.levels

for (bin_col in grep("^bins.", colnames(ctr), value = T))
{
  
  strVals = sapply(ctr[[bin_col]], function(x){num2str[[as.character(x)]]})
  strVals = as.factor(strVals)
  strVals = factor(strVals, levels=as.character(use.levels))
  
  ctr[[bin_col]] = strVals
  
  
}

#print(ctr)

if (logged)
{
  nfunc = function(x){return(log(x))}
  nfuncState = paste("n > ",filterThresh,"; log(n); ", sep="")
  
  yLabText = "Logged Gene Count"
} else {
  nfunc = function(x){return(x)}
  nfuncState = paste("n > ",filterThresh,"; ", sep="")
  
  yLabText = "Gene Count"
}


  ctr[["highlighted"]] = 0
  
if (!is.null(bins.name))
{
  
  write.table(ctr, file=paste(bins.name, "ctr", "txt", sep="."))
  ctr2 = as.data.frame(ctr)
  ctr2$gene = rownames(ctr)
  write_xlsx(ctr2, paste(bins.name, "ctr", "xlsx", sep="."))
}

if (is.null(highlight))
{
  
  ctrFeature = ctr %>%
  dplyr::group_by(!!!syms(sort(grep("^bins.", colnames(ctr), value = T)))) %>%
  dplyr::summarise(
    n = nfunc(n())  ) %>% dplyr::filter(n>nfunc(filterThresh)) %>% dplyr::arrange(desc(n))

} else {
  
  ctr[highlight, c("highlighted")] = 1
  
  ctrFeature = ctr %>%
  dplyr::group_by(!!!syms(c(sort(grep("^bins.", colnames(ctr), value = T)), "highlighted"))) %>%
  dplyr::summarise(
    n = nfunc(n())  ) %>% dplyr::filter(highlighted==1 | n>nfunc(filterThresh)) %>% dplyr::arrange(desc(-highlighted,n))

  print(ctrFeature[ctrFeature$highlighted == 1,])
  
}


allgroups = paste("group", 1:nrow(ctrFeature), sep="")
allgroups = as.factor(allgroups)
allgroups = factor(allgroups, levels=paste("group", 1:nrow(ctrFeature), sep=""))

ctrFeature$group = allgroups

#print(ctrFeature)


if (!is.null(list.name))
{
  grpgenes = list()
  binCols = sort(grep("^bins.", colnames(ctrFeature), value = T))
  for (i in 1:nrow(ctrFeature)) {
  
      
      grpVec = as.vector(ctrFeature[i, binCols])
      grp = as.character(ctrFeature[i, "group"])
      
      for (j in 1:nrow(ctr))
      {
          rVec = as.vector(ctr[j,binCols])
          
          #print(sum(rVec == grpVec))
          if (sum(rVec == grpVec) == length(binCols))
          {
              #print(paste(grp, j, inGenes[j]))
              #print()
              grpgenes[[grp]] = c(grpgenes[[grp]], inGenes[j])
          }
      }
  }
    

    resgrps = list()
    fname = paste(list.name, ".txt", sep="")
    print(fname)
        write(paste(binCols, collapse=", "), fname)

  for (grp in names(grpgenes))
  {
    
      subdf = ctrFeature[ctrFeature$group == paste("group", grp, sep=""), binCols]
      subdf <- sapply(subdf[, binCols], as.character)
      
      dfvec = as.vector(subdf)
      print(dfvec)
      
      resgrps[[grp]] = as.character(paste(paste("group", grp, sep=""), paste(dfvec, collapse=", "), paste(grpgenes[[grp]], collapse=", ")))
  
      write(paste(resgrps[[grp]]), fname, append=TRUE)
  
      
  }
   
}



aesMappings = list("y"="n")
i=1
for (axisCol in sort(grep("^bins.", colnames(ctr), value = T)))
{
  axisName = paste("axis", as.character(i), sep="")
  aesMappings[[axisName]] = axisCol
  i=i+1
}


#print(aesMappings)
print(create_aes(aesMappings))
print(nrow(ctrFeature))

getPalette = colorRampPalette(brewer.pal(10, "Set3"))

colorValues = getPalette(nrow(ctrFeature))
colorValues[ctrFeature$highlighted == 1] = "#000000" #"#FF0000"

ctrFeature$highcolor = colorValues
#print(ctrFeature)

nLabels = length(grep("^bins.", colnames(ctrFeature), value = T))

if (! is.null(map.labels))
{
  use.labels = map.labels[grep("^bins.", colnames(ctrFeature), value = T)]
} else {
  use.labels = grep("^bins.", colnames(ctrFeature), value = T)
}

geneCount = sum(ctrFeature$n)

p=ggplot(as.data.frame(ctrFeature), create_aes(aesMappings)) +
  geom_alluvium(aes(fill = group, colour = group), width = 0, knot.pos = 0, reverse = FALSE, curve_type="quintic") +
  scale_fill_manual(values = colorValues)+
  scale_color_manual(values = colorValues)+
  guides(colour=FALSE,fill=FALSE) +
  geom_stratum(width = 1/16, reverse = FALSE) +
  geom_text(stat = "stratum", aes(label = after_stat(stratum)), reverse = FALSE, angle=90) +
  scale_x_continuous(breaks = 1:nLabels, labels = use.labels)+
  scale_y_continuous(limits = c(0, geneCount), breaks = seq(0, geneCount, by = breaks))+
  labs(
    title=paste("DE Genes per TimePoint (binned logFCs; ", nfuncState, selState, ")", sep="")
  )+
  xlab("Timepoints")+
  ylab(yLabText)+
  theme(
    panel.background = element_blank(),
    axis.text.y = element_text(face="bold", size=14),
    axis.title.y = element_text(face="bold", size=24),
    axis.text.x = element_text(face="bold", size=14),
    axis.title.x = element_text(face="bold", size=24),
  )

return(p)
}


```

```{r}

  isgSubsetDF = allResultsBins[["all"]]
  isgSubsetDF = isgSubsetDF[isgSubsetDF$gene %in% interferonGenes,]
  
  
p1=makeFuzzyPlot(isgSubsetDF, selState = "asympt", list.name=paste(pname, "asympt", sep="_"), filterThresh = 0, logged = F, map.labels=list("bins.asympt.tp1"="TP1", "bins.asympt.tp2"="TP2", "bins.asympt.tp3"="TP3"), breaks=5)
p2=makeFuzzyPlot(isgSubsetDF, selState = "sympt", list.name=paste(pname, "sympt", sep="_"), filterThresh = 0, logged = F, map.labels=list("bins.sympt.tp1"="TP1", "bins.sympt.tp2"="TP2", "bins.sympt.tp3"="TP3"), breaks=5)

p = cowplot::plot_grid(
  p1,p2,
  ncol = 2
)
p

```


```{r}

for (celltypeName in names(allResultsBins))
{
  
  isgSubsetDF = allResultsBins[[celltypeName]]
  isgSubsetDF = isgSubsetDF[isgSubsetDF$gene %in% interferonGenes,]
  
pname = paste("fuzzyfinal_isg/asympt_sympt_isg", "abs", celltypeName, sep="_")
  
  lname = paste(pname, "asympt", sep="_")
p1=makeFuzzyPlot(isgSubsetDF, selState = "asympt", bins.name=lname, list.name=lname, filterThresh = 0, logged = F, map.labels=list("bins.asympt.tp1"="TP1", "bins.asympt.tp2"="TP2", "bins.asympt.tp3"="TP3"), breaks=5)

lname = paste(pname, "sympt", sep="_")
p2=makeFuzzyPlot(isgSubsetDF, selState = "sympt", bins.name=lname, list.name=lname, filterThresh = 0, logged = F, map.labels=list("bins.sympt.tp1"="TP1", "bins.sympt.tp2"="TP2", "bins.sympt.tp3"="TP3"), breaks=5)

p = cowplot::plot_grid(
  p1,p2,
  ncol = 2
)

save_plot(p, pname, fig.width=24, fig.height = 8)

  
pname = paste("fuzzyfinal_isg/asympt_isg", "abs", celltypeName, sep="_")
lname = paste(pname, "asympt", sep="_")
p1=makeFuzzyPlot(isgSubsetDF, selState = "asympt", bins.name=lname, list.name=lname, filterThresh = 0, logged = F, map.labels=list("bins.asympt.tp1"="TP1", "bins.asympt.tp2"="TP2", "bins.asympt.tp3"="TP3"), breaks=5)
save_plot(p1, pname, fig.width=12, fig.height = 8)


pname = paste("fuzzyfinal_isg/sympt_isg", "abs", celltypeName, sep="_")
lname = paste(pname, "sympt", sep="_")

p1=makeFuzzyPlot(isgSubsetDF, selState = "sympt", bins.name=lname, list.name=lname, filterThresh = 0, logged = F, map.labels=list("bins.sympt.tp1"="TP1", "bins.sympt.tp2"="TP2", "bins.sympt.tp3"="TP3"), breaks=5)
save_plot(p1, pname, fig.width=12, fig.height = 8)




pname = paste("fuzzyfinal_isg/covid_isg", "abs", celltypeName, sep="_")
lname = paste(pname, "covid", sep="_")

p1=makeFuzzyPlot(isgSubsetDF, selState = "covid", bins.name=lname, list.name=lname, filterThresh = 0, logged = F, map.labels=list("bins.covid.tp1"="TP1", "bins.covid.tp2"="TP2", "bins.covid.tp3"="TP3"), breaks=5)

save_plot(p1, pname, fig.width=12, fig.height = 8)
}


```


```{r fig.width=24, fig.height=8}

for (celltypeName in names(allResultsBins))
{
  
  lname = paste("fuzzyfinal/asympt_sympt_highlighted", "asympt",celltypeName, sep="_")

p1=makeFuzzyPlot(allResultsBins[[celltypeName]], selState = "asympt",bins.name=lname, filterThresh = 10, logged = F, highlight=interferonGenes,map.labels=list("bins.asympt.tp1"="TP1", "bins.asympt.tp2"="TP2", "bins.asympt.tp3"="TP3"))

  lname = paste("fuzzyfinal/asympt_sympt_highlighted", "sympt",celltypeName, sep="_")

p2=makeFuzzyPlot(allResultsBins[[celltypeName]], selState = "sympt",bins.name=lname, filterThresh = 10, logged = F, highlight=interferonGenes,map.labels=list("bins.sympt.tp1"="TP1", "bins.sympt.tp2"="TP2", "bins.sympt.tp3"="TP3"))

p = cowplot::plot_grid(
  p1,p2,
  ncol = 2
)

save_plot(p, paste("fuzzyfinal/asympt_sympt_highlighted", "abs", celltypeName, sep="_"), fig.width=24, fig.height = 8)

}

```



```{r fig.width=12, fig.height=8}

for (celltypeName in names(allResultsBins))
{
  
    lname = paste("fuzzyfinal/sympt", "abs", celltypeName, sep="_")

  p=makeFuzzyPlot(allResultsBins[[celltypeName]], bins.name = lname, selState = "sympt", filterThresh = 10, logged=F, map.labels=list("bins.sympt.tp1"="TP1", "bins.sympt.tp2"="TP2", "bins.sympt.tp3"="TP3"))
save_plot(p, paste("fuzzyfinal/sympt", "abs", celltypeName, sep="_"), fig.width=12, fig.height = 8)

    lname = paste("fuzzyfinal/asympt", "abs", celltypeName, sep="_")

    p=makeFuzzyPlot(allResultsBins[[celltypeName]], bins.name = lname, selState = "asympt", filterThresh = 10, logged=F, map.labels=list("bins.asympt.tp1"="TP1", "bins.asympt.tp2"="TP2", "bins.asympt.tp3"="TP3"))
save_plot(p, paste("fuzzyfinal/asympt", "abs", celltypeName, sep="_"), fig.width=12, fig.height = 8)

    lname = paste("fuzzyfinal/covid", "abs", celltypeName, sep="_")

    p=makeFuzzyPlot(allResultsBins[[celltypeName]], bins.name = lname, selState = "covid", filterThresh = 10, logged=F, map.labels=list("bins.covid.tp1"="TP1", "bins.covid.tp2"="TP2", "bins.covid.tp3"="TP3"))
save_plot(p, paste("fuzzyfinal/covid", "abs", celltypeName, sep="_"), fig.width=12, fig.height = 8)


    lname = paste("fuzzyfinal/sympt", "log", celltypeName, sep="_")

    p=makeFuzzyPlot(allResultsBins[[celltypeName]], bins.name = lname, selState = "sympt", filterThresh = 10, logged=T, map.labels=list("bins.sympt.tp1"="TP1", "bins.sympt.tp2"="TP2", "bins.sympt.tp3"="TP3"))
save_plot(p, paste("fuzzyfinal/sympt", "log", celltypeName, sep="_"), fig.width=12, fig.height = 8)


    lname = paste("fuzzyfinal/asympt", "log", celltypeName, sep="_")

    p=makeFuzzyPlot(allResultsBins[[celltypeName]], bins.name = lname, selState = "asympt", filterThresh = 10, logged=T, map.labels=list("bins.asympt.tp1"="TP1", "bins.asympt.tp2"="TP2", "bins.asympt.tp3"="TP3"))
save_plot(p, paste("fuzzyfinal/asympt", "log", celltypeName, sep="_"), fig.width=12, fig.height = 8)


    lname = paste("fuzzyfinal/covid", "log", celltypeName, sep="_")

    p=makeFuzzyPlot(allResultsBins[[celltypeName]], bins.name = lname, selState = "covid", filterThresh = 10, logged=T, map.labels=list("bins.covid.tp1"="TP1", "bins.covid.tp2"="TP2", "bins.covid.tp3"="TP3"))
save_plot(p, paste("fuzzyfinal/covid", "log", celltypeName, sep="_"), fig.width=12, fig.height = 8)

}

```

```{r fig.width=12, fig.height=8}

p=makeFuzzyPlot(celltype.results, selState = "asympt", filterThresh = 10, logged = F)
p

```

```{r fig.width=15, fig.height=8}

umap_cellnames_plot2 = DimPlot(obj.integrated, group.by = "celltypenames", label=T)
save_plot(umap_cellnames_plot2, outname = "fuzzyplots/umap_celltypenames_label", fig.width=15, fig.height=8)

umap_cellnames_plot2 = DimPlot(obj.integrated, group.by = "celltypenames", label=F)
save_plot(umap_cellnames_plot2, outname = "fuzzyplots/umap_celltypenames", fig.width=15, fig.height=8)

umap_cellnames_plot2 = DimPlot(obj.integrated, group.by = "unified_cellnames", label=T)
save_plot(umap_cellnames_plot2, outname = "fuzzyplots/umap_unified_cellnames_label", fig.width=15, fig.height=8)

umap_cellnames_plot2 = DimPlot(obj.integrated, group.by = "unified_cellnames", label=F)
save_plot(umap_cellnames_plot2, outname = "fuzzyplots/umap_unified_cellnames", fig.width=15, fig.height=8)


umap_cellnames_plot2 = DimPlot(obj.integrated, group.by = "idents", label=T)
save_plot(umap_cellnames_plot2, outname = "fuzzyplots/umap_idents_label", fig.width=15, fig.height=8)

umap_cellnames_plot2 = DimPlot(obj.integrated, group.by = "idents", label=F)
save_plot(umap_cellnames_plot2, outname = "fuzzyplots/umap_idents", fig.width=15, fig.height=8)

```

```{r}

createFuzzyBins = function(indf, col.fc, col.sig, bounds)
{

  smalldf = indf[, c("gene", col.fc, col.sig)]
  colnames(smalldf) = c("gene", "fc", "sig")

  binFun = function(x)
  {

    fc = as.numeric(x[2])
    sig = as.numeric(x[3])

    #print(x)
    #print(bounds)

    if ((is.na(sig)) || is.nan(sig) ||(sig > 0.05) || is.na(fc) || is.nan(fc))
    {
      return(0)
    }

    curBin = 0

    if (fc > bounds[length(bounds)])
    {
      return(length(bounds))
    }

    for (i in 1:length(bounds))
    {
      #print(paste(i, curBin+i-1, fc, "<", bounds[i], fc < bounds[i]))
      if (fc < bounds[i])
      {
        return(curBin + i -1)
      }
    }

    print(x)
    return(10)
  }


  bins = apply(smalldf, 1, binFun)

  return(bins)

}


```



```{r}
typeIinterferones = c("IFNE", "IFNK", "IFNA1", "IFNA2", "IFNA4", "IFNA5", "IFNA6", "IFNA7", "IFNA8", "IFNA10", "IFNA13", "IFNA14", "IFNA16", "IFNA17", "IFNA21","IFNB1", "IFNW1", "IL6")
typeIIinterferones = c("IFNG")
typeIIIinterferones = c("IFNL1", "IFNL2", "IFNL3")

typeInterferones = list("typeI"=typeIinterferones, "typeII"=typeIIinterferones, "typeIII"=typeIIIinterferones)

typeI_ISG = c("ISG15", "IRF1", "IRF2", "IRF7", "IRF8", "IRF9", "IFIT1", "IFIT2", "IFIT3", "IFITM1", "IFITM2", "IFITM3", "MX1", "IFI44L", "IFI16", "LY6E", "OAS1", "PKR", "BST2", "CXCL10", "ZAP", "RSAD2")
typeII_ISG = c("IFNGR1", "IFNGR2", "IFNAR1", "IFNAR2", "ICAM1", "CXCL9", "NOS2")

typeISGs = list("typeI"=typeI_ISG, "typeII"=typeII_ISG, "kami"=interferonGenes)

```

```{r}
for (typeIFN in names(typeInterferones))
{
  ifnGenes = typeInterferones[[typeIFN]]
  
  print(typeIFN)
  print(ifnGenes)
  print(paste(length(ifnGenes), length(intersect(ifnGenes, rownames(obj.integrated)))))

}


for (typeIFN in names(typeISGs))
{
  ifnGenes = typeISGs[[typeIFN]]
  
  print(typeIFN)
  print(ifnGenes)
  intersectionISG = intersect(ifnGenes, rownames(obj.integrated))
  print(paste(length(ifnGenes), length(intersectionISG), paste(intersectionISG, collapse=",")))

}

```

```{r}
ctrlTPs = obj.integrated$mpoint
ctrlTPs[obj.integrated$disease_state == "CONTROL"] = "0"
obj.integrated$TimePoint = ctrlTPs
```

```{r}

tp0123Colors = c(color.control, color.tp1, color.tp2, color.tp3)

```


```{r fig.width=6, fig.height=6}

for (pgene in c("ICAM1", "IFNGR1", "IFNGR2", "IFNAR1", "IFNAR2"))
{
  p = VlnPlot(subset(obj.integrated, unified_cellnames == "Monocytes;Immune system" & disease_state %in% c("CONTROL", "SYMPTOMATIC")), pgene, pt.size=0, group.by="unified_cellnames", cols = tp0123Colors, split.by="TimePoint")+ylim(0,3.5) + geom_boxplot(color="grey", alpha=0.4, position =position_dodge(width = 0.9)) + stat_summary(fun=mean, geom="point", aes(group=split), position=position_dodge(.9), color="black", size=4) + ggtitle(paste("CTRL+SYMPT", pgene))
  save_plot(p, paste("ifng_story/vlnplot_mono_sympt_", pgene, sep=""), 6,6)
  
  p=plot(VlnPlot(subset(obj.integrated, unified_cellnames == "Monocytes;Immune system" & disease_state %in% c("CONTROL", "ASYMPTOMATIC")), pgene, pt.size=0, group.by="unified_cellnames",cols = tp0123Colors, split.by="TimePoint")+ylim(0,3.5)  ) + geom_boxplot(color="grey", alpha=0.4, position =position_dodge(width = 0.9)) + stat_summary(fun=mean, geom="point", aes(group=split), position=position_dodge(.9), color="black", size=4) + ggtitle(paste("CTRL+ASYMPT", pgene))
  save_plot(p, paste("ifng_story/vlnplot_mono_asympt_", pgene, sep=""), 6,6)

  }
p
```


```{r}
allResultsBins[[celltypeName]][allResultsBins[[celltypeName]]$gene == "ICAM1", ]
```


```{r}

isgI = read.table("interferomes/typeI_isg_human_all", sep="\t", header = T)
isgII = read.table("interferomes/typeII_isg_human_all", sep="\t", header = T)
isgIII = read.table("interferomes/typeIII_isg_human_all", sep="\t", header = T)

isgI_only_genes = setdiff(isgI$Gene.Name, union(isgII$Gene.Name, isgIII$Gene.Name))
isgII_only_genes = setdiff(isgII$Gene.Name, union(isgI$Gene.Name, isgIII$Gene.Name))
isgIII_only_genes = setdiff(isgIII$Gene.Name, union(isgII$Gene.Name, isgI$Gene.Name))

print(paste("isgI", length(isgI_only_genes)))
print(paste("isgII", length(isgII_only_genes)))
print(paste("isgIII", length(isgIII_only_genes)))

isgOnly = list("typeI_isg"=isgI_only_genes, "typeII_isg"=isgII_only_genes, "typeIII_isg"=isgIII_only_genes)
isgInterferome = list("typeI_isg_overlap"=isgI$Gene.Name, "typeII_isg_overlap"=isgII$Gene.Name, "typeIII_isg_overlap"=isgIII$Gene.Name)


```

```{r}
isgZiegler = list("ziegler_ifng"= c("WARS", "CD74", "GBP1", "LAP3", "TAP1", "CXCL11", "PSMB9","UBE2L6","PSMB8","IFI30","STAT1","PSME2","PSMB10","B2M","CXCL10","IFI35","ISG20","GUK1","LGALS17A","S,C25A1","DUSP1","APOL2","IDO1","SLC16A3","PSMA4","PSME1","GBP2","PSMA5","PSMA2","LDHA","HLA-DRA","C20orf24","ISG15"), "ziegler_ifna"=c("ISG15","MX1","IFIT1","IFITM1","IFI44L","OAS1","IFI35","IFIT3","UBE2L6","RP11-396M11.1","ISG20","LAP3","XAF1","STAT1","OASL","USP18","OAS2","MX2","OAS3","CTD-2561J22.2","DDX58","EPSTI1","GBP1","LGALS9","RSAD2","HPX","DANCR"))

isgZiegler$ziegler_ifng = c(isgZiegler$ziegler_ifng, c("IFNGR1", "IFNGR2", "IFNAR1", "IFNAR2", "ICAM1", "CXCL9", "NOS2"))

```


```{r}


for (typeIFN in names(typeISGs))
{
  celltypeName = "Monocytes"
  
  ifnGenes = typeISGs[[typeIFN]]
  
  print(typeIFN)
  print(paste(length(ifnGenes), length(intersect(ifnGenes, rownames(obj.integrated))), length(intersect(ifnGenes, allResultsBins[[celltypeName]]$gene))))

  fullFcDF = allResultsBins[[celltypeName]]
  fcDF = fullFcDF[fullFcDF$gene %in% ifnGenes,]



  lname = paste("rev2_ffcs/sympt_fc_", typeIFN, "_", celltypeName, sep="")

  p=makeFuzzyPlot(fcDF, bins.name = lname, selState = "sympt", filterThresh = 0, logged=F, map.labels=list("bins.sympt.tp1"="TP1", "bins.sympt.tp2"="TP2", "bins.sympt.tp3"="TP3"))
save_plot(p, lname, fig.width=12, fig.height = 8)

  lname = paste("rev2_ffcs/asympt_fc_", typeIFN, "_", celltypeName, sep="")

  p=makeFuzzyPlot(fcDF, bins.name = lname, selState = "asympt", filterThresh = 0, logged=F, map.labels=list("bins.asympt.tp1"="TP1", "bins.asympt.tp2"="TP2", "bins.asympt.tp3"="TP3"))
save_plot(p, lname, fig.width=12, fig.height = 8)
  
  
fcDF[is.na(fcDF)] = 0
fcDF

symptCols = grep("\\.sympt\\.", colnames(fcDF), value=T)
asymptCols = grep("\\.asympt\\.", colnames(fcDF), value=T)

ffcDF = NULL
ffcDFColNames = c("gene")
celltype.logfcs = c()

for (tpName in c("tp1", "tp2", "tp3"))
{
  
  symptCol = paste("avg_log2FC", "sympt", tpName, sep=".")
  asymptCol = paste("avg_log2FC", "asympt", tpName, sep=".")
  
  newLogFC = fcDF[,symptCol]-fcDF[,asymptCol]
  
  celltype.logfcs = c(celltype.logfcs, fullFcDF[, symptCol]-fullFcDF[, asymptCol])
  
  col1name = paste("avg_log2FC", "ffc", tpName, sep=".")
  col2name = paste("p_val_adj", "ffc", tpName, sep=".")
  
  ffcDFColNames = c(ffcDFColNames, col1name, col2name)
  
  if (is.null(ffcDF))
  {
    ffcDF = data.frame(fcDF$gene,
                       newLogFC,
                       newLogFC*0)
  } else {
    ffcDF = cbind(ffcDF, newLogFC)
    ffcDF = cbind(ffcDF, newLogFC*0)
  }
  
  colnames(ffcDF) = ffcDFColNames
}

colnames(ffcDF) = ffcDFColNames
rownames(ffcDF) = rownames(fcDF)


quantileVec = as.numeric(quantile(abs(celltype.logfcs), probs = c(0.25, 0.75), na.rm=T))
quantileVec = c(-rev(quantileVec), quantileVec)
print(quantileVec)

for (tpName in c("tp1", "tp2", "tp3"))
{
  col3name = paste("bins", "ffc", tpName, sep=".")

  binVec = createBins(ffcDF, paste("avg_log2FC.ffc", tpName, sep="."),
                  paste("p_val_adj.ffc", tpName, sep="."), quantileVec)
  ffcDF = cbind(ffcDF, binVec)
  
  ffcDFColNames = c(ffcDFColNames, col3name)
  
}

colnames(ffcDF) = ffcDFColNames
ffcDF

  lname = paste("rev2_ffcs/ffc_", typeIFN, "_", celltypeName, sep="")
  
  ffc.levels=list("-2"="NON-PNEU", "-1"="non-pneu", "0"="No Reg.", "1"="pneu", "2"="PNEU")

  p=makeFuzzyPlot(ffcDF, bins.name = lname, selState = "ffc", filterThresh = 0, logged=F,use.levels=ffc.levels, map.labels=list("bins.ffc.tp1"="TP1", "bins.ffc.tp2"="TP2", "bins.ffc.tp3"="TP3"))
plot(save_plot(p, lname, fig.width=12, fig.height = 8))

write.table(ffcDF[, grep("gene|avg_log2FC|bins", colnames(ffcDF), value=T)], file=paste(lname, "_ffcdf.txt", sep=""), row.names = F)

}


```


```{r fig.width=12, fig.height=8}

p=makeFuzzyPlot(celltype.results[celltype.results$gene %in% interferonGenes,], selState = "sympt", filterThresh = 0,logged = F)

save_plot(p, "fuzzyplots/sympt_isg", fig.width=12, fig.height = 8)

```

```{r fig.width=12, fig.height=8}
p = makeFuzzyPlot(celltype.results[celltype.results$gene %in% interferonGenes,], selState = "asympt", filterThresh = 0,logged = F)

save_plot(p, "fuzzyplots/asympt_isg", fig.width=12, fig.height = 8)

```


```{r}

save.image("/mnt/f/dev/data/Rprojs/covidSC/analysis_end.temporal.ffc.RData", compress = FALSE)


```


```{r}

getExprDataPop = function(markerObj, markerCells, sampleSuffix, slot="data")
{
  expTable = GetAssayData(object = subset(x=markerObj, cells=markerCells), slot = slot)
  allgenes = rownames(expTable)
  cellnames = colnames(expTable)
  expt.r = as(expTable, "dgTMatrix")
  expt.df = data.frame(r = expt.r@i + 1, c = expt.r@j + 1, x = expt.r@x)
  DT <- data.table(expt.df)
  res = DT[, as.list(makesummary(x, sampleSuffix)), by = r]
  res[[paste("anum", sampleSuffix, sep=".")]] = length(cellnames)
  res$gene = allgenes[res$r]
  
  res = res[,r:=NULL]
  
  return(res)
}

get_population_expression_data = function(scobj, group, outname)
{
  
  exprData = list()
  for (cellPop in unique(as.character(unlist(scobj[[group]]))))
  {
    print(cellPop)
    varPop = str_to_lower( str_replace_all(
                            str_replace_all(#
                              str_replace_all( cellPop, "\\(|\\)| |,", "_"),
                              "__", "_"),
                            "_$", "")
                           )
    
    
    cellPopCells = rownames(scobj[[group]][scobj[[group]] == cellPop])
  
    exprData[[varPop]] = getExprDataPop(markerObj=scobj, markerCells=cellPopCells, sampleSuffix=varPop, slot="counts")
  }
  
  
  meanExprData = list()
  
  for (name in names(exprData))
  {
    
    exprDf = as.data.frame(exprData[[name]])
    subdf = exprDf[ ,c("gene", paste("mean", name, sep=".")) ]
  
    meanExprData[[name]] = subdf
  }
  
  cellnames_manualExprDF = Reduce(function(x,y) merge(x = x, y = y, by = "gene"), meanExprData)
  
  write.table(cellnames_manualExprDF, file = paste(outname, ".tsv", sep=""), quote=FALSE, sep = "\t", row.names = F)
  write_xlsx( cellnames_manualExprDF, path = paste(outname, ".xlsx", sep="") )
  
  return(cellnames_manualExprDF)
  
  
}
manualExprDF = get_population_expression_data(obj.integrated, "celltypenames", "population_expr/exprdf_celltypenames")

manualExprDF = get_population_expression_data(subset(obj.integrated, disease_state="SYMPTOMATIC"), "celltypenames", "population_expr/exprdf_celltypenames_sympt")
manualExprDF = get_population_expression_data(subset(obj.integrated, disease_state="ASYMPTOMATIC"), "celltypenames", "population_expr/exprdf_celltypenames_asympt")


```















